---
title: "TF-lib pDNA insert sequencing"
author: "Max Trauernicht"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output: 
  prettydoc::html_pretty:
    theme: leonids
    highlight: github
  #   toc: true
  #   toc_float: true
  #   code_folding: show
  # editor_options:
  #   chunk_output_type: console
---


*knitr document van Steensel lab*

# TF-lib pDNA insert sequencing

# Introduction
I sequenced the complete insert of the pDNA library of pMT09. I already extracted all sequences in front of the 3' adapter from the sequences data and added counts to identical sequences by starcode. I now want to make an overview about how many pDNA insert sequences in the pDNA still match the designed inserts. 


```{r setup, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# Load options and libraries
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(seqinr)
library(ShortRead)
library(plyr)
library(maditr)
library(phylotools)
library(tidyr)
library(readr)
library(dplyr)
library(ggplot2)
library(ggbeeswarm)
library(vwr)
library(d3r)
library(sunburstR)
library(LncFinder)
library(plotly)
library(tibble)
library(stringr)

```


```{r, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
### Custom functions
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}

ReadFasta<-function(file) {
   # Read the file line by line
   fasta<-readLines(file)
   # Identify header lines
   ind<-grep(">", fasta)
   # Identify the sequence lines
   s<-data.frame(ind=ind, from=ind+1, to=c((ind-1)[-1], length(fasta)))
   # Process sequence lines
   seqs<-rep(NA, length(ind))
   for(i in 1:length(ind)) {
      seqs[i]<-paste(fasta[s$from[i]:s$to[i]], collapse="")
   }
   # Create a data frame 
   DF<-data.frame(name=gsub(">", "", fasta[ind]), sequence=seqs)
   # Return the data frame as a result object from the function
   return(DF)
}
```


```{r knits setup, echo=FALSE, warning= FALSE, message=FALSE}
# Save all figures generated in this Rmd document
library(knitr)
filename <- SetFileName("_figures","mt")
dir.create(paste("results/", filename, sep = ""), showWarnings = FALSE)
opts_chunk$set(fig.width = 4, fig.height = 4, 
               dev=c('png', 'pdf'), fig.path = file.path(paste("results/", filename, "/", sep = ""))) 
pdf.options(useDingbats = FALSE)
```



```{r, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
## Data import
# Import reference sequences
ref_seq <- ReadFasta("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/gen-2/mt20210326_oligo_pool_gen2.fasta")

# Remove adapters from reference sequence (cause these are not in the sequencing data)
ref_seq$sequence <- gsub("CGGAGCGAACCGAGTTAG", "", ref_seq$sequence)
ref_seq$sequence <- gsub("CATCGTCGCATCCAAGAG", "", ref_seq$sequence)

# Split up in insert and barcode part
## In my case, the barcode should be the last 13 bases of the sequence
ref_seq$barcode <- gsub(".*([A-Z]{13})$", "\\1", ref_seq$sequence)
ref_seq$insert <- gsub("(.*)[A-Z]{13}$", "\\1", ref_seq$sequence)

# Add control group classifier
match <- c("romanov", "pMT02", "TF-seq", "ctrl", "promega")
ref_seq$ctrl <- "no"
ref_seq$ctrl[grep(paste(match, collapse = "|"), ref_seq$name)] <- "yes"

# Import sequencing files
pDNA_seq_files = list.files('/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/pDNA_insert_seq_2/results/',
                       full.names=T, patter='.*TF.*_counts.tsv')
pDNA_seq <- lapply(pDNA_seq_files, fread, header = FALSE)
names(pDNA_seq)<- gsub('.*//(.*?)_[CGAT]{8}.*_counts.tsv', 
                                    '\\1', 
                                    pDNA_seq_files)

# Generate wide df - each condition attached as new column
for (i in 1:length(pDNA_seq)) {
  if (i == 1) {
  pDNA_seq_df <- data.frame(pDNA_seq[i])
  pDNA_seq_df[3] <- names(pDNA_seq[i])
  names(pDNA_seq_df) <- c("sequence", "count", "name")
  pDNA_seq_df <- reshape2::dcast(pDNA_seq_df, sequence ~ name, value.var = "count")
  }
  else {
  pDNA_seq_df_i <- data.frame(pDNA_seq[i])
  pDNA_seq_df_i[3] <- names(pDNA_seq[i])
  names(pDNA_seq_df_i) <- c("sequence", "count", "name")
  pDNA_seq_df_i <- reshape2::dcast(pDNA_seq_df_i, sequence ~ name, value.var = "count")
  pDNA_seq_df <- merge(pDNA_seq_df, pDNA_seq_df_i, all = T)
  }
}



# Convert to long df - write conditions under each other
pDNA_seq <- melt(pDNA_seq_df, id.vars = "sequence",
              variable.name = "condition", value.name = "counts")


# Split up in insert and barcode part
## In my case, the barcode should be the last 13 bases of the sequence
pDNA_seq$barcode <- gsub(".*([A-Z]{13})$", "\\1", pDNA_seq$sequence)
pDNA_seq$insert <- gsub("(.*)[A-Z]{13}$", "\\1", pDNA_seq$sequence)

# Calculate reads per million
pDNA_seq$counts[is.na(pDNA_seq$counts)] <- 0
for (i in unique(pDNA_seq$condition)) {
  pDNA_seq$rpm[pDNA_seq$condition == i] <- (pDNA_seq$counts[pDNA_seq$condition == i] + 1) / # Adds a pseudocount of 1
    sum(pDNA_seq$counts[pDNA_seq$condition == i]) *1e6
}
```


# Analysis

## What is the barcode distribution of mapped vs. unmapped for both TFs?
```{r read_distribution_per_tf, fig.width=4, fig.height=4, fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Match barcodes with original data
pDNA_seq_bc <- pDNA_seq %>% 
  dplyr::select(barcode, rpm, condition) %>% 
  unique() %>% 
  mutate(id = "pDNA")

# Only keep highest barcode values - a bit of cheating here
pDNA_seq_bc$cond_bc <- paste(pDNA_seq_bc$barcode, pDNA_seq_bc$condition, sep = "_")
pDNA_seq_bc <- pDNA_seq_bc[order(pDNA_seq_bc$cond_bc, -abs(pDNA_seq_bc$rpm) ), ]
pDNA_seq_bc <- pDNA_seq_bc[ !duplicated(pDNA_seq_bc$cond_bc), ]
ref_seq_bc <- ref_seq %>% dplyr::select(name, barcode, ctrl)
match_bc <- merge(ref_seq_bc, pDNA_seq_bc, by = "barcode", all = T)
match_bc <- match_bc[!is.na(match_bc$rpm),]
match_bc$TF <- gsub("_.*", "\\1", match_bc$name)
match_bc$match <- "true"
match_bc$match[is.na(match_bc$name)] <- "false"
match_bc$conf <- "high"
match_bc$conf[match_bc$rpm < 10] <- "low"

# Visualize barcode distribution per TF
ggplot(match_bc %>%
         filter(ctrl == "no"), aes(x = TF, y = rpm)) +
  geom_quasirandom() +
  theme_bw() +
  xlab("reporter matches to:") +
  ylab("reads per million") +
  theme(text = element_text(size = 14), axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, size = 6)) +
  ylim(0,300) + facet_wrap(~condition)

# Plot per TF the amount of high vs. low confidence reporters
x <- match_bc %>%
  filter(match == "true", ctrl == "no") %>%
  mutate(conf_count = as.numeric(ave(TF, TF, conf, condition, FUN = function(x) length(x)))) %>%
  dplyr::select(conf_count, TF, conf, condition) %>%
  unique %>%
  dcast(TF + condition ~ conf, value.var = "conf_count")
x[is.na(x)] <- 0
x <- x %>%
  mutate(sum = high + low,
         percent = (high/sum)*100)

ggplot(x, 
       aes(x = reorder(TF, -percent), y = percent)) +
  geom_bar(stat = "identity", position = "dodge") +
  theme_bw() + 
  theme(text = element_text(size = 14), axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, size = 6)) +
  facet_wrap(~condition)



# Visualize read distribution for all matched reporters
ggplot(match_bc %>%
         filter(match == "true", ctrl == "no"), aes(x = rpm)) +
  geom_density() +
  geom_vline(aes(xintercept = 10), linetype = "dashed")+
  theme_bw() +
  xlim(0, 200) +
  facet_wrap(~condition)

ggplot(match_bc %>%
         filter(match == "true", ctrl == "no"), aes(x = rpm, color = condition)) +
  geom_density() +
  geom_vline(aes(xintercept = 10), linetype = "dashed")+
  theme_bw() +
  scale_color_brewer(palette = "Dark2") +
  xlim(0, 200)

ggplot(match_bc %>%
         filter(match == "true", ctrl == "no", str_detect(condition, "plasmid")), aes(x = rpm, color = condition)) +
  geom_density() +
  geom_vline(aes(xintercept = 10), linetype = "dashed")+
  theme_bw() +
  scale_color_brewer(palette = "Dark2") +
  xlim(0, 200)
```

## Correlate to GC contenct
```{r gc_content, fig.width=4, fig.height=4, fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Load reference file
ref_seq_2 <- seqinr::read.fasta("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr//data/library_design/output/mt20200619_oligo_pool_deep.fasta")

# Compute gc contents
gc <- compute_GC(ref_seq_2)

# Plot gc distribution
density <- density(gc$GC.content)

plot_ly(x = ~density$x, y = ~density$y, type = 'scatter', mode = 'lines', fill = 'tozeroy') %>% 
  layout(xaxis = list(title = 'GC content'),
         yaxis = list(title = 'Density'))

match_seq <- match_bc[!is.na(match_bc$name),]
gc <- gc %>% rownames_to_column(var = "name")
gc <- merge(gc, match_seq)


plot_ly(data = gc, x = ~GC.content, y = ~rpm, color = ifelse(gc$rpm >= 10, "black", "red")) %>% 
  layout(xaxis = list(title = 'GC content'),
         yaxis = list(title = 'reads per million',
                      range=c(0,230)))

```

## Plot how many barcodes are found in pDNA data
```{r barcodes_found, fig.width=4, fig.height=4, fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Filter pDNA data - we should have at least 10 rpm - we want to be sure that we're looking at relevant things
pDNA_seq_filt <- pDNA_seq[pDNA_seq$rpm >= 10,]

# Match barcodes with original data
ref_seq_bc <- ref_seq %>% dplyr::select(name, barcode)
ref_seq_bc$TF <- gsub("_.*", "\\1", ref_seq_bc$name)
pDNA_seq_bc <- pDNA_seq_filt %>% 
  filter(str_detect(condition, "plasmid")) %>% 
  dplyr::select(barcode, rpm) %>% 
  unique() %>% 
  mutate(id = "pDNA")
match_bc <- merge(ref_seq_bc, pDNA_seq_bc, by = "barcode", all = T)

## Identify the unmapped fraction
match_df_ref <- match_bc[!is.na(match_bc$name),]
n_match <- nrow(match_df_ref[!is.na(match_df_ref$id),])
n_nomatch <- nrow(match_df_ref[is.na(match_df_ref$id),])


# Create donut chart
data <- data.frame(
  lbls=c("matched:", "unmatched:"),
  count=c(n_match, n_nomatch)
)
 
# Compute percentages
data$fraction <- data$count / sum(data$count)
data$percentage <- data$fraction * 100


# Compute the cumulative percentages (top of each rectangle)
data$ymax <- cumsum(data$fraction)


# Compute the cumulative percentages (top of each rectangle)
data$ymax <- cumsum(data$fraction)

# Compute the bottom of each rectangle
data$ymin <- c(0, head(data$ymax, n=-1))

# Compute label position
data$labelPosition <- (data$ymax + data$ymin) / 2

# Compute a good label
data$label <- paste0(data$lbls, "\n", round(data$percentage), "%")

# Make the plot
ggplot(data, aes(ymax=ymax, ymin=ymin, xmax=4, xmin=3, fill=lbls)) +
  geom_rect() +
  geom_text(x=2, aes(y=labelPosition, label=label, color=lbls), size=5) + # x here controls label position (inner / outer)
  scale_fill_manual(values = c("#1B998B", "#2D3047")) +
  scale_color_manual(values = c("#1B998B", "#2D3047")) +
  coord_polar(theta="y") +
  xlim(c(0, 4)) +
  theme_void() +
  theme(legend.position = "none")
```


## Plot how many reads match to designed barcodes
```{r, barcode_reads, fig.width=4, fig.height=4, fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Match barcodes with original data
pDNA_seq_bc <- pDNA_seq_filt %>% 
  filter(str_detect(condition, "plasmid")) %>% 
  dplyr::select(barcode, rpm) %>% 
  unique() %>% 
  mutate(id = "pDNA")
ref_seq_bc <- ref_seq %>% dplyr::select(name, barcode)
ref_seq_bc$TF <- gsub("_.*", "\\1", ref_seq_bc$name)
match_bc <- merge(ref_seq_bc, pDNA_seq_bc, by = "barcode", all = T)

## Identify the unmapped fraction
match_df_ref <- match_bc
match_df_ref$rpm[is.na(match_df_ref$rpm)] <- 0
n_match <- sum(match_df_ref$rpm[!is.na(match_df_ref$id)])
n_nomatch <- sum(match_df_ref$rpm[is.na(match_df_ref$id)])


# Create donut chart
data <- data.frame(
  lbls=c("matched:", "unmatched:"),
  count=c(n_match, n_nomatch)
)
 
# Compute percentages
data$fraction <- data$count / sum(data$count)
data$percentage <- data$fraction * 100


# Compute the cumulative percentages (top of each rectangle)
data$ymax <- cumsum(data$fraction)


# Compute the cumulative percentages (top of each rectangle)
data$ymax <- cumsum(data$fraction)

# Compute the bottom of each rectangle
data$ymin <- c(0, head(data$ymax, n=-1))

# Compute label position
data$labelPosition <- (data$ymax + data$ymin) / 2

# Compute a good label
data$label <- paste0(data$lbls, "\n", round(data$percentage), "%")

# Make the plot
ggplot(data, aes(ymax=ymax, ymin=ymin, xmax=4, xmin=3, fill=lbls)) +
  geom_rect() +
  geom_text(x=2, aes(y=labelPosition, label=label, color=lbls), size=5) + # x here controls label position (inner / outer)
  scale_fill_manual(values = c("#1B998B", "#2D3047")) +
  scale_color_manual(values = c("#1B998B", "#2D3047")) +
  coord_polar(theta="y") +
  xlim(c(0, 4)) +
  theme_void() +
  theme(legend.position = "none")
```


## How many raw complete sequences match with the design?
```{r sequences_found, fig.width=4, fig.height=4, fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Match barcodes with original data
ref_seq_seq <- ref_seq %>% dplyr::select(name, sequence)
pDNA_seq_seq <- pDNA_seq_filt %>% 
  filter(str_detect(condition, "plasmid")) %>% 
  dplyr::select(sequence, rpm) %>% 
  unique() %>% 
  mutate(id = "pDNA")
match_seq <- merge(ref_seq_seq, pDNA_seq_seq, by = "sequence", all = T)
match_seq$TF <- gsub("_.*", "\\1", match_seq$name)

## Identify the unmapped fraction
match_df_ref_seq <- match_seq[!is.na(match_seq$name),]
n_match <- nrow(match_df_ref_seq[!is.na(match_df_ref_seq$id),])
n_nomatch <- nrow(match_df_ref_seq[is.na(match_df_ref_seq$id),])

# Create donut chart
data <- data.frame(
  lbls=c("matched:", "unmatched:"),
  count=c(n_match, n_nomatch)
)
 
# Compute percentages
data$fraction <- data$count / sum(data$count)
data$percentage <- data$fraction * 100


# Compute the cumulative percentages (top of each rectangle)
data$ymax <- cumsum(data$fraction)


# Compute the cumulative percentages (top of each rectangle)
data$ymax <- cumsum(data$fraction)

# Compute the bottom of each rectangle
data$ymin <- c(0, head(data$ymax, n=-1))

# Compute label position
data$labelPosition <- (data$ymax + data$ymin) / 2

# Compute a good label
data$label <- paste0(data$lbls, "\n", round(data$percentage), "%")

# Make the plot
ggplot(data, aes(ymax=ymax, ymin=ymin, xmax=4, xmin=3, fill=lbls)) +
  geom_rect() +
  geom_text(x=2, aes(y=labelPosition, label=label, color=lbls), size=5) + # x here controls label position (inner / outer)
  scale_fill_manual(values = c("#1B998B", "#2D3047")) +
  scale_color_manual(values = c("#1B998B", "#2D3047")) +
  coord_polar(theta="y") +
  xlim(c(0, 4)) +
  theme_void() +
  theme(legend.position = "none")
```






## Now we want to know the read distribution of matched/unmatched sequences
```{r reads_matched_sequences, fig.width=4, fig.height=4, fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
## Only select barcodes from design in pDNA data
pDNA_seq_insert <- pDNA_seq[pDNA_seq$barcode %in% ref_seq$barcode,] %>% dplyr::select(sequence, number)
ref_seq_insert <- ref_seq %>% dplyr::select(name, sequence)
match_insert <- merge(pDNA_seq_insert, ref_seq_insert, by = "sequence", all = T)
match_insert$TF[!is.na(match_insert$number)] <- gsub("_.*", "\\1", match_insert$name[!is.na(match_insert$number)])

ggplot(match_insert, aes(x = TF, y = number)) +
  geom_quasirandom(alpha = 0.1) +
  xlab("TF reporter match")+
  theme_bw() +
  ylim(0,1000)

## Identify the unmapped fraction
match_df_reads <- match_insert[!is.na(match_insert$number),]
match_df_reads$TF[is.na(match_df_reads$TF)] <- "no-match"
n_match <- sum(match_df_reads$number[match_df_reads$TF == "Trp53"])
n_match <- sum(match_df_reads$number[match_df_reads$TF == "Gr"])
n_nomatch <- sum(match_df_reads$number[match_df_reads$TF == "no-match"])

# Create donut chart
data <- data.frame(
  lbls=c("matched:", "unmatched:"),
  count=c(n_match, n_nomatch)
)
 
# Compute percentages
data$fraction <- data$count / sum(data$count)
data$percentage <- data$fraction * 100


# Compute the cumulative percentages (top of each rectangle)
data$ymax <- cumsum(data$fraction)


# Compute the cumulative percentages (top of each rectangle)
data$ymax <- cumsum(data$fraction)

# Compute the bottom of each rectangle
data$ymin <- c(0, head(data$ymax, n=-1))

# Compute label position
data$labelPosition <- (data$ymax + data$ymin) / 2

# Compute a good label
data$label <- paste0(data$lbls, "\n", round(data$percentage), "%")

# Make the plot
ggplot(data, aes(ymax=ymax, ymin=ymin, xmax=4, xmin=3, fill=lbls)) +
  geom_rect() +
  geom_text(x=2, aes(y=labelPosition, label=label, color=lbls), size=5) + # x here controls label position (inner / outer)
  scale_fill_manual(values = c("#1B998B", "#2D3047")) +
  scale_color_manual(values = c("#1B998B", "#2D3047")) +
  labs(title = "percentage of reads from unmatched inserts (only matched barcodes)") +
  coord_polar(theta="y") +
  xlim(c(0, 4)) +
  theme_void() +
  theme(legend.position = "none")
```


# Barcodes attached to wrong insert?
```{r, barcode_matching, fig.width=4, fig.height=4, fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Select only matched barcodes
pDNA_seq_insert <- pDNA_seq[pDNA_seq$barcode %in% ref_seq$barcode,]

# Add bc-id to the barcodes found in pDNA
ref_seq_insert <- ref_seq %>% dplyr::select(barcode, name) %>% setnames("name", "bc-match")
pDNA_seq_insert <- merge(pDNA_seq_insert, ref_seq_insert)

# Add insert-id to the inserts found in pDNA
ref_seq_insert <- ref_seq %>% dplyr::select(insert, name) %>% setnames("name", "insert-match")
ref_seq_insert$`insert-match` <- gsub("(.*)_bc_[0-9]$", "\\1", ref_seq_insert$`insert-match`)
ref_seq_insert <- ref_seq_insert %>% unique()
pDNA_seq_insert <- merge(pDNA_seq_insert, ref_seq_insert)

# Count occurences where bc matches insert
pDNA_seq_insert$`bc-match` <- gsub("(.*)_bc_[0-9]$", "\\1", pDNA_seq_insert$`bc-match`)
for (i in 1:nrow(pDNA_seq_insert)) {
  pDNA_seq_insert$match[i] <- identical(pDNA_seq_insert$`bc-match`[i], pDNA_seq_insert$`insert-match`[i])
  
}


pDNA_seq_insert$TF <- gsub("_.*", "\\1", pDNA_seq_insert$`insert-match`)
pDNA_seq_insert <- pDNA_seq_insert[pDNA_seq_insert$TF == "Gr",]

## Identify the unmapped fraction
n_match <- sum(pDNA_seq_insert$number[pDNA_seq_insert$match == T])
n_nomatch <- sum(pDNA_seq_insert$number[pDNA_seq_insert$match == F])

# Create donut chart
data <- data.frame(
  lbls=c("correct:", "incorrect:"),
  count=c(n_match, n_nomatch)
)
 
# Compute percentages
data$fraction <- data$count / sum(data$count)
data$percentage <- data$fraction * 100


# Compute the cumulative percentages (top of each rectangle)
data$ymax <- cumsum(data$fraction)


# Compute the cumulative percentages (top of each rectangle)
data$ymax <- cumsum(data$fraction)

# Compute the bottom of each rectangle
data$ymin <- c(0, head(data$ymax, n=-1))

# Compute label position
data$labelPosition <- (data$ymax + data$ymin) / 2

# Compute a good label
data$label <- paste0(data$lbls, "\n", round(data$percentage), "%")

# Make the plot
ggplot(data, aes(ymax=ymax, ymin=ymin, xmax=4, xmin=3, fill=lbls)) +
  geom_rect() +
  geom_text(x=2, aes(y=labelPosition, label=label, color=lbls), size=5) + # x here controls label position (inner / outer)
  scale_fill_manual(values = c("#1B998B", "#2D3047")) +
  scale_color_manual(values = c("#1B998B", "#2D3047")) +
  labs(title = "percentage of reads where bc attached to wrong insert ") +
  coord_polar(theta="y") +
  xlim(c(0, 4)) +
  theme_void() +
  theme(legend.position = "none")


ggplot(pDNA_seq_insert, aes(x=match, y=number)) +
  geom_quasirandom() +
  labs(title = "Read count distribution", 
       subtitle = "bc attached to correct insert vs. attached to wrong insert")+
  ylab("counts") + xlab("match vs. non-match")+
  ylim(1,1000)+
  theme_bw()
```

# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

