<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>TP53 reporter library scan - detailed analysis</title>

<script src="site_libs/header-attrs-2.13/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/united.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>







<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">TP53 Reporter Scan</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Info</a>
</li>
<li>
  <a href="barcode-processing.html">Data Quality Checks</a>
</li>
<li>
  <a href="cDNA-processing.html">Detailed Reporter Activity Analysis</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">TP53 reporter library scan - detailed
analysis</h1>
<h4 class="author">Max Trauernicht</h4>
<address class="author_afil">
Netherlands Cancer Institute - van Steensel
lab<br><a class="author_email" href="mailto:#"><a
href="mailto:m.trauernicht@nki.nl"
class="email">m.trauernicht@nki.nl</a></a>
</address>
<h4 class="date">17/02/2023</h4>

</div>


<hr />
<div id="introduction" class="section level3">
<h3>Introduction</h3>
<p>~6,000 barcoded TP53 reporters were probed in MCF7 TP53WT/KO cells
and stimulated with Nutlin-3a. I previously processed the raw sequencing
data, quantified the pDNA data and normalized the cDNA data. In this
script, a detailed dissection of the reporter activities will be carried
out to understand how TP53 drives transcription and to identify the most
sensitive TP53 reporters.</p>
<hr />
</div>
<div id="setup" class="section level2 tabset">
<h2 class="tabset">Setup</h2>
<!-- little HTML script to do indentation of the table of contents -->
<script>
    $(document).ready(function() {
      $items = $('div#TOC li');
      $items.each(function(idx) {
        num_ul = $(this).parentsUntil('#TOC').length;
        $(this).css({'text-indent': num_ul * 10, 'padding-left': 0});
      });

    });
</script>
<style type="text/css">
div.sourceCode {
  overflow-x: hidden;
}
</style>
<div id="libraries" class="section level3">
<h3>Libraries</h3>
<hr />
</div>
<div id="functions" class="section level3">
<h3>Functions</h3>
<hr />
</div>
<div id="load-data" class="section level3">
<h3>Load data</h3>
<p><img src="cDNA-processing_files/figure-html/unnamed-chunk-3-1.png" width="100%" style="display: block; margin: auto;" /><img src="cDNA-processing_files/figure-html/unnamed-chunk-3-2.png" width="100%" style="display: block; margin: auto;" /><img src="cDNA-processing_files/figure-html/unnamed-chunk-3-3.png" width="100%" style="display: block; margin: auto;" /></p>
<hr />
</div>
</div>
<div id="figure-1-characterize-p53-activities-per-condition"
class="section level2">
<h2>Figure 1: Characterize P53 activities per condition</h2>
<p>Aim: I want to characterize the reporter activity distributions in
the tested conditions. Does Nutlin boost P53 reporter activity and is
P53 inactive in the KO cells?
<img src="cDNA-processing_files/figure-html/unnamed-chunk-4-1.png" width="100%" style="display: block; margin: auto;" /></p>
<pre><code>## [1] 0.9685877</code></pre>
<pre><code>## [1] 0.9036356</code></pre>
<pre><code>## [1] 0.903232</code></pre>
<p><img src="cDNA-processing_files/figure-html/unnamed-chunk-4-2.png" width="100%" style="display: block; margin: auto;" /><img src="cDNA-processing_files/figure-html/unnamed-chunk-4-3.png" width="100%" style="display: block; margin: auto;" /></p>
<p>Conclusion: 1F: Replicates do correlate well. 1G: Negative controls
are inactive compared to P53 reporters. P53 reporters become more active
in WT cells and even more active upon Nutlin stimulation.</p>
<hr />
</div>
<div
id="figure-2-effect-of-affinity-and-binding-sites-binding-site-positioning"
class="section level2">
<h2>Figure 2: Effect of affinity and binding sites + binding site
positioning</h2>
<p>Aim: How does the binding site affinity, copy number, and their
respective positioning affect reporter activity?</p>
<p><img src="cDNA-processing_files/figure-html/unnamed-chunk-5-1.png" width="100%" style="display: block; margin: auto;" /><img src="cDNA-processing_files/figure-html/unnamed-chunk-5-2.png" width="100%" style="display: block; margin: auto;" /><img src="cDNA-processing_files/figure-html/unnamed-chunk-5-3.png" width="100%" style="display: block; margin: auto;" /><img src="cDNA-processing_files/figure-html/unnamed-chunk-5-4.png" width="100%" style="display: block; margin: auto;" /></p>
<p>Conclusion: BS006 is the most responsive to Nutlin-3a. Addition of
binding sites is super-additive. Positioning of binding sites matters -
putting them directly next to each other is inhibitory, and putting them
close to the TSS leads to higher activity.</p>
<hr />
<p>Figure 3: The effect of the spacer length.</p>
<p>Aim: Show how the spacer length between adjacent binding sites
affects reporter activity.</p>
<p><img src="cDNA-processing_files/figure-html/unnamed-chunk-6-1.png" width="100%" style="display: block; margin: auto;" /><img src="cDNA-processing_files/figure-html/unnamed-chunk-6-2.png" width="100%" style="display: block; margin: auto;" /></p>
<p>Conclusion: Spacer length influences activity periodically. Adjacent
binding sites need to be 180 degrees tilted with respect to each other
to achieve optimal activation.</p>
<hr />
</div>
<div
id="figure-4-the-effect-of-the-minimal-promoter-and-the-spacer-sequence."
class="section level2">
<h2>Figure 4: The effect of the minimal promoter and the spacer
sequence.</h2>
<p>Aim: Show how the P53 reporters interact with the two minimal
promoters and the three spacer sequences.</p>
<p><img src="cDNA-processing_files/figure-html/unnamed-chunk-7-1.png" width="100%" style="display: block; margin: auto;" /><img src="cDNA-processing_files/figure-html/unnamed-chunk-7-2.png" width="100%" style="display: block; margin: auto;" /><img src="cDNA-processing_files/figure-html/unnamed-chunk-7-3.png" width="100%" style="display: block; margin: auto;" /><img src="cDNA-processing_files/figure-html/unnamed-chunk-7-4.png" width="100%" style="display: block; margin: auto;" /><img src="cDNA-processing_files/figure-html/unnamed-chunk-7-5.png" width="100%" style="display: block; margin: auto;" /><img src="cDNA-processing_files/figure-html/unnamed-chunk-7-6.png" width="100%" style="display: block; margin: auto;" /><img src="cDNA-processing_files/figure-html/unnamed-chunk-7-7.png" width="100%" style="display: block; margin: auto;" /></p>
<p>Conclusion: Promoter and spacer sequence influence activity
linearly.</p>
<hr />
</div>
<div id="figure-5-6-linear-model-selection-of-best-reporters"
class="section level2">
<h2>Figure 5 &amp; 6: Linear model + Selection of best reporters</h2>
<p>Aim: Can we explain now every observation using a linear model?
<img src="cDNA-processing_files/figure-html/unnamed-chunk-8-1.png" width="100%" style="display: block; margin: auto;" /><img src="cDNA-processing_files/figure-html/unnamed-chunk-8-2.png" width="100%" style="display: block; margin: auto;" /></p>
<pre><code>## MODEL INFO:
## Observations: 264
## Dependent Variable: log2(reporter_activity)
## Type: OLS linear regression 
## 
## MODEL FIT:
## F(9,254) = 109.03, p = 0.00
## R² = 0.79
## Adj. R² = 0.79 
## 
## Standard errors: OLS
## ---------------------------------------------------------------
##                                     Est.   S.E.   t val.      p
## -------------------------------- ------- ------ -------- ------
## (Intercept)                         4.01   0.08    50.79   0.00
## promotermCMV                        0.67   0.09     7.39   0.00
## background2                        -0.91   0.09   -10.03   0.00
## background3                         0.33   0.09     3.57   0.00
## spacing_degree_transf               0.75   0.04    20.06   0.00
## affinity_id3_med_only               0.13   0.07     1.75   0.08
## affinity_id5_low_only               0.89   0.07    11.97   0.00
## affinity_id7_very-low_only          0.82   0.07    11.06   0.00
## promotermCMV:background2            0.44   0.13     3.45   0.00
## promotermCMV:background3           -0.76   0.13    -5.92   0.00
## ---------------------------------------------------------------</code></pre>
<p><img src="cDNA-processing_files/figure-html/unnamed-chunk-8-3.png" width="100%" style="display: block; margin: auto;" /></p>
<pre><code>## MODEL INFO:
## Observations: 264
## Dependent Variable: log2(reporter_activity)
## Type: OLS linear regression 
## 
## MODEL FIT:
## F(9,254) = 88.29, p = 0.00
## R² = 0.76
## Adj. R² = 0.75 
## 
## Standard errors: OLS
## ---------------------------------------------------------------
##                                     Est.   S.E.   t val.      p
## -------------------------------- ------- ------ -------- ------
## (Intercept)                         2.78   0.10    28.58   0.00
## promotermCMV                        1.11   0.11     9.92   0.00
## background2                        -0.87   0.11    -7.79   0.00
## background3                         0.50   0.11     4.46   0.00
## spacing_degree_transf               0.28   0.05     6.09   0.00
## affinity_id3_med_only              -0.14   0.09    -1.50   0.13
## affinity_id5_low_only               1.43   0.09    15.57   0.00
## affinity_id7_very-low_only         -0.08   0.09    -0.90   0.37
## promotermCMV:background2            0.21   0.16     1.31   0.19
## promotermCMV:background3           -1.13   0.16    -7.10   0.00
## ---------------------------------------------------------------</code></pre>
<p><img src="cDNA-processing_files/figure-html/unnamed-chunk-8-4.png" width="100%" style="display: block; margin: auto;" /><img src="cDNA-processing_files/figure-html/unnamed-chunk-8-5.png" width="100%" style="display: block; margin: auto;" /><img src="cDNA-processing_files/figure-html/unnamed-chunk-8-6.png" width="100%" style="display: block; margin: auto;" /></p>
<p>Conlusion: Top reporters are better than commercial reporters. Linear
model gives insights into which features are important to drive high
expression.</p>
</div>
<div id="session-info" class="section level1">
<h1>Session Info</h1>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">paste</span>(<span class="st">&quot;Run time: &quot;</span>,<span class="fu">format</span>(<span class="fu">Sys.time</span>()<span class="sc">-</span>StartTime))</span></code></pre></div>
<pre><code>## [1] &quot;Run time:  21.24236 secs&quot;</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getwd</span>()</span></code></pre></div>
<pre><code>## [1] &quot;/DATA/usr/m.trauernicht/projects/P53_reporter_scan/analyses&quot;</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">date</span>()</span></code></pre></div>
<pre><code>## [1] &quot;Fri Feb 17 16:09:49 2023&quot;</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 4.0.5 (2021-03-31)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.5 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
##  [1] stats4    grid      parallel  stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] biomaRt_2.46.3              ENCODExplorer_2.16.0        LncFinder_1.1.5             RcmdrMisc_2.7-2             sandwich_3.0-1              car_3.0-12                  carData_3.0-5              
##  [8] sjmisc_2.8.9                ggseqlogo_0.1               heatmaply_1.3.0             magrittr_2.0.3              phylotools_0.2.2            ape_5.6-2                   DNABarcodes_1.20.0         
## [15] gtools_3.9.2                SimRAD_0.96                 zlibbioc_1.36.0             ShortRead_1.48.0            GenomicAlignments_1.26.0    Rsamtools_2.6.0             BiocParallel_1.24.1        
## [22] Biostrings_2.58.0           XVector_0.30.0              universalmotif_1.8.5        seqLogo_1.56.0              seqinr_4.2-8                PCAtools_2.2.0              ggrepel_0.9.1              
## [29] DESeq2_1.30.1               SummarizedExperiment_1.20.0 Biobase_2.50.0              MatrixGenerics_1.2.1        matrixStats_0.62.0          GenomicRanges_1.42.0        GenomeInfoDb_1.26.7        
## [36] IRanges_2.24.1              S4Vectors_0.28.1            BiocGenerics_0.36.1         ggpointdensity_0.1.0        ggbiplot_0.55               scales_1.2.0                factoextra_1.0.7.999       
## [43] shiny_1.7.1                 haven_2.5.0                 vwr_0.3.0                   latticeExtra_0.6-29         lattice_0.20-41             stringdist_0.9.8            data.table_1.14.2          
## [50] ggrastr_1.0.1               jtools_2.1.4                glmnetUtils_1.1.8           glmnet_4.1-4                Matrix_1.5-1                randomForest_4.6-14         plotly_4.10.0              
## [57] ROCR_1.0-11                 tidyr_1.2.0                 stringr_1.4.0               readr_2.1.2                 GGally_2.1.2                gridExtra_2.3               cowplot_1.1.1              
## [64] plyr_1.8.7                  viridis_0.6.2               viridisLite_0.4.0           ggforce_0.3.3               ggbeeswarm_0.6.0            ggpubr_0.4.0                pheatmap_1.0.12            
## [71] tibble_3.1.6                maditr_0.8.3                dplyr_1.0.8                 ggplot2_3.4.0               RColorBrewer_1.1-3         
## 
## loaded via a namespace (and not attached):
##   [1] rappdirs_0.3.3                rtracklayer_1.50.0            ModelMetrics_1.2.2.2          bit64_4.0.5                   knitr_1.38                    irlba_2.3.5                   DelayedArray_0.16.3          
##   [8] rpart_4.1-15                  hwriter_1.3.2.1               hardhat_0.2.0                 RCurl_1.98-1.6                generics_0.1.2                RSQLite_2.2.12                future_1.25.0                
##  [15] proxy_0.4-26                  bit_4.0.4                     tzdb_0.3.0                    xml2_1.3.3                    lubridate_1.8.0               webshot_0.5.3                 httpuv_1.6.5                 
##  [22] assertthat_0.2.1              gower_1.0.0                   xfun_0.30                     jquerylib_0.1.4               hms_1.1.1                     evaluate_0.15                 promises_1.2.0.1             
##  [29] TSP_1.2-0                     progress_1.2.2                fansi_1.0.3                   dbplyr_2.1.1                  dendextend_1.15.2             readxl_1.4.0                  DBI_1.1.2                    
##  [36] geneplotter_1.68.0            htmlwidgets_1.5.4             reshape_0.8.9                 purrr_0.3.4                   ellipsis_0.3.2                crosstalk_1.2.0               backports_1.4.1              
##  [43] insight_0.18.2                annotate_1.68.0               sparseMatrixStats_1.2.1       vctrs_0.5.1                   Cairo_1.5-15                  sjlabelled_1.2.0              abind_1.4-5                  
##  [50] caret_6.0-92                  cachem_1.0.6                  withr_2.5.0                   vroom_1.5.7                   checkmate_2.1.0               prettyunits_1.1.1             cluster_2.1.1                
##  [57] lazyeval_0.2.2                crayon_1.5.1                  genefilter_1.72.1             labeling_0.4.2                recipes_0.2.0                 pkgconfig_2.0.3               tweenr_1.0.2                 
##  [64] nlme_3.1-152                  vipor_0.4.5                   seriation_1.3.5               nnet_7.3-17                   globals_0.14.0                rlang_1.0.6                   lifecycle_1.0.3              
##  [71] registry_0.5-1                BiocFileCache_1.14.0          rsvd_1.0.5                    AnnotationHub_2.22.1          cellranger_1.1.0              polyclip_1.10-0               zoo_1.8-10                   
##  [78] base64enc_0.1-3               beeswarm_0.4.0                png_0.1-7                     bitops_1.0-7                  pROC_1.18.0                   pander_0.6.5                  blob_1.2.3                   
##  [85] DelayedMatrixStats_1.12.3     shape_1.4.6                   parallelly_1.31.1             jpeg_0.1-9                    rstatix_0.7.0                 ggsignif_0.6.3                beachmat_2.6.4               
##  [92] memoise_2.0.1                 compiler_4.0.5                dqrng_0.3.0                   cli_3.4.1                     ade4_1.7-19                   listenv_0.8.0                 htmlTable_2.4.0              
##  [99] Formula_1.2-4                 mgcv_1.8-34                   MASS_7.3-53.1                 tidyselect_1.1.2              stringi_1.7.6                 forcats_0.5.1                 highr_0.9                    
## [106] yaml_2.3.5                    askpass_1.1                   BiocSingular_1.6.0            locfit_1.5-9.4                sass_0.4.1                    tools_4.0.5                   future.apply_1.8.1           
## [113] rstudioapi_0.13               BH_1.78.0-0                   foreach_1.5.2                 foreign_0.8-81                prodlim_2019.11.13            farver_2.1.0                  BiocManager_1.30.19          
## [120] digest_0.6.29                 lava_1.6.10                   nortest_1.0-4                 Rcpp_1.0.8.3                  broom_0.8.0                   BiocVersion_3.12.0            later_1.3.0                  
## [127] httr_1.4.2                    AnnotationDbi_1.52.0          Rdpack_2.3                    colorspace_2.0-3              XML_3.99-0.9                  splines_4.0.5                 xtable_1.8-4                 
## [134] jsonlite_1.8.0                timeDate_3043.102             ipred_0.9-12                  R6_2.5.1                      Hmisc_4.7-0                   pillar_1.7.0                  htmltools_0.5.2              
## [141] mime_0.12                     glue_1.6.2                    fastmap_1.1.0                 interactiveDisplayBase_1.28.0 class_7.3-18                  codetools_0.2-18              utf8_1.2.2                   
## [148] bslib_0.3.1                   curl_4.3.2                    openssl_2.0.0                 survival_3.2-10               rmarkdown_2.13                munsell_0.5.0                 e1071_1.7-9                  
## [155] GenomeInfoDbData_1.2.4        iterators_1.0.14              reshape2_1.4.4                gtable_0.3.0                  rbibutils_2.2.8</code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
