---
title: "Barcode processing - Deep P53/GR scan - stimulation 2"
author: "Max Trauernicht"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output: 
  prettydoc::html_pretty:
    theme: leonids
    highlight: github
  #   toc: true
  #   toc_float: true
  #   code_folding: show
  # editor_options:
  #   chunk_output_type: console
---

*knitr document van Steensel lab*

# TF reporter barcode processing - Deep P53/GR scan - stimulation 2

## Introduction
12,000 TF reporters on pMT06 were transfected into A549 & MCF7 cells, sequencing data yielded barcode counts of these experiments. These counts will be processed in this script. 

```{r setup, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(data.table)
library(plyr)
library(stringr)
library(ggpubr)
library(GGally)
library(vwr)
library(dplyr)
library(tibble)
library(plotly)
library(ggbeeswarm)
library(haven)
library(readr)
library(parallel)
library(RColorBrewer)
library(gridExtra)
library(pheatmap)
library(shiny)
library(factoextra)
library(ggbiplot)
library(ggpointdensity)
library(viridis)
```


```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
#Custom functions
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}


# From Fede:
# ggpairs custom functions
corColor <- function(data, mapping, color = I("black"), sizeRange = c(1, 3), ...) {

  x   <- eval_data_col(data, mapping$x)
  y   <- eval_data_col(data, mapping$y)
  r   <- cor(x, y, "pairwise.complete.obs")
  rt  <- format(r, digits = 3)
  tt  <- as.character(rt)
  cex <- max(sizeRange)

  # helper function to calculate a useable size
  percent_of_range <- function(percent, range) {
    percent * diff(range) + min(range, na.rm = TRUE)
  }

  # plot correlation coefficient
  p <- ggally_text(label = tt, mapping = aes(), xP = 0.5, yP = 0.5,
                   size = I(percent_of_range(cex * abs(r), sizeRange)), color = color, ...) +
    theme(panel.grid.minor=element_blank(),
          panel.grid.major=element_blank())

  corColors <- RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")[2:6]

  if (r <= boundaries[1]) {
    corCol <- corColors[1]
  } else if (r <= boundaries[2]) {
    corCol <- corColors[2]
  } else if (r < boundaries[3]) {
    corCol <- corColors[3]
  } else if (r < boundaries[4]) {
    corCol <- corColors[4]
  } else {
    corCol <- corColors[5]
  }

  p <- p +
    theme(panel.background = element_rect(fill = corCol))

  return(p)
}
```


```{r data import, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Import barcode counts per experiment
bc_files = list.files('/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/gcf6301/results/',
                       full.names=T, patter='*_barcode_counts.tsv')
bc_list <- lapply(bc_files, fread, header = FALSE)
names(bc_list)<- gsub('.*//6301_[0-9]{1,2}_(.*?)_[CGAT]{6}.*_barcode_counts.tsv', 
                                    '\\1', 
                                    bc_files)

# Import barcode annotation
bc_annotation <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/library_design/output/tf_df_complete.csv", header = T) %>% 
  dplyr::select(barcode, tf, oligo.barcode, spacing, promoter,
         position, distance, background, affinity_pos1,
         affinity_pos2, affinity_pos3, affinity_pos4, seq.name) %>%
  setnames("seq.name", "reporter_id")
```

## Analysis
```{r cluster_compare, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Generate long dfs
for (i in 1:length(bc_list)) {
  if (i == 1) {
  bc_df <- data.frame(bc_list[i])
  bc_df[3] <- names(bc_list[i])
  names(bc_df) <- c("barcode", "count", "name")
  bc_df <- reshape2::dcast(bc_df, barcode ~ name, value.var = "count")
  }
  else {
  bc_df_i <- data.frame(bc_list[i])
  bc_df_i[3] <- names(bc_list[i])
  names(bc_df_i) <- c("barcode", "count", "name")
  bc_df_i <- reshape2::dcast(bc_df_i, barcode ~ name, value.var = "count")
  bc_df <- merge(bc_df, bc_df_i, all = T)
  }
}

# Match designed barcodes
bc_df <- merge(bc_df, bc_annotation, all = T)


# Convert to long df
bc_df <- melt(bc_df, id.vars = c("barcode", "tf", "oligo.barcode", "spacing", "distance", "promoter", "background", "position", 
                                 "affinity_pos1", "affinity_pos2", "affinity_pos3", "affinity_pos4", "reporter_id"),
              variable.name = "condition", value.name = "starcode_counts", as.is = T)
bc_df$starcode_counts[is.na(bc_df$starcode_counts)] <- 0


# Rename conditions
bc_df$condition <- gsub("_pMT06", "", bc_df$condition)
```


### Data exploration
```{r data inspection, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# First compute reads per million to estimate the relative counts in their respective sample
for (i in unique(bc_df$condition)) {
  bc_df$rpm[bc_df$condition == i] <- (bc_df$starcode_counts[bc_df$condition == i] + 1) / # Adds a pseudocount of 1
    sum(bc_df$starcode_counts[bc_df$condition == i]) *1e6
}

# I want to show the following:
## 1: Read distribution of matched barcodes vs. unmatched barcode
ggplot(bc_df[-grep("random", bc_df$tf),], aes(x = tf, y = rpm)) +
  geom_jitter(alpha = 0.1) +
  theme_bw() +
  ylim(0,2500) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, size = 6)) +
  facet_wrap(~condition)

bc_df_2 <- bc_df[bc_df$rpm <= 250,]
bc_df_2 <- bc_df_2[bc_df_2$rpm >= 0.5,]
bc_df_2 <- bc_df_2[!is.na(bc_df_2$tf),]

ggplot(bc_df_2, aes(x = rpm)) +
  geom_histogram(binwidth = 20) +
  theme_bw() +
  xlim(0,250)+
  ylim(0,6000)+
  facet_wrap(~condition)+
  theme(strip.background =element_rect(fill="#D6D5C9"))

ggplot(bc_df[bc_df$rpm >= 250 & !is.na(bc_df$tf),], aes(x = rpm)) +
  geom_histogram(binwidth = 20) +
  theme_bw() +
  xlim(250,1000)+
  ylim(0,100)+
  facet_wrap(~condition)+
  theme(strip.background =element_rect(fill="#D6D5C9"))

n_highly_expressed <- data.frame("condition" = unique(bc_df$condition),
                                 "n_bc" = "", stringsAsFactors=FALSE)
for (i in unique(bc_df$condition)) {
  n_highly_expressed$n_bc[n_highly_expressed$condition == i] <- length(bc_df$barcode[bc_df$rpm > 500 & bc_df$condition == i])
}

plot_ly(n_highly_expressed, x = ~condition, y = ~as.numeric(n_bc), type = 'bar',
             marker = list(color = '#D6D5C9',
                           line = list(color = 'rgb(8,48,107)', width = 1.5))) %>% 
  layout(title = "Highly expressed barcodes",
         yaxis = list(title = "Number of barcodes with > 500 rpm"),
         xaxis = list(title = "Condition"))


## 2: How many barcodes can I find back at which cutoff? + What is the percentage of barcode reads that match the design at which cutoff?
## Identify the unmapped fraction
bc_fraction <- data.frame("condition" = unique(bc_df$condition),
                          "bcs_found" = "", stringsAsFactors=FALSE)
rpm_cutoff <- data.frame("cutoff" = seq(0.0001,15,0.5), stringsAsFactors=FALSE)
bc_fraction <- merge(bc_fraction, rpm_cutoff)

for (i in unique(bc_fraction$cutoff)) {
  for (j in unique(bc_df$condition)) {
    bc_n <- bc_df[bc_df$rpm >= i & bc_df$condition == j,]
    bc_fraction$bcs_found[bc_fraction$cutoff == i & bc_fraction$condition == j] <- nrow(bc_n[!is.na(bc_n$tf),])/
      nrow(bc_annotation) *100
  }
}



## How many reads match to designed barcodes?
bc_reads <- data.frame("condition" = unique(bc_df$condition),
                          "bc_reads" = "", stringsAsFactors=FALSE)
bc_reads <- merge(bc_reads, rpm_cutoff)

for (i in unique(bc_reads$cutoff)) {
  for (j in unique(bc_df$condition)) {
    bc_n <- bc_df[bc_df$rpm >= i & bc_df$condition == j,]
    bc_reads$bc_reads[bc_reads$cutoff == i & bc_reads$condition == j] <- sum(bc_n$rpm[!is.na(bc_n$tf)])/
      sum(bc_n$rpm) *100
  }
}

bc_fraction <- merge(bc_fraction, bc_reads)
bc_fraction$bcs_found <- as.numeric(bc_fraction$bcs_found)
bc_fraction$bc_reads <- as.numeric(bc_fraction$bc_reads)

#c("#1B998B", "#2D3047", "#FF9B71", "#ECDD7B")
# Plot to evaluate data quality per cutoff
ggplot(bc_fraction) +
  geom_point(aes(x = cutoff, y = bcs_found), color = '#1B998B', size = 1) +
  geom_line(aes(x = cutoff, y = bcs_found), color = '#1B998B', size = 1) +
  geom_point(aes(x = cutoff, y = bc_reads), color = 'black', size = 1) +
  geom_line(aes(x = cutoff, y = bc_reads), color = 'black', size = 1) +
  theme_bw()+
  xlab("rpm cutoff")+
  ylab("total barcodes (green) and matched barcode reads (black) detected (%)")+
  facet_wrap(~condition)+
  theme(strip.background =element_rect(fill="#D6D5C9"))

## 3: What is the correlation of the 24 cDNA bc counts with the pDNA bc counts? 
pDNA <- data.frame("pDNA" = bc_df$rpm[bc_df$condition == "pMT06-v2_pDNA_1"],
                   "barcode"= bc_df$barcode[bc_df$condition == "pMT06-v2_pDNA_1"], stringsAsFactors=FALSE)
bc_df_2 <- merge(pDNA, bc_df, all = T)

ggplot(bc_df_2, aes(x = pDNA, y = rpm)) +
  geom_bin2d(bins = 100)+
  xlim(0,1000) +
  ylim(0,1000)+
  theme_bw()+
  facet_wrap(~condition)

# Plot only new vs old pDNA data
ggplot(bc_df_2[bc_df_2$condition == "pMT06_pDNA" & bc_df_2$tf == "Trp53",], aes(x = pDNA, y = rpm)) +
    geom_pointdensity(show.legend = F)+
    theme_bw()+
    theme(text = element_text(size=14)) +
    xlab("old pDNA data (e.cloni)")+
    ylab("new pDNA data (clonecatcher)")+
    ggtitle("Old vs. new pDNA data")+
    scale_color_viridis()
ggplot(bc_df_2[bc_df_2$condition == "pMT06_pDNA" & bc_df_2$tf == "Gr",], aes(x = pDNA, y = rpm)) +
    geom_pointdensity(show.legend = F)+
    theme_bw()+
    theme(text = element_text(size=14)) +
    xlab("old pDNA data (e.cloni)")+
    ylab("new pDNA data (clonecatcher)")+
    ggtitle("Old vs. new pDNA data")+
    scale_color_viridis()

# Plot distribution of reads over tfs
test <- bc_df_2[grep("pMT06",bc_df_2$condition),] %>% dplyr::select(starcode_counts,tf,condition) %>% unique()
test <- test[-grep("pMT06-v2_pDNA_2",test$condition),]
test$condition[test$condition == "pMT06-v2_pDNA_1"] <- "clonecatcher"
test$condition[test$condition == "pMT06_pDNA"] <- "e.cloni"
test <- test[!is.na(test$tf),]
test$tf <- gsub("_ran.*", "", test$tf)

ggplot(test, aes(x = condition, y = starcode_counts, fill = tf)) +
  geom_bar(position="fill", stat="identity")+
  theme_bw()+
  theme(text = element_text(size=14)) +
  xlab("")+ylab("")+
  ggtitle("Read distribution per TF")+
  scale_fill_brewer(palette = "Dark2")

cor <- data.frame("condition" = unique(bc_df_2$condition), "cor" = "", stringsAsFactors=FALSE)

for (i in unique(bc_df_2$condition)) {
  x <- bc_df_2$rpm[bc_df_2$condition == i]
  y <- bc_df_2 %>% dplyr::select(barcode, pDNA) %>% unique()
  cor$cor[cor$condition == i] <- cor(x, y$pDNA)
}

bc_df <- merge(bc_df, cor, by = "condition", all = T)

## 4: Correlation plots of the replicates
## Combine replicates of normalized data in 3 different columns
bc_df$rep <- gsub(".*([1-3]{1}$)","\\1",bc_df$condition)
bc_df_rep <- bc_df[!is.na(bc_df$tf),] %>% dplyr::select(rep, rpm, reporter_id, condition) %>% unique()
bc_df_rep$condition <- gsub("(.*?)_r[1-2]$", "\\1", bc_df_rep$condition)
rep1 <- bc_df_rep[bc_df_rep$rep == 1,]
rep2 <- bc_df_rep[bc_df_rep$rep == 2,]
rep1 <- rep1 %>% dplyr::select(-rep)
rep2 <- rep2 %>% dplyr::select(-rep)

names(rep1) <- c("rep1", "reporter", "condition")
names(rep2) <- c("rep2", "reporter", "condition")

bc_df_rep <- merge(rep1, rep2, all = TRUE)
bc_df_rep$neg_ctrl <- "No"
bc_df_rep$neg_ctrl[grep("random", bc_df_rep$reporter)] <- "Yes"

colors <- c("#2D3047", "#1B998B")

ggscatter(bc_df_rep, x = "rep1", y = "rep2",
   add = "reg.line",
   color = "neg_ctrl",
   size = 0.5,
   alpha = 0.2,
   add.params = list(color = "blue", fill = "lightgray"), title = "rep1 vs rep2",
   conf.int = TRUE, ylab = "rep2", xlab = "rep1") + 
  stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
  geom_abline(linetype = "dashed") +
  xlim(0,10000) + ylim(0,10000) +
  scale_color_manual(values = colors)+facet_wrap(~condition)



## 5: What is the sequencing depth of the pDNA data?

# Plot pDNA distribution vs. cutoff
pDNA_fraction <- bc_df[grep("pMT06-v2", bc_df$condition),]
pDNA_fraction <- pDNA_fraction[!is.na(pDNA_fraction$tf),]
pDNA_fraction <- pDNA_fraction %>% dplyr::select(barcode, condition, rpm, tf)
pDNA_fraction <- dcast(pDNA_fraction, barcode + tf ~ condition, value.var = "rpm")
pDNA_fraction <- pDNA_fraction %>% 
  mutate(pDNA = (`pMT06-v2_pDNA_1` + `pMT06-v2_pDNA_2`) / 2) %>%
  dplyr::select(-`pMT06-v2_pDNA_1`, -`pMT06-v2_pDNA_2`)
pDNA_fr <- data.frame("cutoff" = seq(0,100, 1),
                      "bcs_missing" = "", stringsAsFactors=FALSE)
pDNA_fr <- merge(pDNA_fr, c("Trp53", "Gr"))
for (i in pDNA_fr$cutoff) {
  for (j in pDNA_fr$y) {
  pDNA_fr$bcs_missing[pDNA_fr$cutoff == i & pDNA_fr$y == j] <- length(pDNA_fraction$barcode[pDNA_fraction$tf == j]) -
    length(pDNA_fraction$barcode[pDNA_fraction$pDNA >= i & pDNA_fraction$tf == j])
  }
}

ggplot(pDNA_fr) +
  geom_point(aes(x = cutoff, y = as.numeric(bcs_missing)), color = '#1B998B', size = 1) +
  geom_line(aes(x = cutoff, y = as.numeric(bcs_missing)), color = '#1B998B', size = 1) +
  theme_bw()+
  xlab("rpm cutoff")+
  ylab("barcodes excluded from analysis (6,000 in total)") +
  geom_vline(xintercept = 25, linetype = "dashed", color = "black") +
  facet_wrap(~y)
```
**Stuff learned from above figures:**  
-> Based on these figures I will exclude all samples that have less than 300 barcodes with at least 500 normalized counts.

```{r filtering, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Remove samples with low dynamic range
for (i in unique(bc_df$condition)) {
  bc_df$n_bc[bc_df$condition == i] <- length(bc_df$barcode[bc_df$rpm > 500 & bc_df$condition == i])
}
bc_df <- bc_df[bc_df$n_bc >= 300 | bc_df$condition == "pMT06-v2_pDNA_1" | bc_df$condition == "pMT06-v2_pDNA_2",] %>% dplyr::select(-n_bc)

# Remove samples with high pDNA contamination
bc_df <- bc_df[bc_df$cor <= 0.85 | bc_df$condition == "pMT06-v2_pDNA_1" | bc_df$condition == "pMT06-v2_pDNA_2",] %>% dplyr::select(-cor)

# Remove all non-matching reads
bc_df <- bc_df[!is.na(bc_df$tf),]
```



### Normalization of barcode counts:
Divide cDNA barcode counts through pDNA barcode counts, if more than 30 pDNA counts for that barcode
```{r normalization, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Normalize data using pDNA data
# Add pDNA data as separate column
pDNA <- bc_df[grep("pDNA", bc_df$condition),] %>% dplyr::select(barcode, condition, rpm) %>% unique() %>%
  dcast(barcode ~ condition) %>%
  mutate(pDNA_counts_rpm = (`pMT06-v2_pDNA_1` + `pMT06-v2_pDNA_2`) / 2) %>%
  dplyr::select(barcode, pDNA_counts_rpm)
bc_df <- merge(pDNA, bc_df)

## Compute activity by dividing cDNA bc counts through pDNA bc counts
bc_df$activity <- bc_df$rpm / bc_df$pDNA_counts_rpm
```


```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# First identify and remove outlier barcodes - this removes the noise created by faulty barcode clustering etc. 
## Calculate mean and SD for each reporter
bc_df_cDNA <- bc_df[-grep("pDNA", bc_df$condition),]
bc_df_cDNA$reporter_id <- gsub(".{5}$", "", bc_df_cDNA$reporter_id)
bc_df_cDNA$reporter_activity <- ave(bc_df_cDNA$activity, bc_df_cDNA$reporter_id, 
                                bc_df_cDNA$condition, FUN =
                                  function(x) mean(x))
bc_df_cDNA$reporter_activity_sd <- ave(bc_df_cDNA$activity, bc_df_cDNA$reporter_id, 
                              bc_df_cDNA$condition,  FUN =
                                  function(x) sd(x))

## Remove data points that are 2xSD away from the mean
bc_df_cDNA$upper_activity <- bc_df_cDNA$reporter_activity + (2 * bc_df_cDNA$reporter_activity_sd)
bc_df_cDNA$lower_activity <- bc_df_cDNA$reporter_activity - (2 * bc_df_cDNA$reporter_activity_sd)
bc_df_cDNA_filt <- bc_df_cDNA
bc_df_cDNA_filt$activity[bc_df_cDNA_filt$activity >= bc_df_cDNA_filt$upper_activity | 
                                bc_df_cDNA_filt$activity <= bc_df_cDNA_filt$lower_activity] <- NA

## Recalculate mean and sd
bc_df_cDNA_filt$reporter_activity <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id, 
                                bc_df_cDNA_filt$condition, FUN =
                                  function(x) mean(x))
bc_df_cDNA_filt$reporter_activity_sd <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id, 
                              bc_df_cDNA_filt$condition,  FUN =
                                  function(x) sd(x))
```





```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE}
# # Remove barcodes with multiple inserts attached
# bc_exclude <- read.csv("/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/pDNA_seq/bc_exclude.csv") %>% dplyr::select(-X) %>% setnames("x", "barcode")
# exclude <- bc_df_cDNA_filt[!bc_df_cDNA_filt$barcode %in% bc_exclude$barcode,]
# 
# # Reassign barcodes with wrong inserts attached
# bc_replace <- read.csv("/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/pDNA_seq/bc_replace.csv") %>% dplyr::select(-X, -bc.match)
# change.df <- exclude[exclude$barcode %in% bc_replace$barcode,]
# exclude <- exclude[!exclude$barcode %in% bc_replace$barcode,]
# 
# change.df$bc.number <- 9
# change.df <- merge(change.df, bc_replace)
# e <- c("e11", "e93", "e97")
# change.df.e <- change.df[grep(paste(e, collapse = "|"), change.df$insert.match),]
# change.df <- change.df[-grep(paste(e, collapse = "|"), change.df$insert.match),]
# change.df <- change.df %>%
#   mutate(TF = gsub("(^.*?)_.*", "\\1", insert.match),
#          promoter = gsub(".*(minP|mCMV|hBGm|Random)_.*", "\\1", insert.match),
#          distance = gsub(".*_d-([0-9]{1,2}bp)_.*", "\\1", insert.match),
#          spacing = gsub(".*_s-([0-9]{1,2}bp)_.*", "\\1", insert.match),
#          background = gsub(".*([0-9]{1}$)", "\\1", insert.match),
#          reporter_id = paste(TF, spacing, distance, promoter, background, sep = "_"))
# change.df <- change.df %>% dplyr::select(-insert.match)
# 
# change.df.e <- change.df.e %>%
#   mutate(TF = gsub("(.*?)_(minP|mCMV|hBGm|Random)$", "\\1", insert.match),
#          promoter = gsub(".*(minP|mCMV|hBGm|Random)", "\\1", insert.match),
#          spacing = "",
#          distance = "",
#          background = "",
#          reporter_id = paste(TF, promoter, sep ="_"))
# change.df.e <- change.df.e %>% dplyr::select(-insert.match)
# 
# 
# exclude$reporter_id <- gsub("___", "_", exclude$reporter_id)
# exclude$reporter_id <- gsub("_0", "", exclude$reporter_id)
# 
# 
# 
# exclude <- rbind(exclude, change.df, change.df.e)
# 
# bc_df_cDNA_filt <- exclude
```



```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# Annotate the mutated motif of each TF
bc_df_cDNA_filt$neg_ctrls <- "No"
bc_df_cDNA_filt$neg_ctrls[grep("random", bc_df_cDNA_filt$tf)] <- "Yes"

# Annotate random promoter control
bc_df_cDNA_filt$rand_promoter <- "No"
bc_df_cDNA_filt$rand_promoter[grep("random", bc_df_cDNA_filt$promoter)] <- "Yes"
```



```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
## Scale data to 1, use negative controls as reference
mean_neg_ctrl <- mean(bc_df_cDNA_filt$activity[bc_df_cDNA_filt$neg_ctrls == "Yes"], na.rm = T)
bc_df_cDNA_filt$activity <- bc_df_cDNA_filt$activity / mean_neg_ctrl
bc_df_cDNA_filt$mean_activity <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id, 
                                bc_df_cDNA_filt$condition, FUN =
                                  function(x) mean(x))
bc_df_cDNA_filt <- bc_df_cDNA_filt[!is.na(bc_df_cDNA_filt$condition),]
```



### Calculate correlations between technical replicates
```{r correlations_2, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
## Combine replicates in 5 different columns
bc_df_rep <- bc_df_cDNA_filt[bc_df_cDNA_filt$rand_promoter == "No",] %>% 
  dplyr::select(oligo.barcode, activity, tf, condition, reporter_id, pDNA_counts_rpm) %>%
  mutate(pDNA_counts_rpm = ave(pDNA_counts_rpm, reporter_id, FUN = function(x) min(x))) %>% unique()
rep1 <- bc_df_rep[bc_df_rep$oligo.barcode == 1,] %>% dplyr::select(-oligo.barcode) %>% setnames("activity", "bc1")
rep2 <- bc_df_rep[bc_df_rep$oligo.barcode == 2,] %>% dplyr::select(-oligo.barcode) %>% setnames("activity", "bc2")
rep3 <- bc_df_rep[bc_df_rep$oligo.barcode == 3,] %>% dplyr::select(-oligo.barcode) %>% setnames("activity", "bc3")
rep4 <- bc_df_rep[bc_df_rep$oligo.barcode == 4,] %>% dplyr::select(-oligo.barcode) %>% setnames("activity", "bc4")
rep5 <- bc_df_rep[bc_df_rep$oligo.barcode == 5,] %>% dplyr::select(-oligo.barcode) %>% setnames("activity", "bc5")

bc_df_rep <-  Reduce(function(x, y) merge(x, y, all=TRUE), list(rep1, rep2, rep3, rep4, rep5))
bc_df_rep <- bc_df_rep[rowSums(is.na(bc_df_rep)) != ncol(bc_df_rep), ]

## Compute the correlation between two technical replicates for each pDNA cutoff, separately for each TF
cor_df <- data.frame("rpm_cutoff"  = seq(0,100,1), "cor" = "", stringsAsFactors=FALSE)
cor_df <- merge(unique(bc_df_rep$tf), cor_df)

for (i in unique(cor_df$rpm_cutoff)) {
  for (j in unique(cor_df$x)) {
    x <- bc_df_rep[bc_df_rep$pDNA_counts_rpm > i & bc_df_rep$tf == j,]
    if (nrow(x) == 0) {
      cor_df$cor[cor_df$rpm_cutoff == i & cor_df$x == j] <- NA
    } else {
      cor_df$cor[cor_df$rpm_cutoff == i & cor_df$x == j] <- cor(x$bc2, x$bc4, use = "pairwise.complete.obs", method = "spearman")
    }
  }
}

ggplot(cor_df) +
  geom_point(aes(x = rpm_cutoff, y = as.numeric(cor))) +
  theme_bw() +
  ylim(0,1) +
  facet_wrap(~x)

## Compute the correlation between two technical replicates for each pDNA cutoff, for all reporters together
cor_df <- data.frame("rpm_cutoff"  = seq(0,100,1), "cor" = "", stringsAsFactors=FALSE)

for (i in unique(cor_df$rpm_cutoff)) {
    x <- bc_df_rep[bc_df_rep$pDNA_counts_rpm > i,]
    if (nrow(x) == 0) {
      cor_df$cor[cor_df$rpm_cutoff == i] <- NA
    } else {
      cor_df$cor[cor_df$rpm_cutoff == i] <- cor(x$bc2, x$bc4, "pairwise.complete.obs", method = "spearman")
    }
}

ggplot(cor_df) +
  geom_point(aes(x = rpm_cutoff, y = as.numeric(cor))) +
  theme_bw() +
  ylim(0,1) 


## Set a pDNA cutoff based on above plots
bc_df_cDNA_filt$activity[bc_df_cDNA_filt$pDNA_counts_rpm <= 10] <- NA
bc_df_rep <- bc_df_rep[bc_df_rep$pDNA_counts_rpm >= 10,]


# Correlation matrix plot
n <- sample(1:nrow(bc_df_rep), 5000)
boundaries <- seq(from = 0.8, by = 0.05, length.out = 4)
plt <- ggpairs(bc_df_rep %>% dplyr::select(bc1, bc2, bc3, bc4, bc5),
               upper = list(continuous = corColor),
               lower = list(continuous = function(data, mapping, ...) {
                   ggally_points(data = data[n, ], mapping = mapping, alpha = 0.1, size = 0.5) +
                   geom_abline(slope = 1, lty = "dashed", col = "red") +
                   theme_bw()}),
               diag = list(continuous = function(data, mapping, ...) {
                   ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red") +
                   theme_bw()})) +
  ggtitle("Correlation Between Technial Replicates") +
  theme(text = element_text(size = 20)) +
  xlab("Reporter activity") +
  ylab("Reporter activity") + 
  theme_light()

print(plt)
```



### Data quality plots - correlation between replicates
```{r correlations_3, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# Correlation plots of the replicates
## Combine replicates of normalized data in 3 different columns
bc_df_rep <- bc_df_cDNA_filt %>% dplyr::select(rep, reporter_activity, reporter_id, condition) %>% unique()
bc_df_rep$condition <- gsub("(.*?)_r[1-3]$", "\\1", bc_df_rep$condition)
rep1 <- bc_df_rep[bc_df_rep$rep == 1,]
rep2 <- bc_df_rep[bc_df_rep$rep == 2,]
rep1 <- rep1 %>% dplyr::select(-rep)
rep2 <- rep2 %>% dplyr::select(-rep)

names(rep1) <- c("rep1", "reporter", "condition")
names(rep2) <- c("rep2", "reporter", "condition")

bc_df_rep <- merge(rep1, rep2, all = TRUE)
bc_df_rep$neg_ctrl <- "No"
bc_df_rep$neg_ctrl[grep("neg", bc_df_rep$reporter)] <- "Yes"

colors <- c("#2D3047", "#1B998B")

ggscatter(bc_df_rep, x = "rep1", y = "rep2",
   add = "reg.line",
   color = "neg_ctrl",
   size = 0.5,
   add.params = list(color = "blue", fill = "lightgray"), title = "rep1 vs rep2",
   conf.int = TRUE, ylab = "rep2", xlab = "rep1") + 
  stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
  geom_abline(linetype = "dashed") +
  xlim(0,10) + ylim(0,10) +
  scale_color_manual(values = colors)+facet_wrap(~condition)
```

```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# # Create similarity heatmap
# bc_df_cDNA_filt <- bc_df_cDNA_filt[!is.na(bc_df_cDNA_filt$condition),]
# cor_heatmap <- bc_df_cDNA_filt %>% dplyr::select(barcode, condition, activity)
# cor_heatmap <- dcast(cor_heatmap, barcode~condition) %>% column_to_rownames("barcode")
# cor_heatmap <- round(cor(cor_heatmap, use = "pairwise.complete.obs"),2)
# cor_heatmap_mcf <- cor_heatmap[grep("MCF", colnames(cor_heatmap)),grep("MCF", rownames(cor_heatmap))]
# cor_heatmap_a549 <- cor_heatmap[grep("A549", colnames(cor_heatmap)),grep("A549", rownames(cor_heatmap))]
# cor_heatmap_mES <- cor_heatmap[grep("mES", colnames(cor_heatmap)),grep("mES", rownames(cor_heatmap))]
# pheatmap(cor_heatmap)
# pheatmap(cor_heatmap_mcf)
# pheatmap(cor_heatmap_a549)
# pheatmap(cor_heatmap_mES)


# Mean of the three replicates
bc_df_cDNA_filt$condition <- gsub("(.*?)_rep[1-3]$", "\\1", bc_df_cDNA_filt$condition)
bc_df_cDNA_filt$reporter_activity <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id, bc_df_cDNA_filt$condition, FUN = function(x) mean(x))
bc_df_cDNA_filt$reporter_activity_sd <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id, bc_df_cDNA_filt$condition, FUN = function(x) sd(x))
```

```{r data export, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# Polish export dataframe
bc_df_cDNA_filt <- bc_df_cDNA_filt %>% 
  dplyr::select(-upper_activity, -lower_activity) %>% 
  mutate(log_activity = log2(activity),
         log_reporter_activity = log2(reporter_activity))


# Export bc_df for cDNA analysis
filename <- SetFileName("_reporter_activity_filt", "mt")
setwd("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/gcf6301/results/")
write.csv(bc_df_cDNA_filt, file = paste(filename,".csv", sep = ""), row.names = F)
```

## Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

