---
title: "genomic P53 response element scoring"
author: 
  - name: "Max Trauernicht"
    email: "m.trauernicht@nki.nl"
    affiliation: "Netherlands Cancer Institute - van Steensel lab"
date: '`r format(Sys.time(), "%d/%m/%Y")`'
output: 
  html_document:
    theme: united
    highlight: pygments
    fig_caption: yes
    code_folding: hide
    df_print: kable
    toc: true
    toc_depth: 4
    toc_float:
      collapsed: false
---

---

### Aim
I want to score the affinity genomic P53 response elements using the same algorithm that was used to create P53 reporters. I then want to conclude if I can say anything about P53 gene regulation with the P53 reporters that I generated or not.

---

## Setup {.tabset}

<!-- little HTML script to do indentation of the table of contents -->
<script>
    $(document).ready(function() {
      $items = $('div#TOC li');
      $items.each(function(idx) {
        num_ul = $(this).parentsUntil('#TOC').length;
        $(this).css({'text-indent': num_ul * 10, 'padding-left': 0});
      });

    });
</script>

```{css, echo = FALSE}
div.sourceCode {
  overflow-x: hidden;
}
```

### Libraries

```{r setup, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(GenomicRanges)
library(BSgenome.Hsapiens.UCSC.hg38)
library(GenomicFeatures)
library(BSgenome)
library(spgs)
library(dplyr)
library(ggplot2)
library(reshape)
library(ggbeeswarm)
library(biomaRt)
library(readr)
library(stringr)
library(maditr)
library(phylotools)
library(plyr)
library(tidyr)
library(ggpubr)
```


### Custom functions

```{r, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}

ReadFasta<-function(file) {
   # Read the file line by line
   fasta<-readLines(file)
   # Identify header lines
   ind<-grep(">", fasta)
   # Identify the sequence lines
   s<-data.frame(ind=ind, from=ind+1, to=c((ind-1)[-1], length(fasta)))
   # Process sequence lines
   seqs<-rep(NA, length(ind))
   for(i in 1:length(ind)) {
      seqs[i]<-paste(fasta[s$from[i]:s$to[i]], collapse="")
   }
   # Create a data frame 
   DF<-data.frame(name=gsub(">", "", fasta[ind]), sequence=seqs)
   # Return the data frame as a result object from the function
   return(DF)
}
```


### Extract all TSSs 

```{r, fig.width=10, fig.height=7, fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Import selected TSSs from Christ, gencode.v27
# Original locatioN: DATA/scratch/usr/c.leemans/projects/meta_gene_plots/cl20200811_TSS_window_TOP_DRB/selection/
tss_annotated <- read.delim2("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/gencode.v27-fantom-selection.txt", header = F, col.names = c("chr", "TSS_position", "transcript_id", "strand", "gene_id", "expr_avg", "n_tissues"))

# Select 500 bp window upstream of TSS
tss_annotated$promoter_start <- 
  ifelse(tss_annotated$strand == "+", tss_annotated$TSS_position - 499, tss_annotated$TSS_position + 499)

# Save promoter_start as the lowest value
tss_annotated$start <- pmin(tss_annotated$TSS_position, tss_annotated$promoter_start)
tss_annotated$end <- pmax(tss_annotated$TSS_position, tss_annotated$promoter_start)
tss_annotated$width <- 1 + (tss_annotated$end - tss_annotated$start)

# Extract genomic sequence matching the selected promoter_start to TSS locations
gr <- makeGRangesFromDataFrame(tss_annotated)

tss_annotated$seq <- getSeq(BSgenome.Hsapiens.UCSC.hg38::Hsapiens, names = gr, as.character = T)
# # Save file
# save(tss_annotated, file = "/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/mt20210308_hg38_promoters-500bp.RData")
```

### Import scored TSSs

```{r, fig.width=10, fig.height=7, fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Import data
load("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/p53_tss_per_position.Rda")

# # Put the scores of all TSSs into one large data frame - only keep the forward matching
# tss_scores <- data.frame(matrix(ncol = 477, nrow = 1))
# names(tss_scores) <- 1:477
# 
# count <- 0
# for (i in unique(names(tss.per.pos$Scores))) {
#   x <- data.frame(data.frame(tss.per.pos$Scores[i]))[1,]
#   names(x) <- 1:ncol(x)
#   tss_scores <- rbind(tss_scores, x)
#   count <- count + length(i)
#   print(count)
# }
# 
# tss_scores <- tss_scores[-1,]
# rownames(tss_scores) <- names(tss.per.pos$Scores)
# tss_scores <- tss_scores %>%
#   rownames_to_column(var = "transcript_id")
load("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/mt20200309_tss_scored_affinity.RData")


tss_scores <- merge(tss_annotated, tss_scores)

tss_scores[, "max"] <- apply(tss_scores[, 13:489], 1, max)
```


### Import MCF7 RNA-seq gene list

```{r mcf7_up_down_scoring, fig.width=4, fig.height=4, fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Import genes that are differentially expressed according to Aerts paper (Janky et al., iRegulon)
geneSet1 <- read.delim2("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/mcf7_nutlin_upregulated_GSE47041.txt",
                        stringsAsFactors = F) %>%
  mutate(nutlin = "upregulated")
geneSet2 <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/mcf7_nutlin_downregulated.csv",
                      stringsAsFactors = F) %>%
  mutate(nutlin = "downregulated")

# Annotate the tss data frame with the gene symbols
# ensembl <- useMart("ensembl")
# mart <- useDataset("hsapiens_gene_ensembl", mart = ensembl)
# genes <- tss_scores$gene_id
# genes <- gsub("[.].*","",genes)
# G_list <- getBM(filters= "ensembl_gene_id", attributes= c("ensembl_gene_id","hgnc_symbol"),values=genes,mart= mart)
# save(G_list, file = "/DATA/usr/m.trauernicht/data/gene_id_hgnc_converstion.RData")
load("/DATA/usr/m.trauernicht/data/gene_id_hgnc_converstion.RData")
G_list <- G_list %>% 
  dplyr::select("genes" = hgnc_symbol, ensembl_gene_id)
tss_scores$ensembl_gene_id <- gsub("[.].*","",tss_scores$gene_id)
tss_scores <- merge(tss_scores, G_list, all = T)


# Add up-or downregulated information to tss data frame
tss_scores <- Reduce(function(x, y) merge(x, y, all = T, by = "genes"), list(tss_scores, geneSet1, geneSet2))
tss_scores$nutlin.x[is.na(tss_scores$nutlin.x)] <- ""
tss_scores$nutlin.y[is.na(tss_scores$nutlin.y)] <- ""
tss_scores$nutlin <- paste(tss_scores$nutlin.x, tss_scores$nutlin.y, sep = "")
tss_scores$nutlin[tss_scores$nutlin == ""] <- "unchanged"
tss_scores <- tss_scores %>%
  dplyr::select(-nutlin.x, -nutlin.y) %>% 
  unique()
```



### Load the P53-reporter affinities

```{r, fig.width=4, fig.height=4, fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Import data
load("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/p53_per_position.Rda")

# Put the scores of all TSSs into one large data frame - only keep the forward matching
# reporter_scores <- data.frame(matrix(ncol = 215, nrow = 1))
# names(reporter_scores) <- 1:215
# 
# for (i in unique(names(per.pos$Scores))) {
#   x <- data.frame(data.frame(per.pos$Scores[i]))[1,]
#   names(x) <- 1:ncol(x)
#   rownames(x) <- i
#   reporter_scores <- rbind.fill(reporter_scores, x)
# }
# reporter_scores <- reporter_scores[-1,]
# rownames(reporter_scores) <- names(per.pos$Scores)
# reporter_scores <- reporter_scores %>%
#   rownames_to_column(var = "reporter_id")
load("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/reporter_affinity_scores.RData")
```


### Extract P53 RE sequences

```{r ChIP_extraction, fig.width=4, fig.height=4, fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Curated list of P53 targets (Nguyen et al. 2018, NAS, Supplementary Table ST8)
nguyen_p53_REs <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/ST8-p53 cistrome targets.csv") %>%
  mutate(p53RE.center = round((p53RE.start + p53RE.end) / 2),
         p53RE.start = p53RE.center - 500,
         p53RE.end = p53RE.center + 500,
         width = p53RE.end - p53RE.start) %>%
  dplyr::select('chr' = Chr, 'start' = p53RE.start, 'end' = p53RE.end, 'gene' = Current.p53.cistrome.Gene.Symbol, width)

gr <- makeGRangesFromDataFrame(nguyen_p53_REs)

nguyen_p53_REs$seq <- getSeq(BSgenome.Hsapiens.UCSC.hg19::Hsapiens, names = gr, as.character = T)

nguyen_p53_REs <- nguyen_p53_REs %>%
  mutate(name = paste("nguyen", gene, sep = "_")) %>%
  dplyr::select(name, seq)


# Curated list of P53 targets (Haran et al. 2018)
haran_p53_REs <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/haran_p53RE_curated.csv") %>%
  dplyr::select("gene" = Gene.site.name, RE.chromosomal.location) %>%
  mutate(RE.chromosomal.location = gsub("crr", "chr", RE.chromosomal.location))
haran_p53_REs <- haran_p53_REs %>%
  mutate(RE.chromosomal.location = gsub(" -", "-", RE.chromosomal.location))
haran_p53_REs <- haran_p53_REs %>%
  mutate(RE.chromosomal.location = gsub(" ", "-", RE.chromosomal.location))
haran_p53_REs <- haran_p53_REs %>%
  mutate(chr = gsub("(chr[A-z0-9]{1,2}).*", "\\1", RE.chromosomal.location))
haran_p53_REs <- haran_p53_REs %>%
  mutate(start = as.numeric(gsub(".*:(.*)-.*", "\\1", RE.chromosomal.location)))
haran_p53_REs <- haran_p53_REs %>%
  mutate(end = as.numeric(gsub(".*-(.*)", "\\1", RE.chromosomal.location)))
haran_p53_REs <- haran_p53_REs %>%
  mutate(center = round((start + end)/2))
haran_p53_REs <- haran_p53_REs %>%
  mutate(start = center - 500)
haran_p53_REs <- haran_p53_REs %>%
  mutate(end = center + 500)
haran_p53_REs <- haran_p53_REs %>%
  mutate(width = end - start)
haran_p53_REs <- haran_p53_REs %>% 
  dplyr::select(gene, chr, start, end, width) %>%
  unique()
haran_p53_REs <- haran_p53_REs[-251,]

gr <- makeGRangesFromDataFrame(haran_p53_REs)

haran_p53_REs$seq <- getSeq(BSgenome.Hsapiens.UCSC.hg19::Hsapiens, names = gr, as.character = T)

haran_p53_REs <- haran_p53_REs %>%
  mutate(name = paste("haran", gene, sep = "_")) %>%
  dplyr::select(name, seq)


p53_REs <- rbind(nguyen_p53_REs, haran_p53_REs)
p53_REs <- p53_REs[!duplicated(p53_REs$seq),]


# Save to score sequences
p53_REs_export <- p53_REs

# Add other sequences: pMT02 reporters
pMT02_P53 <- ReadFasta("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/output/mt20210111_oligo_pool.fasta") %>%
  filter(str_detect(name, "Trp53")) %>%
  setnames("sequence", "seq") %>%
  mutate(name = paste("pMT02", name, sep = "_"))


# Add other sequences: published reporter sequences
pMT09_P53 <- ReadFasta("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/gen-2/mt20210114_oligo_pool_gen2.fasta") %>%
  filter(str_detect(name, "P53")) %>%
  setnames("sequence", "seq") %>%
  mutate(name = paste("pMT09", name, sep = "_"))

p53_REs_export <- rbind.fill(p53_REs_export, pMT02_P53)
p53_REs_export <- rbind.fill(p53_REs_export, pMT09_P53)


# Add other sequences: regions from Younger et al.
younger_p53_REs <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/younger_et_al_genomic_p53_REs.csv") %>%
  setnames("Sequence", "seq") %>%
  mutate(name = paste("younger", Chrom, Start, sep = "_")) %>%
  dplyr::select(name, seq)

p53_REs_export <- rbind.fill(p53_REs_export, younger_p53_REs)

# Add other sequences: regions from Aerts paper
aerts_p53_REs <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/aerts_P53_REs.csv") %>%
  dplyr::select("name" = corresponding.peak.name, chr, start, stop) %>%
  filter(str_detect(chr, "chr"))

gr <- makeGRangesFromDataFrame(aerts_p53_REs)

aerts_p53_REs$seq <- getSeq(BSgenome.Hsapiens.UCSC.hg19::Hsapiens, names = gr, as.character = T)

aerts_p53_REs <- aerts_p53_REs %>%
  mutate(name = paste("aerts_design", name, sep = "_"))


aerts_p53_REs_peak <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/aerts_P53_REs.csv") %>%
  dplyr::select("name" = corresponding.peak.name, "chr" = Chr, "start" = start.1, "stop" = stop.1) %>%
  filter(str_detect(chr, "chr"))

gr <- makeGRangesFromDataFrame(aerts_p53_REs_peak)

aerts_p53_REs_peak$seq <- getSeq(BSgenome.Hsapiens.UCSC.hg19::Hsapiens, names = gr, as.character = T)

aerts_p53_REs_peak <- aerts_p53_REs_peak %>%
  mutate(name = paste("aerts_peak", name, sep = "_"))

aerts_p53_REs <- rbind.fill(aerts_p53_REs_peak, aerts_p53_REs)

aerts_p53_REs <- aerts_p53_REs %>%
  dplyr::select(name, seq)

aerts_p53_REs <- aerts_p53_REs[!duplicated(aerts_p53_REs$seq),]

p53_REs_export <- rbind.fill(p53_REs_export, aerts_p53_REs)

p53_REs_export$seq <- toupper(p53_REs_export$seq)

# # Save file
# save(p53_REs_export, file = "/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/mt20210412_p53_REs.RData")
```


### Import P53-RE scores

```{r, fig.width=10, fig.height=7, fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Import data
load("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/p53_REs_per_position.Rda")

# # Put the scores of all REs into one large data frame - only keep the forward matching
# RE_scores <- data.frame(matrix(ncol = 1502, nrow = 1))
# names(RE_scores) <- 1:1502
# 
# count <- 0
# for (i in unique(names(re.per.pos$Scores))) {
#   x <- data.frame(data.frame(re.per.pos$Scores[i]))[1,]
#   names(x) <- 1:ncol(x)
#   RE_scores <- rbind.fill(RE_scores, x)
#   count <- count + length(i)
#   print(count)
#   RE_scores$RE[count+1] <- i
# }
# 
# RE_scores <- RE_scores[-1,]
# 
# save(RE_scores, file = "/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/mt20200504_RE_scores_df.RData")

load("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/mt20200504_RE_scores_df.RData")

RE_scores[, "max"] <- apply(RE_scores[, 1:1502], 1, max, na.rm = T)
```





## Maximum affinity at TSSs

```{r, fig.width=4, fig.height=4, fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# What is the maximum affinity at all TSSs? Is there a difference between nutlin-upregulated genes and unchanged genes?
ggplot(tss_scores %>% dplyr::select(nutlin, max) %>% unique(),
       aes(x = nutlin, y = max)) +
  geom_quasirandom() +
  theme_bw()
```
There's not a clearly visible difference between Nutlin-upregulated and unchanged P53 binding in the promoter sequences. We need to check whether there's any specific enrichment at the exact affinities of my reporters to get a better picture.

---

## Highlights: TSS affinity score plotting

```{r individual_promoter_plots, fig.width=4, fig.height=4, fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Filter only relevant genes
tss_scores_2 <- tss_scores[tss_scores$genes != "",]
tss_scores_2 <- tss_scores_2[tss_scores_2$nutlin != "unchanged",]

# Transform to long df
tss_scores_2 <- melt(tss_scores_2, id.vars = c("genes", "ensembl_gene_id", "transcript_id", 
                                             "chr", "TSS_position", "strand", "gene_id", "expr_avg", 
                                             "n_tissues", "promoter_start", "start", "end", "width", "seq", "max", "nutlin"),
                     variable.name = "position")

# Plot enrichment graph for Aerts top hit: CDKN1A

## Import molotool.autosome.ru predicted P53 binding
p53_cdkn1a <- read_tsv("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/CDKN1A_promoter_P53.tsv")

ggplot(data = tss_scores_2 %>% filter(transcript_id == "ENST00000615513.4"), aes(x = rev(as.numeric(position)), y = value)) +
  geom_line() +
  geom_segment(data = p53_cdkn1a %>% filter(Strand == "+", `-log10(P-value)` >= 2), 
               aes(x=488-Start, xend=488-End,y=-0.00001,yend=-0.00001,
                   color = `-log10(P-value)`), size = 2, lineend = "round") +
  theme_bw() +
  xlab("distance to TSS (bp)")+
  scale_color_gradient(low = "#F9E9E4", high = "#DD6B48")+
  ylab("affinity relative to perfect match")+
  ggtitle("P53 affinity across CDKN1A promoter")

# Try some other genes:
other_transcripts <- c("DCP1B","SPATA18")
ggplot(data = tss_scores_2 %>% filter(genes %in% other_transcripts), aes(x = rev(as.numeric(position)), y = value)) +
  geom_line() +
  theme_bw() +
  xlab("distance to TSS (bp)")+
  ylab("affinity relative to perfect match")+
  ggtitle("P53 affinity across promoter")+
  facet_wrap(~genes)

# TRIM22, the gene with the highest score (from all upregulated genes)
ggplot(data = tss_scores_2 %>% filter(transcript_id == "ENST00000414641.5"), aes(x = rev(as.numeric(position)), y = value)) +
  geom_line() +
  theme_bw() +
  xlab("distance to TSS (bp)")+
  scale_color_gradient(low = "#F9E9E4", high = "#DD6B48")+
  ylab("affinity relative to perfect match")+
  ggtitle("P53 affinity across TRIM22 promoter")

# Plot the same plot for the 6 highest scoring promoters
highest_promoters <- tss_scores %>% arrange(desc(max)) %>% 
  slice(1:6)
highest_promoters <- highest_promoters$transcript_id

## Import molotool.autosome.ru predicted P53 binding
p53_CEP85L <- read_tsv("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/CEP85L_P53.tsv") %>%
  filter(Strand == "-") %>%
  dplyr::select(`-log10(P-value)`, Start, End) %>%
  mutate(TF = "ENST00000472713.1")
p53_ENS <- read_tsv("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/ENST00000565189.1_P53.tsv") %>%
  filter(Strand == "+") %>%
  dplyr::select(`-log10(P-value)`, Start, End) %>%
  mutate(TF = "ENST00000565189.1")
p53_MDM4 <- read_tsv("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/MDM4_P53.tsv") %>%
  filter(Strand == "+") %>%
  dplyr::select(`-log10(P-value)`, Start, End) %>%
  mutate(TF = "ENST00000507825.3")
p53_SBF2 <- read_tsv("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/SBF2_P53.tsv") %>%
  filter(Strand == "-") %>%
  dplyr::select(`-log10(P-value)`, Start, End) %>%
  mutate(TF = "ENST00000527019.5")
p53_TAFA2 <- read_tsv("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/TAFA2_P53.tsv")%>%
  filter(Strand == "-") %>%
  dplyr::select(`-log10(P-value)`, Start, End) %>%
  mutate(TF = "ENST00000416284.7")
p53_TRIM22 <- read_tsv("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/TRIM22_P53.tsv")%>%
  filter(Strand == "+") %>%
  dplyr::select(`-log10(P-value)`, Start, End) %>%
  mutate(TF = "ENST00000414641.5")
molo_scoring <- rbind(p53_CEP85L, p53_ENS, p53_MDM4, p53_SBF2, p53_TAFA2, p53_TRIM22)

for (i in unique(molo_scoring$TF)) {
  p<- ggplot(data = tss_scores %>% 
         filter(transcript_id == i) %>%
         melt(id.vars = c("genes", "ensembl_gene_id", "transcript_id",
                          "chr", "TSS_position", "strand", "gene_id", 
                          "expr_avg", "n_tissues", "promoter_start", 
                          "start", "end", "width", "seq", "max", "nutlin"),
                     variable.name = "position"), 
       aes(x = rev(as.numeric(position)), y = value)) +
  geom_line() +
  geom_segment(data = molo_scoring %>% filter(TF == i), 
               aes(x=488-Start, xend=488-End,y=-0.01,yend=-0.01,
                   color = `-log10(P-value)`), size = 2, lineend = "round") +
  theme_bw() +
  xlab("distance to TSS (bp)")+
  scale_color_gradient(low = "#F9E9E4", high = "#DD6B48")+
  ylab("affinity relative to perfect match")+
    ylim(-0.01,0.15)+
  ggtitle("P53 affinity at the highest-scoring promoters", subtitle = i)
  
  print(p)
}
```

---

## P53 reporter affinity plot

```{r reporter_scoring, fig.width=4, fig.height=4, fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
load("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/cDNA_df.RData")
cDNA_df <- cDNA_df %>%
  dplyr::select("reporter_id" = reporter_id_3, neg_ctrls, affinity_id) %>%
  unique()
reporter_scores <- merge(reporter_scores, cDNA_df, all = T)

reporter_scores_sel <- reporter_scores %>%
  filter(neg_ctrls == "No", str_detect(affinity_id, "2|3"))

reporter_scores_sel <- melt(reporter_scores_sel, variable.name = "position", 
                            value.name = "affinity",
                            id.vars = c("neg_ctrls", "affinity_id", "reporter_id"))

example_reporter <- 
  reporter_scores_sel[reporter_scores_sel$reporter_id %in%
                        c("Trp53_mCMV_p_0_s_0_d_10_bg_1_a1_2_a2_2_a3_2_a4_2_bc_1",
                          "Trp53_mCMV_p_0_s_0_d_10_bg_1_a1_3_a2_3_a3_3_a4_3_bc_1"),]

colors <- c(color = '#A1B0AB', "#373930")
ggplot(example_reporter, aes(x = as.numeric(position), y = affinity, color = affinity_id)) +
  geom_point()+
  xlim(0,215)+
  theme_bw()+
  ylab("affinity relative to perfect match")+
  scale_color_manual(values = colors)+
  xlab("reporter position")+
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))+
  theme(legend.position = "bottom")
```
Since the genomic promoter elements are not in the same affinity-range as the reporter sequences, it seems like a good idea to re-score different sequences that have stronger P53 binding sites. So, I will now score curated lists of genomic P53 response elements instead and also take along published P53 reporter sequences (TF-seq, 13-copy, PUMA, ...) and my pMT02 reporter sequences. 

---

## maximum P53-RE scores

```{r max_scores_all_promoters, fig.width=4, fig.height=4, fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Select highest scoring sequences
RE_scores_filt <- RE_scores %>%
  filter(max >= 0.000025) %>%
  melt(id.vars = c("RE", "max"), variable.name = "position", value.name = "affinity") %>%
  na.omit()

RE_scores_max <- RE_scores_filt %>%
  dplyr::select(RE, max) %>%
  unique()

# Include reporter affinities
reporter_scores_max <- reporter_scores
reporter_scores_max[, "max"] <- apply(reporter_scores[, 2:ncol(reporter_scores)], 1, max, na.rm = T)
reporter_scores_max <- reporter_scores_max %>%
  melt(id.vars = c("reporter_id", "max"), variable.name = "position", value.name = "affinity") %>%
  na.omit() %>%
  dplyr::select(reporter_id, max) %>%
  unique()
reporter_scores_max <- reporter_scores_max[!duplicated(reporter_scores_max$max),]
reporter_scores_max <- reporter_scores_max[1:5,]
reporter_scores_max <- reporter_scores_max %>%
  setnames("reporter_id", "RE") %>%
  mutate(group = "pMT06-RE")

# Curate the different sources
RE_scores_max <- RE_scores_max[!duplicated(RE_scores_max$max),]
RE_scores_max$group <- "chip_peaks_aerts"
RE_scores_max$group[grep("haran|nguyen", RE_scores_max$RE)] <- "curated REs"
RE_scores_max$group[grep("younger", RE_scores_max$RE)] <- "chip_peaks_younger"
RE_scores_max$group[grep("promega|romanov", RE_scores_max$RE)] <- "published REs"
RE_scores_max$group[grep("pMT02", RE_scores_max$RE)] <- "pMT02-RE"
RE_scores_max <- rbind(reporter_scores_max, RE_scores_max)

ggplot(RE_scores_max %>%
         filter(group != "pMT02-RE", group != "published REs"), aes(x = group, y = max)) +
    
  geom_hline(yintercept = 1, linetype = "dashed", alpha = 0.4) +
  geom_rect(aes(ymin = 3.708171e-01, ymax = 1,
                xmin = -Inf, xmax = Inf), fill = '#adb5bd', alpha = 0.5) +
  
  geom_hline(yintercept = 3.708171e-01, linetype = "dashed", alpha = 0.4) +
  geom_rect(aes(ymin = 1.382304e-01, ymax = 3.708171e-01,
                xmin = -Inf, xmax = Inf), fill = '#ced4da', alpha = 0.4) +
  
  geom_hline(yintercept = 1.382304e-01, linetype = "dashed", alpha = 0.4) +
  geom_rect(aes(ymin = 5.752735e-02, ymax = 1.382304e-01,
                xmin = -Inf, xmax = Inf), fill = '#dee2e6', alpha = 0.3) +
  
  geom_hline(yintercept = 5.752735e-02, linetype = "dashed", alpha = 0.4) +
  geom_rect(aes(ymin = 3.539419e-06, ymax = 5.752735e-02,
                xmin = -Inf, xmax = Inf), fill = '#e9ecef', alpha = 0.2) +
  
  geom_hline(yintercept = 3.539419e-06, linetype = "dashed", alpha = 0.4) +
  
  geom_quasirandom(width = 0.4, varwidth = T) +
  xlab("")+
  ylab("RE affinity relative to perfect match")+
  theme_classic() +
  coord_flip()

```
Looks like there are already some higher affinities - let's have a closer look

---

## Highlights: affinity scores of genomic P53-REs

```{r}
p <- RE_scores_filt %>% filter(RE == "nguyen_CDKN1A") 

ggplot(data = p, 
       aes(x = -((nrow(p)/2)-1):(nrow(p)/2), 
           y = affinity)) +
  geom_line() +
  theme_bw() +
  xlab("distance to RE center (bp)")+
  scale_color_gradient(low = "#F9E9E4", high = "#DD6B48")+
  ylab("affinity relative to perfect match")+
  xlim(-50,50) +
  ggtitle("P53 affinity across CDKN1A RE")


p <- RE_scores_filt %>% filter(RE == "younger_chr7_121151311") 

ggplot(data = p, 
       aes(x = -((nrow(p)/2)-1):(nrow(p)/2), 
           y = affinity)) +
  geom_line() +
  theme_bw() +
  xlab("distance to RE center (bp)")+
  scale_color_gradient(low = "#F9E9E4", high = "#DD6B48")+
  ylab("affinity relative to perfect match")+
  ggtitle("P53 affinity across younger_chr7_121151311")

p <- RE_scores_filt %>% filter(RE == "nguyen_GADD45A") 

ggplot(data = p, 
       aes(x = -((nrow(p)/2)-1):(nrow(p)/2), 
           y = affinity)) +
  geom_line() +
  theme_bw() +
  xlab("distance to RE center (bp)")+
  scale_color_gradient(low = "#F9E9E4", high = "#DD6B48")+
  ylab("affinity relative to perfect match")+
  xlim(-50,50) +
  labs(title = "P53 affinity across GADD45A RE",
       subtitle = "promega reporter & only 1 mismatch from consensus")


p <- RE_scores_filt %>% filter(RE == "haran_MDM2_RE1")

ggplot(data = p, 
       aes(x = -((nrow(p)/2)-1):(nrow(p)/2), 
           y = affinity)) +
  geom_line() +
  theme_bw() +
  xlab("distance to RE center (bp)")+
  scale_color_gradient(low = "#F9E9E4", high = "#DD6B48")+
  ylab("affinity relative to perfect match")+
  xlim(-50,50) +
  labs(title = "P53 affinity across MDM2 RE",
       subtitle = "two adjacent binding sites")

p <- RE_scores_filt %>% filter(RE == "nguyen_FOSL1")

ggplot(data = p, 
       aes(x = -((nrow(p)/2)-1):(nrow(p)/2), 
           y = affinity)) +
  geom_line() +
  theme_bw() +
  xlab("distance to RE center (bp)")+
  scale_color_gradient(low = "#F9E9E4", high = "#DD6B48")+
  ylab("affinity relative to perfect match")+
  xlim(-50,50) +
  labs(title = "P53 affinity across FOSL1 RE",
       subtitle = "two adjacent binding sites")


# Take the highest 20
relevant <- c("aerts", "younger", "haran", "nguyen")
p <- RE_scores_filt[grep(paste(relevant, collapse = "|"), RE_scores_filt$RE),] %>% 
  arrange(desc(max)) %>% 
  top_n(20)

p <- RE_scores_filt[RE_scores_filt$RE %in% p$RE,]

ggplot(data = p %>% 
         arrange(desc(max)), 
       aes(x = as.numeric(position), 
           y = affinity)) +
  geom_line() +
  theme_bw() +
  xlab("position (bp)")+
  scale_color_gradient(low = "#F9E9E4", high = "#DD6B48")+
  ylab("affinity relative to perfect match")+
  labs(title = "P53 affinity - highest matches") +
  facet_wrap(~RE, scales = "free_x", ncol = 3)

ggplot(data = p, 
       aes(x = as.numeric(position), 
           y = affinity,
           color = RE)) +
  geom_line() +
  theme_bw() +
  xlab("distance to RE center (bp)")+
  ylab("affinity relative to perfect match")+
  labs(title = "P53 affinity across FOSL1 RE",
       subtitle = "two adjacent binding sites")

```

---

## Number of binding sites per P53-RE

```{r}
# Select sequences with scores higher than 0.00025 and check the number of positions higher than 0.00025 for this sequence
RE_scores_filt <- RE_scores %>%
  filter(max >= 0.000025) %>%
  melt(id.vars = c("RE", "max"), variable.name = "position", value.name = "affinity") %>%
  na.omit()


RE_scores_filt <- RE_scores_filt %>%
  filter(affinity >= 0.000025)

RE_scores_filt <- RE_scores_filt[-grep("promega|romanov", RE_scores_filt$RE),] 
RE_scores_filt <- RE_scores_filt[-grep("pMT02", RE_scores_filt$RE),]

# Boxplot of number of REs with 1,2,3,4 binding sites
x <- data.frame(RE = unique(RE_scores_filt$RE), n = "")
for(i in unique(RE_scores_filt$RE)){

  dif <- max(as.numeric(RE_scores_filt$position[RE_scores_filt$RE == i])) - min(as.numeric(RE_scores_filt$position[RE_scores_filt$RE == i]))

  
  if(dif <= 20 & dif > 0) {
    x$n[x$RE == i] <- 1
  } else{
  
  x$n[x$RE == i] <- nrow(RE_scores_filt[RE_scores_filt$RE == i,])
  }
}

x$n <- as.numeric(x$n)
x$n[x$n >= 7] <- 7

ggplot(x, aes(x = n)) +
  geom_bar(stat = "count", fill = "#E29578") +
  theme_classic() +
  scale_x_continuous(name="number of binding sites", breaks = 1:7, labels = c("1", "2", "3", "4", "5", "6", ">=7")) +
  theme(text = element_text(size=16)) +
  coord_flip()

# Plot distance between binding sites
y <- x %>% filter(n == 2)
RE_scores_dist <- RE_scores_filt[RE_scores_filt$RE %in% y$RE,]

for(i in unique(RE_scores_dist$RE)){

  RE_scores_dist$dist[RE_scores_dist$RE == i] <- 
    max(as.numeric(RE_scores_dist$position[RE_scores_dist$RE == i])) - min(as.numeric(RE_scores_dist$position[RE_scores_dist$RE == i])) - 24
}

RE_scores_dist <- RE_scores_dist %>%
  dplyr::select(RE, dist) %>%
  unique()

ggplot(RE_scores_dist %>% filter(dist <= 75), aes(x = dist)) +
  geom_histogram(aes(y = ..density..), binwidth = 1, fill = "white", color = "black") +
  geom_density(adjust = 0.1, fill = "#E29578", alpha = 0.4)+
  theme_bw() + 
  theme(text = element_text(size=16)) +
  scale_x_continuous(name="distance between 2 binding sites", breaks = seq(0, 70, 5))


```

---

## Number of sequences with perfect 2xCATG match

```{r}
# Extract sequences that have 2X CATG
CATG_REs <- p53_REs_export %>%
  filter(p53_REs_export$name %in% RE_scores_filt$RE)

CATG_REs <- CATG_REs[grep("CATG[A-Z]{6}CATG", CATG_REs$seq),]
```


---

## Compute cluster-buster scores

```{r cluster_buster, fig.width=4, fig.height=4, fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Make a file with genomic REs and the chosen reporter REs 
reporter_REs <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/parameter_files/TF_motifs.csv") %>%
  filter(TF == "Trp53") %>%
  mutate(seq.name = paste(TF, Affinity, sep = "_")) %>%
  dplyr::select(seq.name, 'seq.text' = Motif)

p53_REs <- p53_REs_export %>%
  dplyr::select('seq.name' = name, 'seq.text' = seq)

p53_REs <- rbind(p53_REs, reporter_REs)
dat2fasta(p53_REs, outfile = "/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/p53_REs.fasta")  

# cluster-buster code:
# ./cbust -m 0 -c 0 -f 5 /DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/p53_pfms /DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/p53_REs.fasta > /DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/p53_2.tsv


# Scan with ~6 different p53 motifs - plot each mean score per affinity categorie in a quasirandom plot
reporter_cbust_scores <- read_tsv("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/mt20210413_p53_RE_scoring.tsv") %>%
  setnames("# chrom", "gene") %>%
  filter(cluster_or_motif == "motif", cluster_id_or_motif_name != "M11198_2.00_transfac",
         cluster_id_or_motif_name != "MA0106.1_SELEX", cluster_id_or_motif_name != "M11197_2.00_transfac") %>%
  mutate(width = relative_end__bed - relative_start__bed) %>%
  group_by(gene, cluster_id_or_motif_name) %>%
  top_n(1, cluster_or_motif_score) %>%
  ungroup() %>%
  dplyr::select(gene, 'pwm' = cluster_id_or_motif_name, 'score' = cluster_or_motif_score, width, 'start' = relative_start__bed, 
                'end' = relative_end__bed, motif_sequence) %>%
  unique()

# Annotate REs classes
reporter_cbust_scores$response_element <- "nguyen_REs"
reporter_cbust_scores$response_element[grep("haran", reporter_cbust_scores$gene)] <- "haran_REs"
reporter_cbust_scores$response_element[grep("younger", reporter_cbust_scores$gene)] <- "younger_REs"
reporter_cbust_scores$response_element[grep("aerts", reporter_cbust_scores$gene)] <- "aerts_REs"
reporter_cbust_scores$response_element[grep("Trp53", reporter_cbust_scores$gene)] <- gsub("Trp53_([0-4]{1})", "\\1", reporter_cbust_scores$gene[grep("Trp53", reporter_cbust_scores$gene)])
reporter_cbust_scores <- reporter_cbust_scores[-grep("pMT02", reporter_cbust_scores$gene),]


reporter_cbust_scores$response_element[reporter_cbust_scores$response_element == "0"] <- "4_high-affinity reporter"
reporter_cbust_scores$response_element[reporter_cbust_scores$response_element == "1"] <- "3_med-affinity reporter"
reporter_cbust_scores$response_element[reporter_cbust_scores$response_element == "2"] <- "2_low-affinity reporter"
reporter_cbust_scores$response_element[reporter_cbust_scores$response_element == "3"] <- "1_very_low-affinity reporter"
reporter_cbust_scores$response_element[reporter_cbust_scores$response_element == "4"] <- "0_zero-affinity reporter"
reporter_cbust_scores$puma <- "No"
puma <- c("BBC3")
reporter_cbust_scores$puma[grep(paste(puma, collapse = "|"), reporter_cbust_scores$gene)] <- "Yes"

# Compute cluster buster score relative to max per pwm
reporter_cbust_scores$rel_score <- ave(reporter_cbust_scores$score, reporter_cbust_scores$pwm, FUN = function(x) x/max(x))



# Plot the motif score per affinity 
colors <- c("black", "red")
ggplot(reporter_cbust_scores, aes(y = rel_score, x = response_element, color = puma)) +
  geom_quasirandom(dodge.width = 0.75, alpha = 0.4) +
  geom_boxplot(alpha = 0.4) +
  ylab("cluster-buster score relative to maximum per pwm") +
  scale_color_manual(values = colors) +
  theme_bw() +
  coord_flip() +
  facet_wrap(~pwm)
```




---

### Session Info

```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

