---
title: "Barcode processing - P53 Deep Scan"
author: 
  - name: "Max Trauernicht"
    email: "m.trauernicht@nki.nl"
    affiliation: "Netherlands Cancer Institute - van Steensel lab"
date: '`r format(Sys.time(), "%d/%m/%Y")`'
output: 
  html_document:
    theme: united
    highlight: pygments
    fig_caption: yes
    code_folding: hide
    df_print: kable
    toc: true
    toc_depth: 4
    toc_float:
      collapsed: false
---

---

### Aim
pMT06 was transfected into various cell types and mRNA barcode counts were sequenced together with pMT06 pDNA counts. In four different sequencing runs, data for all P53 & GR reporters was collected, and will be analyzed here. 

---

## Setup {.tabset}

<!-- little HTML script to do indentation of the table of contents -->
<script>
    $(document).ready(function() {
      $items = $('div#TOC li');
      $items.each(function(idx) {
        num_ul = $(this).parentsUntil('#TOC').length;
        $(this).css({'text-indent': num_ul * 10, 'padding-left': 0});
      });

    });
</script>

```{css, echo = FALSE}
div.sourceCode {
  overflow-x: hidden;
}
```



### Libraries

```{r setup, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(data.table)
library(plyr)
library(stringr)
library(ggpubr)
library(GGally)
library(vwr)
library(dplyr)
library(tibble)
library(plotly)
library(ggbeeswarm)
library(haven)
library(readr)
library(parallel)
library(RColorBrewer)
library(gridExtra)
library(pheatmap)
library(shiny)
library(factoextra)
library(ggbiplot)
library(ggpointdensity)
library(viridis)
library(tidyr)
library(DESeq2)
library(PCAtools)
```


### Functions

```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
#Custom functions
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}


# From Fede:
# ggpairs custom functions
corColor <- function(data, mapping, color = I("black"), sizeRange = c(1, 3), ...) {

  x   <- eval_data_col(data, mapping$x)
  y   <- eval_data_col(data, mapping$y)
  r   <- cor(x, y, "pairwise.complete.obs")
  rt  <- format(r, digits = 3)
  tt  <- as.character(rt)
  cex <- max(sizeRange)

  # helper function to calculate a useable size
  percent_of_range <- function(percent, range) {
    percent * diff(range) + min(range, na.rm = TRUE)
  }

  # plot correlation coefficient
  p <- ggally_text(label = tt, mapping = aes(), xP = 0.5, yP = 0.5,
                   size = I(percent_of_range(cex * abs(r), sizeRange)), color = color, ...) +
    theme(panel.grid.minor=element_blank(),
          panel.grid.major=element_blank())

  corColors <- RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")[2:6]

  if (r <= boundaries[1]) {
    corCol <- corColors[1]
  } else if (r <= boundaries[2]) {
    corCol <- corColors[2]
  } else if (r < boundaries[3]) {
    corCol <- corColors[3]
  } else if (r < boundaries[4]) {
    corCol <- corColors[4]
  } else {
    corCol <- corColors[5]
  }

  p <- p +
    theme(panel.background = element_rect(fill = corCol))

  return(p)
}
```


### Loading data

```{r data import, fig.width=10, fig.height=7, fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Import barcode counts per condition - gcf6210
bc_files = list.files('/DATA/usr/m.trauernicht/projects/P53_reporter_scan/data/gcf6210/results/',
                       full.names=T, patter='*_barcode_counts.tsv')
bc_list <- lapply(bc_files, fread, header = FALSE)
names(bc_list)<- gsub('.*//6210_[0-9]{1,2}_(.*?)_[CGAT]{6}.*_barcode_counts.tsv', 
                                    '\\1', 
                                    bc_files)
for (i in 1:(length(bc_list))) {
  bc_list[[i]]$gcf <- "gcf6210"
}


# Import barcode counts per condition - gcf6301
bc_files = list.files('/DATA/usr/m.trauernicht/projects/P53_reporter_scan/data/gcf6301/results/',
                       full.names=T, patter='*_barcode_counts.tsv')
bc_list_2 <- lapply(bc_files, fread, header = FALSE)
names(bc_list_2)<- gsub('.*//6301_[0-9]{1,2}_(.*?)_[CGAT]{6}.*_barcode_counts.tsv',
                                    '\\1',
                                    bc_files)
for (i in 1:(length(bc_list_2))) {
  bc_list_2[[i]]$gcf <- "gcf6301"
}
bc_list <- c(bc_list, bc_list_2)


# Import gcf6412 sequencing data
bc_files = list.files('/DATA/usr/m.trauernicht/projects/P53_reporter_scan/data/gcf6412/results/',
                       full.names=T, patter='*_barcode_counts.tsv')
bc_list_3 <- lapply(bc_files, fread, header = FALSE)
names(bc_list_3)<- gsub('.*//6412_[0-9]{1,2}_(.*?)_[CGAT]{6}.*_barcode_counts.tsv',
                                    '\\1',
                                    bc_files)
bc_list_3 <- bc_list_3[names(bc_list_3) != 
                          "/DATA/usr/m.trauernicht/projects/P53_reporter_scan/data/gcf6412/results//6412_pool_GATCxG-AGCGCTAG_S2_barcode_counts.tsv"]
for (i in 1:(length(bc_list_3))) {
  bc_list_3[[i]]$gcf <- "gcf6412"
}
bc_list <- c(bc_list, bc_list_3)


# Import gcf6502 sequencing data
bc_files = list.files('/DATA/usr/m.trauernicht/projects/P53_reporter_scan/data/gcf6502/results_d1/',
                       full.names=T, patter='.*pMT06.*_barcode_counts.tsv')
bc_list_4 <- lapply(bc_files, fread, header = FALSE)
names(bc_list_4)<- gsub('.*//6502_[0-9]{1,2}_pMT06_(.*?)_[CGAT]{6}.*_barcode_counts.tsv',
                                    '\\1',
                                    bc_files)
for (i in 1:(length(bc_list_4))) {
  bc_list_4[[i]]$gcf <- "gcf6501"
}
bc_list <- c(bc_list, bc_list_4)


# Import barcode annotation
bc_annotation <- read.csv2("/DATA/usr/m.trauernicht/projects/P53_reporter_scan/data/library_design/output/tf_df_complete.csv", header = T) %>% 
  dplyr::select(barcode, tf, oligo.barcode, 
                spacing, promoter,
         position, distance, background, affinity_pos1,
         affinity_pos2, affinity_pos3, affinity_pos4, seq.name) %>%
  setnames("seq.name", "reporter_id")
```

### Creating count data frames

```{r cluster_compare, fig.width=10, fig.height=7, fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Generate long dfs (I merge instead of rbind the data frames because I want to include barcodes with 0 counts)
#bc_df <- bind_rows(bc_list, .id = "column_label")

for (i in 1:length(bc_list)) {
  if (i == 1) {
  bc_df <- data.frame(bc_list[i])
  bc_df[3] <- paste(names(bc_list[i]), unique(bc_list[[i]]$gcf), sep = "_")
  bc_df <- bc_df[,c(1,2,3)]
  names(bc_df) <- c("barcode", "starcode_counts", "sample")
  bc_df <- reshape2::dcast(bc_df, barcode ~ sample, value.var = "starcode_counts")
  }
  else {
  bc_df_i <- data.frame(bc_list[i])
  bc_df_i[3] <- paste(names(bc_list[i]), unique(bc_list[[i]]$gcf), sep = "_")
  bc_df_i <- bc_df_i[,c(1,2,3)]
  names(bc_df_i) <- c("barcode", "starcode_counts", "sample")
  bc_df_i <- reshape2::dcast(bc_df_i, barcode ~ sample, value.var = "starcode_counts")
  bc_df <- merge(bc_df, bc_df_i, all = T)
  }
}

# Put in a long format
bc_df <- bc_df %>%
  reshape2::melt(id.vars = "barcode", value.name = "starcode_counts", variable.name = "sample")

# Match designed barcodes
bc_df <- merge(bc_df, bc_annotation, all = T)

# Remove non-matched barcodes
bc_df <- bc_df[!is.na(bc_df$tf),]

# Remove pMT02 data
bc_df <- bc_df[-grep("pMT02", bc_df$sample),]

# Rename conditions
bc_df$sample <- gsub("_pMT06", "", bc_df$sample)
bc_df$sample <- gsub("MCF7-", "MCF7_", bc_df$sample)
bc_df$sample <- gsub("A549_Dex-", "A549_Dex", bc_df$sample)
bc_df$sample <- gsub("Nut_r", "Nutlin_r", bc_df$sample)

# Split up in replicates, conditions, and samples
bc_df$condition <- gsub("_r.*", "", bc_df$sample)
bc_df$sample <- gsub("(_r)ep(.*)", "\\1\\2", bc_df$sample)
bc_df$rep <- gsub(".*_(r.*)_gcf.*", "\\1", bc_df$sample)
bc_df$gcf <- gsub(".*(gcf.*)", "\\1", bc_df$sample)

# # Only keep P53 data
# bc_df <- bc_df[-grep("Gr", bc_df$tf),]
# bc_df$tf <- gsub("Trp", "P", bc_df$tf)

# First compute reads per million to estimate the relative counts in their respective sample
bc_df$starcode_counts[is.na(bc_df$starcode_counts)] <- 0
for (i in unique(bc_df$sample)) {
  bc_df$rpm[bc_df$sample == i] <- (bc_df$starcode_counts[bc_df$sample == i] + 1) / # Adds a pseudocount of 1
    sum(bc_df$starcode_counts[bc_df$sample == i]) *1e6
}
```


## Investigate variance of the read counts between individual samples

```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Make dataframe to compare raw counts between conditions
bc_df_filt <- bc_df %>%
  filter(tf == "Trp53")
bc_df_filt <- bc_df_filt[!is.na(bc_df_filt$condition),]
bc_df_cor <- bc_df_filt %>%
  dplyr::select(rpm, barcode, sample)

bc_df_cor <- bc_df_cor %>%
  mutate(rpm = as.integer(rpm)) %>%
  filter(str_detect(sample, "gcf6412") | str_detect(sample,"gcf6501")) %>%
  mutate(sample = gsub("MCF7_PFTa", "MCF7_WT_PFT", sample)) %>% 
  mutate(sample = gsub("MyTaq_pDNA", "pDNA_pDNA_pDNA", sample)) %>% 
  mutate(sample = gsub("pDNA-GR", "pDNA_pDNA_pDNA_r1", sample)) %>% 
  mutate(sample = gsub("PFTa", "PFT", sample)) %>% 
  mutate(sample = gsub("MCF7_DMSO", "MCF7_WT_DMSO", sample)) %>% 
  mutate(sample = gsub("MCF7_Nutlin", "MCF7_WT_Nutlin", sample)) %>% 
  filter(str_detect(sample, "MCF7_WT")| str_detect(sample, "pDNA")) %>%
  spread(sample, rpm) 

bc_df_cor[is.na(bc_df_cor)] <- 0


# Convert into data.frame with row.names for deseq2
bc_df_cor2 <- bc_df_cor %>%
  column_to_rownames("barcode")


#######################################
## Initialize deseq2

# Prepare metadata ready for deseq
metadata <- data.frame("sample_name" = names(bc_df_cor2)) %>%
  #filter(sample_name != "pDNA_1", sample_name != "pDNA_2") %>%
  mutate(cell = gsub("_.*", "", sample_name),
         condition = gsub("^.*_(.*)_r[1-3].*", "\\1", sample_name),
         replicate = gsub(".*_(r[1-3])_.*", "\\1", sample_name),
         gcf = gsub(".*(gcf.*)", "\\1", sample_name))
  #filter(stringr::str_detect(condition, "pDNA", negate = T))


# Initialize
bc_df_cor3 <- bc_df_cor2[,-grep("pDNA", names(bc_df_cor2))]
bc_df_cor2$mean <- rowMeans(bc_df_cor2)
bc_df_cor2 <- bc_df_cor2 %>%
  filter(mean > 200) %>%
  dplyr::select(-mean)


rnaseq_dds <- DESeqDataSetFromMatrix(countData = bc_df_cor2,
                                     colData = metadata,
                                     design= ~ condition)

# Execute deseq2
rnaseq_dds <- DESeq(rnaseq_dds)


#######################################
## PCA analysis - quality control

# Get the "normalized" values and create PCA plot
# Note that "normalized" is simply log2 + 0.01 transformed normalized counts!
rnaseq_dds_norm <- normTransform(rnaseq_dds, pc = 0.01)
plt <- plotPCA(rnaseq_dds_norm, intgroup = c("gcf"), ntop = 50)

plt + 
  theme_bw() +
  theme(aspect.ratio = 1) +
  scale_shape_manual()

## PCA plot alternative
rnaseq_dds_norm <- assay(vst(rnaseq_dds))
p <- pca(bc_df_cor2, metadata = metadata %>%
           column_to_rownames("sample_name"), removeVar = 0.1)
biplot(p, 
       colby = "condition", 
       legendPosition = "right",
       legendLabSize = 8,
       legendTitleSize = 8,
       legendIconSize = 3, 
       lab = "")


## Correlation heatmap
bc_df_cor2$mean <- rowMeans(bc_df_cor2)
bc_df_cor2 <- bc_df_cor2 %>%
  filter(mean > 200) %>%
  dplyr::select(-mean)

cor_df <- cor(bc_df_cor2, method = "pearson")

pheatmap(cor_df, border_color = "black")

plot(bc_df_cor2$MCF7_WT_DMSO_r1_gcf6412, bc_df_cor2$MCF7_WT_DMSO_r3_gcf6412)

# Get cluster information
clusters <- hclust(dist(cor_df))
cutree(clusters, h = 7) 
plot(clusters)
abline(h = 3, col = "red")

# Apply cutree() to hclust.pokemon: cut.pokemon
cut.pokemon<-cutree(clusters,k=3)
##############################################################
# Initialize total within sum of squares error: wss
wss <- 0
# Look over 1 to 15 possible clusters
for (i in 1:13) {
  # Fit the model: km.pokemon
  km.pokemon <- kmeans(cor_df, centers = i, nstart = 20, iter.max = 50)
  # Save the within cluster sum of squares
  wss[i] <- km.pokemon$tot.withinss
}
# Produce a scree plot
plot(1:13, wss, type = "b", 
     xlab = "Number of Clusters", 
     ylab = "Within groups sum of squares")

```
It looks like the the most variance is between the sequencing runs itself. But this makes sense because I used two different pDNA libraries for the 4 sequencing runs. Some samples correlate with pDNA counts - these samples have pDNA contamination and need to be removed.


---

## Read distribution

```{r read_distribution, fig.width=10, fig.height=7, fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# I want to show the following:
## 1: Read distribution of matched barcodes vs. unmatched barcode
bc_df_reads <- bc_df[!is.na(bc_df$tf),] 
bc_df_reads <- bc_df_reads %>% 
  dplyr::group_by(sample, gcf) %>% 
  mutate(seq_sum = sum(starcode_counts))

plot_ly(bc_df_reads[bc_df_reads$gcf == "gcf6210",] %>% dplyr::select(sample, seq_sum) %>% unique(), x = ~sample, y = ~seq_sum, type = 'bar',
             marker = list(color = '#D6D5C9',
                           line = list(color = 'rgb(8,48,107)', width = 1.5))) %>% 
  layout(title = "Percentage of reads in sequencing run 1",
         yaxis = list(title = "Percentage of matched reads"),
         xaxis = list(title = "sample"))

plot_ly(bc_df_reads[bc_df_reads$gcf == "gcf6301",] %>% dplyr::select(sample, seq_sum) %>% unique(), x = ~sample, y = ~seq_sum, type = 'bar',
             marker = list(color = '#D6D5C9',
                           line = list(color = 'rgb(8,48,107)', width = 1.5))) %>% 
  layout(title = "Percentage of reads in sequencing run 2",
         yaxis = list(title = "Percentage of matched reads"),
         xaxis = list(title = "sample"))

plot_ly(bc_df_reads[bc_df_reads$gcf == "gcf6412",] %>% dplyr::select(sample, seq_sum) %>% unique(), x = ~sample, y = ~seq_sum, type = 'bar',
             marker = list(color = '#D6D5C9',
                           line = list(color = 'rgb(8,48,107)', width = 1.5))) %>% 
  layout(title = "Percentage of reads in sequencing run 3",
         yaxis = list(title = "Percentage of matched reads"),
         xaxis = list(title = "sample"))

plot_ly(bc_df_reads[bc_df_reads$gcf == "gcf6501",] %>% dplyr::select(sample, seq_sum) %>% unique(), x = ~sample, y = ~seq_sum, type = 'bar',
             marker = list(color = '#D6D5C9',
                           line = list(color = 'rgb(8,48,107)', width = 1.5))) %>% 
  layout(title = "Percentage of reads in sequencing run 4",
         yaxis = list(title = "Percentage of matched reads"),
         xaxis = list(title = "sample"))

for (i in unique(bc_df$gcf)) {
  p <- ggplot(bc_df[!is.na(bc_df$tf) & bc_df$gcf == i,], aes(x = tf, y = rpm)) +
  geom_jitter(alpha = 0.1) +
  theme_bw() +
  ylim(0,2500) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, size = 6)) +
  facet_wrap(~sample) + ggtitle(i)
  print(p)
}

bc_df_2 <- bc_df[bc_df$rpm <= 250,]
bc_df_2 <- bc_df_2[bc_df_2$rpm >= 0.5,]
bc_df_2 <- bc_df_2[!is.na(bc_df_2$tf),]

for (i in unique(bc_df$gcf)) {
  p <- ggplot(bc_df_2[bc_df_2$gcf == i,], aes(x = rpm)) +
  geom_histogram(binwidth = 10) +
  theme_bw() +
  xlim(0,250)+
  ylim(0,1000)+
  facet_wrap(~sample)+
  theme(strip.background =element_rect(fill="#D6D5C9")) +
    ggtitle(i)
  print(p)
}

for (i in unique(bc_df$gcf)) {
  p <- ggplot(bc_df[bc_df$rpm >= 1000 & !is.na(bc_df$tf),] %>%
                filter(gcf == i), aes(x = rpm)) +
    geom_histogram(binwidth = 40) +
    theme_bw() +
    xlim(1000,2000)+
    ylim(0,25)+
    facet_wrap(~sample)+
    theme(strip.background =element_rect(fill="#D6D5C9")) +
    ggtitle(i)
  print(p)
}

n_highly_expressed <- data.frame("sample" = unique(bc_df$sample),
                                 "n_bc" = "", stringsAsFactors=FALSE)
for (i in unique(bc_df$sample)) {
  n_highly_expressed$n_bc[n_highly_expressed$sample == i] <- 
    length(bc_df$barcode[bc_df$rpm > 1000 & bc_df$sample == i])
}

plot_ly(n_highly_expressed, x = ~sample, y = ~as.numeric(n_bc), type = 'bar',
             marker = list(color = '#D6D5C9',
                           line = list(color = 'rgb(8,48,107)', width = 1.5))) %>% 
  layout(title = "Highly expressed barcodes",
         yaxis = list(title = "Number of barcodes with > 500 rpm"),
         xaxis = list(title = "sample"))

n_highly_expressed <- data.frame("sample" = unique(bc_df$sample),
                                 "n_bc" = "", stringsAsFactors=FALSE)
bc_df_2 <- bc_df[grep("random", bc_df$tf),]
for (i in unique(bc_df$sample)) {
  n_highly_expressed$n_bc[n_highly_expressed$sample == i] <- 
    length(bc_df_2$barcode[bc_df_2$rpm > 300 & bc_df_2$sample == i])
}

plot_ly(n_highly_expressed, x = ~sample, y = ~as.numeric(n_bc), type = 'bar',
             marker = list(color = '#D6D5C9',
                           line = list(color = 'rgb(8,48,107)', width = 1.5))) %>% 
  layout(title = "Highly expressed barcodes from random motifs",
         yaxis = list(title = "Number of barcodes with > 300 rpm"),
         xaxis = list(title = "sample"))

```

---

## Read distribution per cutoff

```{r cutoff_read_distribution, fig.width=10, fig.height=7, fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
## 2: How many barcodes can I find back at which cutoff? + What is the percentage of barcode reads that match the design at which cutoff?
## Identify the unmapped fraction
bc_fraction <- data.frame("sample" = unique(bc_df$sample),
                          "bcs_found" = "", stringsAsFactors=FALSE)
rpm_cutoff <- data.frame("cutoff" = seq(0.0001,10,1), stringsAsFactors=FALSE)
bc_fraction <- merge(bc_fraction, rpm_cutoff)

bc_df_filt <- bc_df[!is.na(bc_df$tf),]

for (i in unique(bc_fraction$cutoff)) {
  for (j in unique(bc_df_filt$sample)) {
    bc_fraction$bcs_found[bc_fraction$cutoff == i & bc_fraction$sample == j] <- nrow(bc_df_filt[bc_df_filt$rpm >= i & bc_df_filt$sample == j & bc_df_filt$tf == "Trp53",])/
      nrow(bc_annotation[bc_annotation$tf == "Trp53",]) *100
  }
}



## How many reads match to designed barcodes?
bc_reads <- data.frame("sample" = unique(bc_df$sample),
                          "bc_reads" = "", stringsAsFactors=FALSE)
bc_reads <- merge(bc_reads, rpm_cutoff)

for (i in unique(bc_reads$cutoff)) {
  for (j in unique(bc_df_filt$sample)) {
    bc_reads$bc_reads[bc_reads$cutoff == i & bc_reads$sample == j] <- sum(bc_df_filt$rpm[bc_df_filt$rpm >= i & bc_df_filt$sample == j])/
      sum(bc_df$rpm[bc_df$rpm >= i & bc_df$sample == j]) *100
  }
}

bc_fraction <- merge(bc_fraction, bc_reads)
bc_fraction$bcs_found <- as.numeric(bc_fraction$bcs_found)
bc_fraction$bc_reads <- as.numeric(bc_fraction$bc_reads)
bc_fraction$gcf <- gsub(".*(gcf.*)", "\\1", bc_fraction$sample)

#c("#1B998B", "#2D3047", "#FF9B71", "#ECDD7B")
# Plot to evaluate data quality per cutoff
for (i in unique(bc_fraction$gcf)) {
  p <- ggplot(bc_fraction[bc_fraction$gcf == i,]) +
  geom_point(aes(x = cutoff, y = bcs_found), color = '#1B998B', size = 1) +
  geom_line(aes(x = cutoff, y = bcs_found), color = '#1B998B', size = 1) +
  geom_point(aes(x = cutoff, y = bc_reads), color = 'black', size = 1) +
  geom_line(aes(x = cutoff, y = bc_reads), color = 'black', size = 1) +
  theme_bw()+
  xlab("rpm cutoff")+
  ylab("total barcodes (green) and matched barcode reads (black) detected (%)")+
  facet_wrap(~sample)+
  theme(strip.background =element_rect(fill="#D6D5C9"))
  print(p)
}
```

---

## pDNA-cDNA correlation

```{r pDNA_cDNA_correlation, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
## 3: What is the correlation of the 24 cDNA bc counts with the pDNA bc counts? 
pDNA <- data.frame("pDNA_lib1" = bc_df$rpm[bc_df$sample == "pMT06_pDNA_gcf6301"],
                   "barcode"= bc_df$barcode[bc_df$sample == "pMT06_pDNA_gcf6301"], stringsAsFactors=FALSE)
bc_df_2 <- merge(pDNA, bc_df, all = T)

pDNA <- data.frame("pDNA_lib2" = bc_df$rpm[bc_df$sample == "pDNA-GR_gcf6412"],
                   "barcode"= bc_df$barcode[bc_df$sample == "pDNA-GR_gcf6412"], stringsAsFactors=FALSE)
bc_df_2 <- merge(pDNA, bc_df_2, all = T)

pDNA <- data.frame("pDNA_lib3" = bc_df$rpm[bc_df$sample == "MyTaq_pDNA_r2_gcf6501"],
                   "barcode"= bc_df$barcode[bc_df$sample == "MyTaq_pDNA_r2_gcf6501"], stringsAsFactors=FALSE)
bc_df_2 <- merge(pDNA, bc_df_2, all = T)

ggplot(bc_df_2[bc_df_2$gcf == "gcf6210" | bc_df_2$gcf == "gcf6301",], aes(x = pDNA_lib1, y = rpm)) +
  geom_bin2d(bins = 100)+
  xlim(0,1000) +
  ylim(0,1000)+
  theme_bw()+
  facet_wrap(~sample)

ggplot(bc_df_2 %>%
         filter(str_detect(sample, "gcf6412")), aes(x = pDNA_lib2, y = rpm)) +
  geom_bin2d(bins = 100)+
  xlim(0,1000) +
  ylim(0,1000)+
  theme_bw()+
  facet_wrap(~sample)

ggplot(bc_df_2 %>%
         filter(str_detect(sample, "gcf6501")), aes(x = pDNA_lib3, y = rpm)) +
  geom_bin2d(bins = 100)+
  xlim(0,1000) +
  ylim(0,5000)+
  theme_bw()+
  facet_wrap(~sample)


# How do the highly represented pDNA_barcodes deviate from the barcode reads in the cDNA data? 
## I do this because the correlation of lowly representdied barcodes is very noisy
pDNA_bias <- data.frame("sample" = unique(bc_df_2$sample[grep("gcf6210|gcf6301", bc_df_2$sample)]),
                        "cor" = "", stringsAsFactors = F)
bc_df_2 <- na.omit(bc_df_2)
for (i in unique(pDNA_bias$sample)) {
  pDNA_bias$cor[pDNA_bias$sample == i] <- cor(bc_df_2$rpm[bc_df_2$pDNA_lib1 >= 250 & bc_df_2$sample == i], 
                                                 bc_df_2$pDNA_lib1[bc_df_2$pDNA_lib1 >= 250 & bc_df_2$sample == i])
}

plot_ly(pDNA_bias, x = ~sample, y = ~as.numeric(cor), type = 'bar',
             marker = list(color = '#D6D5C9',
                           line = list(color = 'rgb(8,48,107)', width = 1.5))) %>% 
  layout(title = "pDNA-cDNA correlation of highly represented barcodes",
         yaxis = list(title = "Correlation of pDNA vs. cDNA read counts (only barcodes with >250 pDNA counts)"),
         xaxis = list(title = "sample"))

pDNA_bias_2 <- data.frame("sample" = unique(bc_df_2$sample[grep("gcf6412|gcf6501", bc_df_2$sample)]),
                        "cor" = "", stringsAsFactors = F)
for (i in unique(pDNA_bias_2$sample)) {
  pDNA_bias_2$cor[pDNA_bias_2$sample == i] <- cor(bc_df_2$rpm[bc_df_2$pDNA_lib2 >= 250 & bc_df_2$sample == i], 
                                                 bc_df_2$pDNA_lib2[bc_df_2$pDNA_lib2 >= 250 & bc_df_2$sample == i])
}

plot_ly(pDNA_bias_2, x = ~sample, y = ~as.numeric(cor), type = 'bar',
             marker = list(color = '#D6D5C9',
                           line = list(color = 'rgb(8,48,107)', width = 1.5))) %>% 
  layout(title = "pDNA-cDNA correlation of highly represented barcodes",
         yaxis = list(title = "Correlation of pDNA vs. cDNA read counts (only barcodes with >250 pDNA counts)"),
         xaxis = list(title = "sample"))

pDNA_bias <- rbind(pDNA_bias, pDNA_bias_2)

pDNA_bias <- pDNA_bias %>%
  filter(as.numeric(cor) <= 0.7 | str_detect(sample, "pDNA")) 

bc_df <- bc_df %>%
  filter(sample %in% pDNA_bias$sample)

# Old vs new pDNA data
ggplot(bc_df_2 %>% dplyr::select(barcode, pDNA_lib1, pDNA_lib2) %>% unique(), aes(x = pDNA_lib1, y = pDNA_lib2)) +
    geom_pointdensity(show.legend = F, adjust = 100)+
    theme_bw()+
    theme(text = element_text(size=14)) +
    xlab("old pDNA data (e.cloni)")+
    ylab("new pDNA data (clonecatcher)")+
    ggtitle("Old vs. new pDNA data")+
    scale_color_viridis()

ggplot(bc_df_2 %>% dplyr::select(barcode, pDNA_lib3, pDNA_lib2) %>% unique(), aes(x = pDNA_lib3, y = pDNA_lib2)) +
    geom_pointdensity(show.legend = F, adjust = 100)+
    theme_bw()+
    theme(text = element_text(size=14)) +
    xlab("new pDNA data rep2")+
    ylab("new pDNA data rep1")+
    ggtitle("rep1 vs. rep2 new pDNA data")+
    scale_color_viridis()
```

---

## Replicate correlation

```{r replicate_correlation_rpm, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
## 4: Correlation plots of the replicates
## Combine replicates of normalized data in 3 different columns
bc_df_rep <- bc_df[!is.na(bc_df$tf),] %>% 
  dplyr::select(rep, rpm, barcode, sample) %>% 
  mutate(sample = gsub("r[1-3]_", "", sample)) %>%
  filter(str_detect(rep, "r[1-3]")) %>%
  unique() %>%
  spread(rep, rpm)

# Correlation matrix plot
n <- sample(1:nrow(bc_df_rep), 5000)
boundaries <- seq(from = 0.8, by = 0.05, length.out = 4)
# bc_df_rep <- bc_df_rep %>% filter(str_detect(condition, "pDNA", negate = T))
# not_all_na <- function(x) any(!is.na(x))
for (i in unique(bc_df_rep$sample)){
  plt <- ggpairs(bc_df_rep[bc_df_rep$sample == i,]  %>%
                   dplyr::select(r1, r2, r3),
               upper = list(continuous = corColor),
               lower = list(continuous = function(data, mapping, ...) {
                   ggally_points(data = data[n, ], mapping = mapping, alpha = 0.1, size = 0.5) +
                   geom_abline(slope = 1, lty = "dashed", col = "red") +
                   theme_bw()}),
               diag = list(continuous = function(data, mapping, ...) {
                   ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red") +
                   theme_bw()})) +
  ggtitle(paste("Correlation Between Replicates, Condition:", i)) +
  theme(text = element_text(size = 10)) +
  xlab("rpm") +
  ylab("rpm")
  #theme_light()
  
  print(plt)
}
```

**Stuff learned from above figures:**  
Some samples can be excluded from further analysis because they don't contain useful information, these samples are:  
MCF7-KO-DMSO: rep2_seq1, rep1_seq1(?), r2_seq2(?), r1_seq3, r3_seq3  
MCF7-KO-Nutlin: rep2_seq1, rep3_seq1(?), r1_seq2, r1_seq3, r3_seq3  
MCF7-WT-DMSO: rep2_seq1, rep3_seq1  
MCF7-WT-Nutlin: rep3_seq1, r2_seq2, r1_seq3  
A549_DMSO: r2_seq3, r3_seq3  
A549_Dex10: r2_seq3  
A549_Dex100: r2_seq3  
A549-Dex-1: r1_seq2, r2_seq3  
mES-N2B27-HQ: rep1_seq1  
mES-N2B27-RA: rep1_seq1, rep2_seq1  

---

### Annotation of the reporters

```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# Annotate the mutated motif of each TF
bc_df$neg_ctrls <- "No"
bc_df$neg_ctrls[grep("random", bc_df$tf)] <- "Yes"

# Annotate random promoter control
bc_df$rand_promoter <- "No"
bc_df$rand_promoter[grep("random", bc_df$promoter)] <- "Yes"

## Mark O'Connell controls
bc_df$positive_ctrl <- "No"
bc_df$positive_ctrl[bc_df$position == 60 | bc_df$position == 70] <- "Yes"


# Annotate affinity ids
## Mixed pool
bc_df$motif_id <- paste(bc_df$affinity_pos1, bc_df$affinity_pos2, bc_df$affinity_pos3, bc_df$affinity_pos4, sep = "_")
bc_df$affinity_id <- "other"

high_med <- read.csv("/DATA/usr/m.trauernicht/projects/P53_reporter_scan/data/library_design/output/high_sequences.csv") %>%
  mutate(motif_id = paste(affinity_pos1, affinity_pos2, affinity_pos3, affinity_pos4, sep = "_"))
bc_df$affinity_id[bc_df$motif_id %in% high_med$motif_id] <- "2_high_med"

med_low <- read.csv("/DATA/usr/m.trauernicht/projects/P53_reporter_scan/data/library_design/output/med_sequences.csv") %>%
  mutate(motif_id = paste(affinity_pos1, affinity_pos2, affinity_pos3, affinity_pos4, sep = "_"))
bc_df$affinity_id[bc_df$motif_id %in% med_low$motif_id] <- "4_med_low"

low_verylow <- read.csv("/DATA/usr/m.trauernicht/projects/P53_reporter_scan/data/library_design/output/low_sequences.csv") %>%
  mutate(motif_id = paste(affinity_pos1, affinity_pos2, affinity_pos3, affinity_pos4, sep = "_"))
bc_df$affinity_id[bc_df$motif_id %in% low_verylow$motif_id] <- "6_low_very-low"

verylow_zero <- read.csv("/DATA/usr/m.trauernicht/projects/P53_reporter_scan/data/library_design/output/verylow_sequences.csv") %>%
  mutate(motif_id = paste(affinity_pos1, affinity_pos2, affinity_pos3, affinity_pos4, sep = "_"))
bc_df$affinity_id[bc_df$motif_id %in% verylow_zero$motif_id] <- "8_very-low_null"

## 4x identical copy reporters
bc_df$affinity_id[bc_df$affinity_pos1 == 0 & bc_df$affinity_pos2 == 0 & bc_df$affinity_pos3 == 0 & bc_df$affinity_pos4 == 0] <- "1_high_only"
bc_df$affinity_id[bc_df$affinity_pos1 == 1 & bc_df$affinity_pos2 == 1 & bc_df$affinity_pos3 == 1 & bc_df$affinity_pos4 == 1] <- "3_med_only"
bc_df$affinity_id[bc_df$affinity_pos1 == 2 & bc_df$affinity_pos2 == 2 & bc_df$affinity_pos3 == 2 & bc_df$affinity_pos4 == 2] <- "5_low_only"
bc_df$affinity_id[bc_df$affinity_pos1 == 3 & bc_df$affinity_pos2 == 3 & bc_df$affinity_pos3 == 3 & bc_df$affinity_pos4 == 3] <- "7_very-low_only"
bc_df$affinity_id[bc_df$affinity_pos1 == 4 & bc_df$affinity_pos2 == 4 & bc_df$affinity_pos3 == 4 & bc_df$affinity_pos4 == 4] <- "9_null_only"

## Other mixed pools
bc_df$affinity_id[bc_df$affinity_pos1 == 0 & bc_df$affinity_pos2 == 0 & bc_df$affinity_pos3 != 0 & bc_df$affinity_pos3 != 1 & 
                      bc_df$affinity_pos4 != 0 & bc_df$affinity_pos4 != 1] <- "10_high_start"

bc_df$affinity_id[bc_df$affinity_pos1 == 0 & bc_df$affinity_pos4 == 0 & bc_df$affinity_pos2 != 0 & bc_df$affinity_pos2 != 1 & 
                      bc_df$affinity_pos3 != 0 & bc_df$affinity_pos3 != 1] <- "11_high_mid"
bc_df$affinity_id[bc_df$affinity_pos1 != 0 & bc_df$affinity_pos1 != 1 & bc_df$affinity_pos2 != 0 & bc_df$affinity_pos2 != 1 & 
                      bc_df$affinity_pos3 == 0 & bc_df$affinity_pos4 == 0] <- "12_high_end"



### Add affinity values for each reporter
reporter_affinity_p53 <- read_tsv("/DATA/usr/m.trauernicht/projects/P53_reporter_scan/data/parameter_files/p53scores.tsv") %>%
  dplyr::select('reporter_id_2' = Name, SumAffinity) 

reporter_affinity_p53_complete <-
  read_tsv("/DATA/usr/m.trauernicht/projects/P53_reporter_scan/data/parameter_files/p53scores_complete.tsv") %>%
  dplyr::select('reporter_id' = Name, 'reporter_affinity' = SumAffinity)

bc_df <- bc_df %>%
  mutate(reporter_id_2 = gsub("(^.*_).*?_(p_.*)","\\1\\2", reporter_id),
         reporter_id_2 = gsub("_bc_[0-9]{1}$", "", reporter_id_2))

bc_df <- merge(bc_df, reporter_affinity_p53, all = T, by = "reporter_id_2")
bc_df <- merge(bc_df, reporter_affinity_p53_complete, all = T, by = "reporter_id")


# Add affinity parameters
## Activity vs. cumulative predicted affinity
affinity_df <- read.csv2("/DATA/usr/m.trauernicht/projects/P53_reporter_scan/data/library_design/mt20201123_affinity.csv", header = T, stringsAsFactors = F)
affinity_df$affinity[affinity_df$id == 4] <- 0
affinity_df$TF[affinity_df$TF == "Trp53"] <- "P53"
affinity_df$TF[affinity_df$TF == "Gr"] <- "GR"

bc_df$tf <- gsub("Trp53", "P53", bc_df$tf)
bc_df$reporter_id <- gsub("Trp53", "P53", bc_df$reporter_id)
bc_df$tf <- gsub("Gr", "GR", bc_df$tf)
bc_df$reporter_id <- gsub("Gr", "GR", bc_df$reporter_id)

### Complicated way of adding cum_affinity information to the df
for (i in unique(bc_df$affinity_pos1)) {
  for (j in unique(bc_df$affinity_pos2)) {
    for (k in unique(bc_df$affinity_pos3)) {
      for (l in unique(bc_df$affinity_pos4)) {
        for (m in unique(affinity_df$TF)) {
    bc_df$cum_affinity[bc_df$affinity_pos1 == i & bc_df$affinity_pos2 == j & bc_df$affinity_pos3 == k &
                         bc_df$affinity_pos4 == l & bc_df$tf == m] <- 
      affinity_df$affinity[affinity_df$id == i & affinity_df$TF == m] +
      affinity_df$affinity[affinity_df$id == j & affinity_df$TF == m] +
      affinity_df$affinity[affinity_df$id == k & affinity_df$TF == m] +
      affinity_df$affinity[affinity_df$id == l & affinity_df$TF == m]
        }
      }
    }
  }
}

### Also add the ddG and max_aff information
bc_df <- bc_df %>%
  group_by(reporter_id) %>%
  mutate(ddG = affinity_pos1 + affinity_pos2 + affinity_pos3 + affinity_pos4,
         max_aff = max(affinity_pos1,affinity_pos2,affinity_pos3,affinity_pos4),
         cum_affinity = cum_affinity/4 *100,
         cum_affinity = round(cum_affinity, 2))

# Add number of non-zero affinity binding sites
bc_df <- bc_df[!is.na(bc_df$sample),]
bc_df$n_sites <- 4
bc_df$n_sites[bc_df$affinity_pos1 == 4] <- bc_df$n_sites[bc_df$affinity_pos1 == 4] - 1
bc_df$n_sites[bc_df$affinity_pos2 == 4] <- bc_df$n_sites[bc_df$affinity_pos2 == 4] - 1
bc_df$n_sites[bc_df$affinity_pos3 == 4] <- bc_df$n_sites[bc_df$affinity_pos3 == 4] - 1
bc_df$n_sites[bc_df$affinity_pos4 == 4] <- bc_df$n_sites[bc_df$affinity_pos4 == 4] - 1
```


---


### Scaling the data
```{r}
# # It might be hard to compare the KO with the WT barcode counts because in th KO we don't really have active elements -> 'inactive' elements will take up most of the reads
# ## How can we scale the different conditions correctly?
# ggplot(bc_df %>% 
#          filter(neg_ctrls == "Yes", str_detect(tf, "53"), str_detect(sample, "MCF")) %>% 
#          dplyr::select(tf, sample, rpm, reporter_id, gcf) %>% 
#          unique(), 
#        aes(x = sample, y = rpm, color = tf)) + 
#   geom_quasirandom(dodge.width = 0.75) + 
#   scale_color_brewer(palette = "Dark2") + 
#   theme_bw() + 
#   theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
#   facet_wrap(~gcf)
# 
# 
# bc_df_scale <- bc_df %>%
#   filter(neg_ctrls == "Yes", str_detect(tf, "53")) %>% 
#   dplyr::select(sample, rpm, reporter_id, gcf) %>%
#   unique() %>% 
#   group_by(sample) %>%
#   mutate(rpm = mean(rpm)) %>%
#   ungroup() %>%
#   dplyr::select(-reporter_id) %>%
#   unique() %>%
#   dplyr::select("background_rpm" = rpm, sample) %>%
#   unique() %>%
#   filter(str_detect(sample, "pDNA", negate = T))
# 
# bc_df <- merge(bc_df, bc_df_scale, all = T)
# 
# bc_df <- bc_df %>%
#   mutate(rpm_norm = rpm / background_rpm)
# 
# ggplot(bc_df %>% 
#          filter(neg_ctrls == "Yes", str_detect(tf, "53"), str_detect(sample, "MCF")) %>% 
#          dplyr::select(tf, sample, rpm_norm, reporter_id, gcf) %>% 
#          unique(), 
#        aes(x = sample, y = rpm_norm, color = tf)) + 
#   geom_quasirandom(dodge.width = 0.75) + 
#   scale_color_brewer(palette = "Dark2") + 
#   theme_bw() + 
#   theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
#   facet_wrap(~gcf)
# # looks quite ok, MCF7_WT_DMSO_r1_gcf6412 has some outliers - maybe I need to remove this sample
```





---

### Normalization of barcode counts:
Divide cDNA barcode counts through pDNA barcode counts
```{r normalization, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Normalize data using pDNA data

## Remove all non-matching reads now
bc_df <- bc_df[!is.na(bc_df$tf),]

## Add pDNA data as separate column


### First correlate all pDNA data to ensure that all samples have a good quality
pDNA_lib1 <- bc_df[grep("pDNA", bc_df$sample),] %>% 
  dplyr::select(barcode, sample, rpm) %>% 
  unique() %>%
  dcast(barcode ~ sample)

ggpairs(pDNA_lib1 %>%
                   dplyr::select(-barcode),
               upper = list(continuous = corColor),
               lower = list(continuous = function(data, mapping, ...) {
                   ggally_points(data = data[n, ], mapping = mapping, alpha = 0.1, size = 0.5) +
                   geom_abline(slope = 1, lty = "dashed", col = "red") +
                   theme_bw()}),
               diag = list(continuous = function(data, mapping, ...) {
                   ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red") +
                   theme_bw()})) +
  ggtitle("Correlation Between Replicates, Condition") +
  theme(text = element_text(size = 10)) +
  xlab("rpm") +
  ylab("rpm")


pDNA_lib1_sel <- pDNA_lib1 %>%
  mutate(pDNA_counts_rpm_lib1 = 
           (pMT06_pDNA_r1_gcf6210 + pMT06_pDNA_r2_gcf6210 + `pMT06-v2_pDNA_1_gcf6301` + `pMT06-v2_pDNA_2_gcf6301` + pMT06_pDNA_gcf6301) / 5) %>%
  dplyr::select(barcode, pDNA_counts_rpm_lib1)


bc_df <- merge(pDNA_lib1_sel, bc_df, by = "barcode")

pDNA_gcf6412 <- pDNA_lib1 %>%
  dplyr::select(barcode, 'pDNA_counts_rpm_gcf6412' = `pDNA-GR_gcf6412`) 

bc_df <- merge(pDNA_gcf6412, bc_df, by = "barcode")

pDNA_gcf6501 <- pDNA_lib1 %>%
  dplyr::select(barcode, 'pDNA_counts_rpm_gcf6501' = MyTaq_pDNA_r2_gcf6501) 

bc_df <- merge(pDNA_gcf6501, bc_df, by = "barcode")

## Compute activity by dividing cDNA bc counts through pDNA bc counts (since I have two totally different pDNA libraries, I need to do it for the two libraries separately)
bc_df$activity <- 0

bc_df$activity[bc_df$gcf == "gcf6210" | bc_df$gcf == "gcf6301"] <- bc_df$rpm[bc_df$gcf == "gcf6210" | bc_df$gcf == "gcf6301"] / 
  bc_df$pDNA_counts_rpm_lib1[bc_df$gcf == "gcf6210" | bc_df$gcf == "gcf6301"]

bc_df$activity[bc_df$gcf == "gcf6412"] <- bc_df$rpm[bc_df$gcf == "gcf6412"] / 
  bc_df$pDNA_counts_rpm_gcf6412[bc_df$gcf == "gcf6412"]

bc_df$activity[bc_df$gcf == "gcf6501"] <- bc_df$rpm[bc_df$gcf == "gcf6501"] / 
  bc_df$pDNA_counts_rpm_gcf6501[bc_df$gcf == "gcf6501"]
```

---

## Characterize reporter activities

```{r tf_activity, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# Histogram - show only high activities per condition
ggplot(bc_df[bc_df$activity >= 5,], aes(x = activity)) +
    geom_histogram(binwidth = 1) +
    theme_bw() +
    xlim(5,15)+
    facet_wrap(~sample)+
    theme(strip.background =element_rect(fill="#D6D5C9"))

# Barplot - show how many active reporters there are per condition
bc_df_2 <- bc_df %>%
  filter(activity >= 4) %>%
  group_by(sample) %>%
  mutate(active_reporters = length(unique(barcode)))

plot_ly(bc_df_2 %>% 
          dplyr::select(sample, active_reporters) %>% 
          unique(), 
        x = ~sample, y = ~active_reporters, type = 'bar',
             marker = list(color = '#D6D5C9',
                           line = list(color = 'rgb(8,48,107)', width = 1.5))) %>% 
  layout(title = "Number of highly active barcodes",
         yaxis = list(title = "Barcodes with cDNA/pDNA > 4"),
         xaxis = list(title = "Condition"))

# Barplot counting high activity barcodes from random motifs
bc_df_2 <- bc_df %>%
  filter(activity >= 4,
         str_detect(tf, "random")) %>%
  group_by(sample) %>%
  mutate(active_reporters = length(unique(barcode))) %>%
  dplyr::select(sample, active_reporters) %>%
  unique()

plot_ly(bc_df_2, x = ~sample, y = ~as.numeric(active_reporters), type = 'bar',
             marker = list(color = '#D6D5C9',
                           line = list(color = 'rgb(8,48,107)', width = 1.5))) %>% 
  layout(title = "Highly active barcodes from random motifs",
         yaxis = list(title = "Number of barcodes with activity > 4"),
         xaxis = list(title = "sample"))

#bc_df <- bc_df[-grep("MCF7_KO_DMSO_r1_gcf6412", bc_df$sample),]

# Correlation activities vs. pDNA data
ggplot(bc_df[bc_df$gcf == "gcf6210" | bc_df$gcf == "gcf6301",], aes(x = pDNA_counts_rpm_lib1, y = activity)) +
  geom_bin2d(bins = 100)+
  ylim(0,10)+
  theme_bw()+
  facet_wrap(~sample)

ggplot(bc_df[bc_df$gcf == "gcf6412",], aes(x = pDNA_counts_rpm_gcf6412, y = activity)) +
  geom_bin2d(bins = 100)+
  ylim(0,50)+
  theme_bw()+
  facet_wrap(~sample)

ggplot(bc_df[bc_df$gcf == "gcf6501",], aes(x = pDNA_counts_rpm_gcf6501, y = activity)) +
  geom_bin2d(bins = 100)+
  ylim(0,50)+
  theme_bw()+
  facet_wrap(~sample)

# Plot number of barcodes with activity > 3 and high pDNA count
bc_df_2 <- bc_df %>%
  filter(activity >= 3) %>%
  filter(pDNA_counts_rpm_gcf6501 >= 200) %>%
  group_by(sample) %>%
  mutate(active_reporters = length(unique(barcode)))

plot_ly(bc_df_2 %>% 
          dplyr::select(sample, active_reporters) %>% 
          unique(), 
        x = ~sample, y = ~active_reporters, type = 'bar',
             marker = list(color = '#D6D5C9',
                           line = list(color = 'rgb(8,48,107)', width = 1.5))) %>% 
  layout(title = "Number of highly active barcodes",
         yaxis = list(title = "Barcodes with cDNA/pDNA > 4"),
         xaxis = list(title = "sample"))

```

---

## Reporter activity correlations

```{r compare_activities, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
## 4: Correlation plots of the replicates
## Combine replicates of normalized data in 3 different columns
bc_df_rep <- bc_df[!is.na(bc_df$tf),] %>% 
  dplyr::select(rep, activity, barcode, sample) %>% 
  mutate(sample = gsub("r[1-3]_", "", sample)) %>%
  filter(str_detect(rep, "r[1-3]")) %>%
  unique() %>%
  spread(rep, activity)

# Correlation matrix plot
n <- sample(1:nrow(bc_df_rep), 5000)
boundaries <- seq(from = 0.8, by = 0.05, length.out = 4)
for (i in unique(bc_df_rep$sample)){
  plt <- ggpairs(bc_df_rep[bc_df_rep$sample == i,]  %>%
                   dplyr::select(r1, r2, r3),
                 upper = list(continuous = corColor),
               lower = list(continuous = function(data, mapping, ...) {
                   ggally_points(data = data[n, ], mapping = mapping, alpha = 0.1, size = 0.5) +
                   geom_abline(slope = 1, lty = "dashed", col = "red") +
                   theme_bw()}),
               diag = list(continuous = function(data, mapping, ...) {
                   ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red") +
                   theme_bw()})) +
  ggtitle(paste("Correlation Between Replicates, Condition:", i)) +
  theme(text = element_text(size = 10)) +
  xlab("Reporter activity") +
  ylab("Reporter activity") 
  #theme_light()
  
  print(plt)
}
```

---

## Removing outliers

```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# First identify and remove outlier barcodes - this removes the noise created by faulty barcode clustering etc. 
## Calculate mean and SD for each reporter
bc_df_2 <- bc_df[bc_df$gcf == "gcf6210" | bc_df$gcf == "gcf6301",] %>%
  filter(pDNA_counts_rpm_lib1 > 5)
bc_df_3 <- bc_df[bc_df$gcf == "gcf6412",] %>%
  filter(pDNA_counts_rpm_gcf6412 > 5)
bc_df_4 <- bc_df[bc_df$gcf == "gcf6501",] %>%
  filter(pDNA_counts_rpm_gcf6501 > 5)
bc_df_2 <- rbind(bc_df_2, bc_df_3, bc_df_4)

bc_df_cDNA <- bc_df_2[-grep("pDNA", bc_df_2$sample),]
bc_df_cDNA$reporter_id <- gsub(".{5}$", "", bc_df_cDNA$reporter_id)
bc_df_cDNA$reporter_activity <- ave(bc_df_cDNA$activity, bc_df_cDNA$reporter_id, 
                                bc_df_cDNA$sample, FUN =
                                  function(x) mean(x))
bc_df_cDNA$reporter_activity_sd <- ave(bc_df_cDNA$activity, bc_df_cDNA$reporter_id, 
                              bc_df_cDNA$sample,  FUN =
                                  function(x) sd(x))

# ## Remove data points that are 2xSD away from the mean
# bc_df_cDNA$upper_activity <- bc_df_cDNA$reporter_activity + (1.75 * bc_df_cDNA$reporter_activity_sd)
# bc_df_cDNA$lower_activity <- bc_df_cDNA$reporter_activity - (1.75 * bc_df_cDNA$reporter_activity_sd)
# bc_df_cDNA_filt <- bc_df_cDNA %>%
#   mutate(outlier = "No")
# bc_df_cDNA_filt$outlier[bc_df_cDNA_filt$activity >= bc_df_cDNA_filt$upper_activity | 
#                                 bc_df_cDNA_filt$activity <= bc_df_cDNA_filt$lower_activity] <- "Yes"

## Remove reporters with a higher SD than the mean activity
bc_df_cDNA_filt <- bc_df_cDNA %>%
  mutate(outlier = "No")
bc_df_cDNA_filt$outlier[bc_df_cDNA_filt$reporter_activity_sd >= (bc_df_cDNA_filt$reporter_activity)] <- "Yes"

## Plot the fraction of outliers per sample
n_outlier <- bc_df_cDNA_filt %>%
  filter(reporter_activity > 0.5) %>%
  dplyr::select(outlier, sample, reporter_id) %>%
  filter(outlier == "Yes") %>%
  unique() %>% 
  group_by(sample) %>%
  mutate(n_outlier = length(outlier)) %>%
  dplyr::select(-outlier, -reporter_id) %>%
  unique()

bc_df_cDNA_filt <- merge(bc_df_cDNA_filt, n_outlier, all = T)
bc_df_cDNA_filt$n_outlier[is.na(bc_df_cDNA_filt$n_outlier)] <- 0

plot_ly(bc_df_cDNA_filt %>%
          dplyr::select(n_outlier, sample) %>%
          unique(),
        x = ~sample, y = ~n_outlier, type = 'bar',
             marker = list(color = '#D6D5C9',
                           line = list(color = 'rgb(8,48,107)', width = 1.5))) %>% 
  layout(title = "Reporters with SD > mean activity per condition",
         yaxis = list(title = "Number of outliers"),
         xaxis = list(title = "Condition"))
  
# ## Only keep conditions with low noise levels
# bc_df_cDNA_filt <- bc_df_cDNA_filt %>%
#   filter(n_outlier < 200)

# From previous script:
bc_df_cDNA_filt <- bc_df_cDNA_filt %>%
  filter(sample %in% c("A549_Dex1_r1_gcf6412", "A549_Dex1_r2_gcf6210", "A549_Dex1_r2_gcf6301", "A549_Dex1_r3_gcf6210", "A549_Dex10_r1_gcf6301",
                       "A549_Dex10_r1_gcf6412", "A549_Dex10_r2_gcf6210", "A549_Dex10_r3_gcf6210", "A549_Dex10_r3_gcf6412", "A549_Dex100_r1_gcf6210", 
                       "A549_Dex100_r1_gcf6301", "A549_Dex100_r2_gcf6210","A549_Dex100_r2_gcf6301","A549_Dex100_r3_gcf6210","A549_Dex100_r3_gcf6412",
                       "A549_DMSO_r1_gcf6210","A549_DMSO_r1_gcf6301", "A549_DMSO_r1_gcf6412", "A549_DMSO_r2_gcf6210", "A549_DMSO_r2_gcf6301", 
                       "A549_DMSO_r3_gcf6210", "MCF7_DMSO_r1_gcf6501","MCF7_DMSO_r2_gcf6501", "MCF7_KO_DMSO_r1_gcf6501", "MCF7_KO_DMSO_r1_gcf6412", "MCF7_KO_DMSO_r2_gcf6301", 
                       "MCF7_KO_DMSO_r2_gcf6501", "MCF7_KO_Nutlin_r1_gcf6501", "MCF7_KO_Nutlin_r2_gcf6301", "MCF7_KO_Nutlin_r2_gcf6501",
                       "MCF7_KO_PFTa_r1_gcf6501", "MCF7_KO_PFTa_r2_gcf6501", "MCF7_Nutlin_r1_gcf6501", "MCF7_Nutlin_r2_gcf6501", "MCF7_PFTa_r1_gcf6501", 
                       "MCF7_PFTa_r2_gcf6501", "MCF7_WT_DMSO_r1_gcf6210", "MCF7_WT_DMSO_r1_gcf6301", "MCF7_WT_DMSO_r1_gcf6412", "MCF7_WT_DMSO_r2_gcf6301",
                       "MCF7_WT_DMSO_r3_gcf6412", "MCF7_WT_Nutlin_r1_gcf6210", "MCF7_WT_Nutlin_r2_gcf6210", "MCF7_WT_Nutlin_r2_gcf6301", "MCF7_WT_Nutlin_r3_gcf6412",
                       "mES_DMSO_r2_gcf6210", "mES_DMSO_r3_gcf6210", "mES_N2B27-HQ_r2_gcf6210", "mES_N2B27-HQ_r3_gcf6210", "mES_N2B27-RA_r3_gcf6210"))

## Recalculate mean and sd
bc_df_cDNA_filt$reporter_activity <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id, 
                                bc_df_cDNA_filt$sample, FUN =
                                  function(x) mean(x))
bc_df_cDNA_filt$reporter_activity_sd <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id, 
                              bc_df_cDNA_filt$sample,  FUN =
                                  function(x) sd(x))
```

---

## Technical replicate correlations

```{r technical_replicate_correlations, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
## Combine replicates in 5 different columns
bc_df_rep <- bc_df_cDNA_filt %>% 
  filter(rand_promoter == "No", outlier == "No") %>%
  dplyr::select(oligo.barcode, activity, tf, sample, reporter_id, pDNA_counts_rpm_lib1) %>%
  mutate(pDNA_counts_rpm_lib1 = ave(pDNA_counts_rpm_lib1, reporter_id, FUN = function(x) min(x))) %>% 
  unique() %>%
  spread(oligo.barcode, activity)

## Compute the correlation between two technical replicates for each pDNA cutoff, separately for each TF
cor_df <- data.frame("rpm_cutoff"  = seq(0,100,1), "cor" = "", stringsAsFactors=FALSE)
cor_df <- merge(unique(bc_df_rep$tf), cor_df)

for (i in unique(cor_df$rpm_cutoff)) {
  for (j in unique(cor_df$x)) {
    x <- bc_df_rep[bc_df_rep$pDNA_counts_rpm > i & bc_df_rep$tf == j,]
    if (nrow(x) == 0) {
      cor_df$cor[cor_df$rpm_cutoff == i & cor_df$x == j] <- NA
    } else {
      cor_df$cor[cor_df$rpm_cutoff == i & cor_df$x == j] <- cor(x$`2`, x$`4`, use = "pairwise.complete.obs", method = "spearman")
    }
  }
}

ggplot(cor_df) +
  geom_point(aes(x = rpm_cutoff, y = as.numeric(cor))) +
  theme_bw() +
  ylim(0,1) +
  facet_wrap(~x)

## Compute the correlation between two technical replicates for each pDNA cutoff, for all reporters together
cor_df <- data.frame("rpm_cutoff"  = seq(0,100,1), "cor" = "", stringsAsFactors=FALSE)

for (i in unique(cor_df$rpm_cutoff)) {
    x <- bc_df_rep[bc_df_rep$pDNA_counts_rpm > i,]
    if (nrow(x) == 0) {
      cor_df$cor[cor_df$rpm_cutoff == i] <- NA
    } else {
      cor_df$cor[cor_df$rpm_cutoff == i] <- cor(x$`2`, x$`4`, "pairwise.complete.obs", method = "spearman")
    }
}

ggplot(cor_df) +
  geom_point(aes(x = rpm_cutoff, y = as.numeric(cor))) +
  theme_bw() +
  ylim(0,1) 


## Set a pDNA cutoff based on above plots
bc_df_cDNA_filt$noisy <- "No"
bc_df_cDNA_filt$noisy[bc_df_cDNA_filt$pDNA_counts_rpm <= 25] <- "Yes"
bc_df_rep <- bc_df_rep[bc_df_rep$pDNA_counts_rpm >= 25,]


# Correlation matrix plot
n <- sample(1:nrow(bc_df_rep), 5000)
boundaries <- seq(from = 0.8, by = 0.05, length.out = 4)
plt <- ggpairs(bc_df_rep %>% dplyr::select("1", "2", "3", "4", "5"),
               upper = list(continuous = corColor),
               lower = list(continuous = function(data, mapping, ...) {
                   ggally_points(data = data[n, ], mapping = mapping, alpha = 0.1, size = 0.5) +
                   geom_abline(slope = 1, lty = "dashed", col = "red") +
                   theme_bw()}),
               diag = list(continuous = function(data, mapping, ...) {
                   ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red") +
                   theme_bw()})) +
  ggtitle("Correlation Between Technial Replicates") +
  theme(text = element_text(size = 20)) +
  xlab("Reporter activity") +
  ylab("Reporter activity")

print(plt)
```

---

### Export data

```{r data export, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# Polish export dataframe
bc_df_cDNA_filt <- bc_df_cDNA_filt %>% 
  mutate(log_activity = log2(activity),
         log_reporter_activity = log2(reporter_activity))


# Export bc_df for cDNA analysis
filename <- SetFileName("_reporter_activity_filt", "mt")
setwd("/DATA/usr/m.trauernicht/projects/P53_reporter_scan/data/gcf6502/results/")
write.csv(bc_df_cDNA_filt, file = paste(filename,".csv", sep = ""), row.names = F)
```

---

### Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

