---
title: "cDNA reads processing - Deep P53/GR scan - stimulation 2"
author: "Max Trauernicht"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output: 
  prettydoc::html_pretty:
    theme: leonids
    highlight: github
  #   toc: true
  #   toc_float: true
  #   code_folding: show
  # editor_options:
  #   chunk_output_type: console
---

*knitr document van Steensel lab*

# TF reporter cDNA reads processing - Deep P53/GR scan - stimulation 2

# Introduction
I previously processed the raw sequencing data, quantified the pDNA data and normalized the cDNA data. In this script, I want to have a detailed look at the cDNA data from a general perspective.

```{r setup, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(RColorBrewer)
library(ggplot2)
library(dplyr)
library(maditr)
library(tibble)
library(pheatmap)
library(ggpubr)
library(ggbeeswarm)
library(ggforce)
library(viridis)
library(plyr)
library(cowplot)
library(gridExtra)
library(GGally)
library(readr)
library(stringr)
library(tidyr)
library(ROCR)
library(plotly)
library(randomForest)
library(glmnet)
library(glmnetUtils)
```


```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}


cbind.fill <- function(...){
    nm <- list(...) 
    nm <- lapply(nm, as.matrix)
    n <- max(sapply(nm, nrow)) 
    do.call(cbind, lapply(nm, function (x) 
        rbind(x, matrix(, n-nrow(x), ncol(x))))) 
}

# Function to substring the right part of the motif
substrRight <- function(x, n){
  substr(x, nchar(x)-n+1, nchar(x))
}


# Function to load PWM matrix
get_pwm_feature_matrix <- function(motif_meta_fn, fimo_fn, db = 2) {

  # validate args
  valid_dbs <- 1:2
  if(!db %in% valid_dbs)
    stop('Invalid db (database version). Please use db=1 (maintained for backward compatibility only) or db=2')

  # db=1 is maintained for backward compatibility only
  if(db == 1) {

    # read in motif metadata
    motif_meta    <- read.csv(motif_meta_fn)

    # check whether motif metadata contain essential annotations
    if(!all(c('PWM.ID', 'Cognate.TF') %in% colnames(motif_meta))) {
      message('The motif metadata file does not contain the essential columns PWM.ID and Cognate.TF')
    }

    motif_minimal <- motif_meta[, c('PWM.ID', 'Cognate.TF')]

    # load fimo output --> extract motif id, sequence id and p-value
    df <- read.table(fimo_fn)
    df <- df[, c(1, 2, 7)]

    colnames(df) <- c('PWM.ID', 'seqid', 'pval')

    # add TF id
    df <- merge(df, motif_minimal, by = 'PWM.ID')

    # group motif hits by sequence id
    l <- split(df, df[['seqid']])

    # multiple PWM and multiple hits possible. Reduce hits to one per TF, keeping best p-val only
    l <- lapply(l, function(x) {
      x_by_tf <- split(x, x[['Cognate.TF']], drop = TRUE)
      x_by_tf <- lapply(x_by_tf, function(y) y[which.min(y$pval), ])
      do.call('rbind', x_by_tf)
    })

    # initialize feature matrix
    n_tf          <- motif_minimal[['Cognate.TF']] %>%
      unique %>%
      length
    n_seq         <- length(l)
    pwm           <- matrix(1, nrow = n_seq, ncol = n_tf)
    colnames(pwm) <- (motif_minimal[['Cognate.TF']] %>% unique)

    # replace :: from names of composite motifs
    colnames(pwm) <- str_replace_all(colnames(pwm), '::', '_')

    # fill in feature matrix
    for(i in 1 : n_seq) {
      pwm[i, l[[i]][['Cognate.TF']]] <- l[[i]]$pval
    }

    # -log10 transform
    pwm           <- -1 * log10(pwm)

    # coerce to tib and return
    tib_fimo <- as_data_frame(pwm) %>%
      mutate(id = names(l))
      dplyr::select(id, everything())

  }

  # db = 2 (default)
  else {

    # load metadata
    tib_meta    <- read_csv(motif_meta_fn) %>%
      # extract tf symbol from motif id (Cognate_TF unsafe, it can be empty) and replace :: occurrences
      mutate(tf_symbol = str_remove(ID, '_[0-9]*'),
             tf_symbol = str_replace(tf_symbol, '::', '_')) %>%
      dplyr::select(motif_id = `PWM ID`, tf_symbol)

    # load fimo results
    tib_fimo <- read_tsv(fimo_fn) %>%
      # extract motif id, sequence id and p-value
      dplyr::select(motif_id, sequence_name, pval = `p-value`)

    # add tf symbol to fimo results
    tib_fimo <- tib_fimo %>%
      left_join(tib_meta, by = 'motif_id') %>%
      # remove hits with missing motif id (composite pwms)
      filter(!is.na(tf_symbol))

    # select best hit for each motif and sequence
    tib_fimo <- tib_fimo %>%
      group_by(sequence_name, tf_symbol) %>%
      dplyr::slice(which.min(pval)) %>%
      ungroup()

    # spread into feature matrix
    tib_fimo <- tib_fimo %>%
      mutate(pval = -1 * log10(pval)) %>%
      dplyr::select(-motif_id) %>%
      spread(key = tf_symbol, value = pval, fill = 0, drop = TRUE) %>%
      # perform cosmetics on the id
      mutate(id = sequence_name) %>%
      dplyr::select(-c(sequence_name)) %>%
      dplyr::select(id, everything())

  }

  return(tib_fimo)

}

# From Fede:
# ggpairs custom functions
corColor <- function(data, mapping, color = I("black"), sizeRange = c(1, 3), ...) {

  x   <- eval_data_col(data, mapping$x)
  y   <- eval_data_col(data, mapping$y)
  r   <- cor(x, y, "pairwise.complete.obs")
  rt  <- format(r, digits = 3)
  tt  <- as.character(rt)
  cex <- max(sizeRange)

  # helper function to calculate a useable size
  percent_of_range <- function(percent, range) {
    percent * diff(range) + min(range, na.rm = TRUE)
  }

  # plot correlation coefficient
  p <- ggally_text(label = tt, mapping = aes(), xP = 0.5, yP = 0.5,
                   size = I(percent_of_range(cex * abs(r), sizeRange)), color = color, ...) +
    theme(panel.grid.minor=element_blank(),
          panel.grid.major=element_blank())

  corColors <- RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")[2:6]

  if (r <= boundaries[1]) {
    corCol <- corColors[1]
  } else if (r <= boundaries[2]) {
    corCol <- corColors[2]
  } else if (r < boundaries[3]) {
    corCol <- corColors[3]
  } else if (r < boundaries[4]) {
    corCol <- corColors[4]
  } else {
    corCol <- corColors[5]
  }

  p <- p +
    theme(panel.background = element_rect(fill = corCol))

  return(p)
}

```


```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# Import processed bc counts from the preprocessing step
cDNA_df <- read.csv("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/gcf6210/results/mt20210118_reporter_activity_filt.csv", header = T) %>% setnames("condition", "sample")
cDNA_df <- cDNA_df[!is.na(cDNA_df$tf),]
cDNA_df$activity[is.na(cDNA_df$activity)] <- 0


# Add annotations
## Mark O'Connell controls
cDNA_df$positive_ctrl <- "No"
cDNA_df$positive_ctrl[cDNA_df$position == 60 | cDNA_df$position == 70] <- "Yes"
cDNA_df$tf <- as.character(cDNA_df$tf)


### I saw that the KO and the WT reporter activities are drastically different
### Hypothetically, the random reporters should have the same baseline in all conditions
# Therefore: I will now scale the reporter activities to the same level based on the random reporters
cDNA_df_neg <- cDNA_df %>%
  dplyr::select(sample, reporter_activity, reporter_id, neg_ctrls, rand_promoter, outlier) %>%
  filter(neg_ctrls == "Yes", outlier == "No") %>%
  unique() %>% 
  group_by(sample) %>%
  mutate(reporter_activity = mean(reporter_activity)) %>%
  ungroup() %>%
  dplyr::select(-reporter_id) %>%
  unique() %>%
  dplyr::select("background_activity" = reporter_activity, sample) 

cDNA_df <- merge(cDNA_df, cDNA_df_neg)
cDNA_df <- cDNA_df %>%
  mutate(reporter_activity = reporter_activity / background_activity)

# Compute means per sample
cDNA_df$condition = gsub("(.*)_r[1-3]{1}$", "\\1", cDNA_df$sample)
cDNA_df$condition = gsub("_rep[1-3]{1}$", "", cDNA_df$condition)
cDNA_df$reporter_activity_condition <- ave(cDNA_df$reporter_activity, cDNA_df$condition, cDNA_df$reporter_id, 
                                           FUN = function(x) mean(x))

# Check if the conditions are scaled correctly
ggplot(cDNA_df %>% 
         filter(neg_ctrls == "Yes", outlier == "No") %>% 
         dplyr::select(tf, condition, reporter_activity_condition, reporter_id) %>% 
         unique(), 
       aes(x = condition, y = reporter_activity_condition, color = tf)) + 
  geom_quasirandom(dodge.width = 0.75) + 
  scale_color_brewer(palette = "Dark2") + 
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))


# Annotate affinity ids
cDNA_df <- cDNA_df
cDNA_df$affinity_id <- "0_mixed_any"
cDNA_df$affinity_id[cDNA_df$affinity_pos1 == 0 & cDNA_df$affinity_pos2 == 0 & cDNA_df$affinity_pos3 == 4 & cDNA_df$affinity_pos4 == 4] <- "6_high_start"
cDNA_df$affinity_id[cDNA_df$affinity_pos1 == 4 & cDNA_df$affinity_pos2 == 4 & cDNA_df$affinity_pos3 == 0 & cDNA_df$affinity_pos4 == 0] <- "7_high_end"
cDNA_df$affinity_id[cDNA_df$affinity_pos1 == 3 & cDNA_df$affinity_pos2 == 3 & cDNA_df$affinity_pos3 == 3 & cDNA_df$affinity_pos4 == 3] <- "2_very_low_only"
cDNA_df$affinity_id[cDNA_df$affinity_pos1 == 0 & cDNA_df$affinity_pos2 == 0 & cDNA_df$affinity_pos3 == 0 & cDNA_df$affinity_pos4 == 0] <- "5_high_only"
cDNA_df$affinity_id[cDNA_df$affinity_pos1 == 1 & cDNA_df$affinity_pos2 == 1 & cDNA_df$affinity_pos3 == 1 & cDNA_df$affinity_pos4 == 1] <- "4_med_only"
cDNA_df$affinity_id[cDNA_df$affinity_pos1 == 2 & cDNA_df$affinity_pos2 == 2 & cDNA_df$affinity_pos3 == 2 & cDNA_df$affinity_pos4 == 2] <- "3_low_only"
cDNA_df$affinity_id[cDNA_df$affinity_pos1 == 4 & cDNA_df$affinity_pos2 == 4 & cDNA_df$affinity_pos3 == 4 & cDNA_df$affinity_pos4 == 4] <- "1_null_only"

# Add affinity parameters
## Activity vs. cumulative predicted affinity
affinity_df <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/library_design/mt20201123_affinity.csv", header = T, stringsAsFactors = F)
affinity_df$affinity[affinity_df$id == 4] <- 0
affinity_df$TF[affinity_df$TF == "Trp53"] <- "P53"
affinity_df$TF[affinity_df$TF == "Gr"] <- "GR"

### Complicated way of adding cum_affinity information to the df
for (i in unique(cDNA_df$affinity_pos1)) {
  for (j in unique(cDNA_df$affinity_pos2)) {
    for (k in unique(cDNA_df$affinity_pos3)) {
      for (l in unique(cDNA_df$affinity_pos4)) {
        for (m in unique(affinity_df$TF)) {
    cDNA_df$cum_affinity[cDNA_df$affinity_pos1 == i & cDNA_df$affinity_pos2 == j & cDNA_df$affinity_pos3 == k &
                         cDNA_df$affinity_pos4 == l & cDNA_df$tf == m] <- 
      affinity_df$affinity[affinity_df$id == i & affinity_df$TF == m] +
      affinity_df$affinity[affinity_df$id == j & affinity_df$TF == m] +
      affinity_df$affinity[affinity_df$id == k & affinity_df$TF == m] +
      affinity_df$affinity[affinity_df$id == l & affinity_df$TF == m]
        }
      }
    }
  }
}

### Also add the ddG and max_aff information
cDNA_df <- cDNA_df %>%
  group_by(reporter_id) %>%
  mutate(ddG = affinity_pos1 + affinity_pos2 + affinity_pos3 + affinity_pos4,
         max_aff = max(affinity_pos1,affinity_pos2,affinity_pos3,affinity_pos4),
         cum_affinity = cum_affinity/4 *100,
         cum_affinity = round(cum_affinity, 2))
```

## Analysis

### First insights into data distribution - reporter activity distribution plots
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
## reporter activity distribution over all experiments
colors <- c("#1B998B", "#2D3047", "#FF9B71")

## reporter activity distribution for each experiment individually
ggplot(data = cDNA_df, aes(x = reporter_activity_condition)) + 
  geom_density(fill = "#3D9F83") + 
  xlab("reporter activity") + 
  labs(title = "reporter activity distribution") +
  xlim(0,10) + 
  theme_bw() + 
  facet_wrap(~condition)

## reporter activity distribution - highlight negative motif controls
ggplot(cDNA_df %>%
         dplyr::select(reporter_activity, positive_ctrl, promoter, condition, neg_ctrls, outlier) %>%
         filter(positive_ctrl == "No", promoter != "random", outlier == "No", str_detect(condition, "MCF")) %>%
         ungroup() %>%
         unique(), 
       aes(x = neg_ctrls, y = reporter_activity)) + 
  geom_quasirandom(alpha = 0.4) + 
  geom_boxplot(alpha = 0.4)+
  xlab("reporter_activity") + 
  labs(title = "reporter activity distribution - negative ctrls") +
  ylim(0,40) + 
  theme_bw()+
  theme(text = element_text(size = 12)) + 
  facet_wrap(~condition)

## the same plot but now no facet
ggplot(cDNA_df %>%
         dplyr::select(reporter_activity, positive_ctrl, promoter, condition, neg_ctrls, outlier) %>%
         filter(positive_ctrl == "No", promoter != "random", outlier == "No", str_detect(condition, "MCF")) %>%
         setnames("neg_ctrls", "random_motif")%>%
         ungroup() %>%
         unique(), 
       aes(x = condition, y = reporter_activity, color = random_motif)) + 
  geom_quasirandom(alpha = 0.4, dodge.width = 1) + 
  geom_boxplot(alpha = 0.4, position = position_dodge(1))+
  ylab("reporter activity (a.u.)") + 
  labs(title = "reporter activity distribution - negative ctrls") +
  ylim(0,60) + 
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))+
  scale_color_manual(values = colors)+
  theme(text = element_text(size = 12))

## reporter activity per sample
ggplot(cDNA_df %>%
         filter(outlier == "No", neg_ctrls == "No", promoter == "mCMV", positive_ctrl == "No", 
                str_detect(condition, "MCF"), affinity_id != "1_null_only") %>%
         dplyr::select(condition, reporter_activity_condition, reporter_id) %>%
         unique(), 
       aes(x = condition, y = reporter_activity_condition)) + 
  geom_quasirandom() + 
  geom_boxplot(alpha = 0.4)+
  ylab("reporter activity (a.u.)") +
  ylim(0,25) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))

## reporter activity distribution - compare promoters
ggplot(cDNA_df %>%
         filter(outlier == "No", neg_ctrls == "No",
                str_detect(condition, "MCF"), affinity_id != "1_null_only") %>%
         dplyr::select(condition, reporter_activity_condition, reporter_id, positive_ctrl, promoter) %>%
         unique(),
       aes(x = condition, y = reporter_activity_condition, color = promoter)) + 
  geom_quasirandom(dodge.width = 1) +
  geom_boxplot(position = position_dodge(1), alpha = 0.4)+
  xlab("") + 
  ylab("reporter activity (a.u.)") +
  labs(title = "reporter activity distribution per minimal promoter") +
  scale_color_manual(values = colors) +
  ylim(0,25) +
  theme_bw()+
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))


## reporter activity distribution - highlight positive controls (O'Connell) vs negative controls
ggplot(cDNA_df %>%
         dplyr::select(reporter_activity, positive_ctrl, condition, neg_ctrls, promoter) %>%
         filter(neg_ctrls == "Yes" | positive_ctrl == "Yes") %>%
         unique(), 
       aes(x = "condition", y = reporter_activity, color = positive_ctrl, alpha = 0.4)) + 
  geom_quasirandom(dodge.width = 1) + 
  xlab("cDNA/pDNA") + 
  labs(title = "reporter activity distribution - promega ctrl vs negative ctrls") +
  scale_color_manual(values = colors) +
  ylim(0,15) + 
  theme_bw()+ 
  facet_wrap(~condition)


## reporter activity per condition - positive control
ggplot(cDNA_df %>%
         filter(outlier == "No", neg_ctrls == "No", promoter == "mCMV", 
                str_detect(condition, "MCF"), affinity_id != "1_null_only") %>%
         dplyr::select(condition, reporter_activity_condition, reporter_id, positive_ctrl) %>%
         unique(), 
       aes(x = condition, y = reporter_activity_condition, color = positive_ctrl)) + 
  geom_quasirandom(dodge.width = 1) + 
  geom_boxplot(alpha = 0.4, position = position_dodge(1))+
  labs(title = "reporter activity distribution vs. positive controls") +
  xlab("")+
  ylab("reporter activity (a.u.)")+
  ylim(0,25) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
  scale_color_manual(values = colors)

## activity per affinity at position 4
ggplot(cDNA_df %>%
         filter(outlier == "No", neg_ctrls == "No", promoter == "mCMV", positive_ctrl == "No", 
                str_detect(condition, "MCF"), affinity_id != "1_null_only") %>%
         dplyr::select(tf, reporter_activity_condition, affinity_pos4, reporter_id, condition) %>% 
         unique()) +
  geom_quasirandom(aes(x = affinity_pos4, y = reporter_activity_condition)) +
  theme_bw() +
  labs(title = "activity at position 4 per affinity (only P53-mCMV)")+
  facet_wrap(~condition)

## activity per spacing
ggplot(cDNA_df %>%
         filter(outlier == "No", neg_ctrls == "No", promoter == "mCMV", positive_ctrl == "No", 
                str_detect(condition, "MCF"), str_detect(affinity_id, "[2-5]_"), background == 1, position == 0) %>%
         dplyr::select(tf, reporter_activity_condition, spacing, reporter_id, condition) %>% 
         unique(), 
       aes(x = factor(spacing), y = reporter_activity_condition)) +
  geom_quasirandom()+
  geom_boxplot(alpha = .4) + 
  theme_bw() +
  ggtitle("reporter activity per spacing (P53-mCMV only)")+
  xlab("spacing length between motifs (bp)")+
  ylab("reporter activity (a.u.)")+
  facet_wrap(~condition)

## activity per spacing at different affinities
ggplot(cDNA_df %>%
         filter(outlier == "No", neg_ctrls == "No", promoter == "mCMV", positive_ctrl == "No", 
                str_detect(condition, "MCF7-WT"), str_detect(affinity_id, "[1,2,3]_"), background == 1, position == 0) %>%
         dplyr::select(tf, reporter_activity_condition, spacing, reporter_id, condition, affinity_id) %>% 
         unique(), 
       aes(x = spacing, y = reporter_activity_condition, color = affinity_id)) +
  geom_point()+
  geom_smooth(se = F)+
  theme_bw() +
  ggtitle("reporter activity per spacing")+
  xlab("spacing length between motifs (bp)")+
  ylab("reporter activity (a.u.)")+
  scale_color_brewer(palette = "Dark2")+
  facet_wrap(~condition)

## activity per position
ggplot(cDNA_df %>%
         filter(outlier == "No", neg_ctrls == "No", promoter == "mCMV", 
                str_detect(condition, "MCF"), affinity_id != "1_null_only") %>%
         dplyr::select(position, reporter_activity_condition, reporter_id, condition, positive_ctrl) %>% 
         unique(), 
         aes(x = factor(position), y = reporter_activity_condition, color = positive_ctrl)) +
  geom_quasirandom()+
  geom_boxplot(alpha = .4) + 
  theme_bw() +
  scale_color_manual(values = colors) +
  ggtitle("activity per position (P53-mCMV only)")+
  facet_wrap(~condition)

## activity per background
ggplot(cDNA_df %>%
         filter(outlier == "No", neg_ctrls == "No", promoter == "mCMV", positive_ctrl == "No", 
                str_detect(condition, "MCF"), affinity_id != "1_null_only", spacing == 7) %>%
         dplyr::select(tf, reporter_activity_condition, background, reporter_id, condition) %>% 
         unique(), aes(x = factor(background), y = reporter_activity_condition)) +
  geom_quasirandom() +
  geom_boxplot(alpha = .4) + 
  theme_bw() +
  labs(title = "reporter activity per background")+
  xlab("background sequence")+
  ylab("reporter activity (a.u.)")+
  facet_wrap(~condition)

## background per affinity
ggplot(cDNA_df %>%
         filter(outlier == "No", neg_ctrls == "No", promoter == "mCMV", positive_ctrl == "No", 
                str_detect(condition, "MCF"), str_detect(affinity_id, "[1-3]_"), spacing == 7) %>%
         dplyr::select(tf, reporter_activity_condition, background, reporter_id, condition, affinity_id) %>% 
         unique(), aes(x = factor(background), y = reporter_activity_condition, color = affinity_id)) +
  geom_quasirandom(dodge.width = .75) +
  geom_boxplot(position = position_dodge(.75))+
  scale_color_manual(values = colors)+
  theme_bw() +
  labs(title = "reporter activity per background")+
  xlab("background sequence")+
  ylab("reporter activity (a.u.)")+
  facet_wrap(~condition)


## Activity per affinity group
ggplot(cDNA_df %>%
         filter(outlier == "No", neg_ctrls == "No", promoter == "mCMV", positive_ctrl == "No", 
                str_detect(condition, "MCF")) %>%
         dplyr::select(tf, reporter_activity_condition, background, reporter_id, condition, affinity_id) %>%
         unique(), 
       aes(x = affinity_id, y = reporter_activity_condition)) +
  geom_quasirandom()+
  geom_boxplot(alpha = .4) + 
  ylim(0,25)+
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  labs(title = "activity distributions of affinities (only P53-mCMV)")+
  ggtitle("activity per affinity (P53-mCMV only)")+
  facet_wrap(~condition)

## Investigate low-only more closely
ggplot(cDNA_df %>%
         filter(outlier == "No", neg_ctrls == "No", promoter == "mCMV", positive_ctrl == "No", 
                str_detect(condition, "MCF"), str_detect(affinity_id, "[2-3]_")) %>%
         dplyr::select(reporter_activity_condition, background, reporter_id, condition, affinity_id) %>%
         unique(), 
       aes(x = condition, y = reporter_activity_condition, color = affinity_id)) +
  geom_quasirandom(dodge.width = (0.75))+
  geom_boxplot(alpha = .4, position = position_dodge(0.75)) + 
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
  labs(title = "reporter activity per affinity")+
  ylab("reporter activity (a.u.)")+
  xlab("")+
  scale_color_manual(values = colors)
  
## Activity per ddG
ggplot(cDNA_df %>%
         filter(outlier == "No", neg_ctrls == "No", promoter == "mCMV", positive_ctrl == "No", 
                str_detect(condition, "MCF"), affinity_id != "1_null_only") %>%
         dplyr::select(reporter_activity_condition, reporter_id, condition, ddG) %>%
         unique(),
       aes(x = factor(ddG), y = reporter_activity_condition)) +
  geom_quasirandom() +
  geom_boxplot(alpha = .4)+
  theme_bw() +
  ylim(0,60)+
  labs(title = "reporter activity per ddG (P53_mCMV)")+
  facet_wrap(~condition)
  
## Do the pDNA counts impact the activity?
ggplot(cDNA_df %>%
         filter(outlier == "No", neg_ctrls == "No", promoter == "mCMV", positive_ctrl == "No", 
                str_detect(condition, "MCF"), affinity_id != "1_null_only") %>%
         dplyr::select(pDNA_counts_rpm, reporter_id, ddG) %>%
         unique(),
       aes(x = factor(ddG), y = pDNA_counts_rpm)) +
  geom_quasirandom() +
  geom_boxplot(alpha = .4)+
  theme_bw() +
  labs(title = "reporter activity per ddG (P53_mCMV)")

## Activity per max_affinity
ggplot(cDNA_df %>%
         filter(outlier == "No", neg_ctrls == "No", promoter == "mCMV", positive_ctrl == "No", 
                str_detect(condition, "MCF"), affinity_id != "1_null_only") %>%
         dplyr::select(reporter_activity_condition, reporter_id, condition, max_aff) %>%
         unique(),
       aes(x = factor(max_aff), y = reporter_activity_condition)) +
  geom_quasirandom() +
  geom_boxplot(alpha = 0.3)+
  theme_bw() +
  ylim(0,60)+
  labs(title = "reporter activity max affinity (P53_mCMV)")+
  facet_wrap(~condition)

## Activity per cumulative affinity
ggplot(cDNA_df %>%
         filter(outlier == "No", neg_ctrls == "No", promoter == "mCMV", positive_ctrl == "No", 
                str_detect(condition, "MCF")) %>%
         dplyr::select(reporter_activity_condition, reporter_id, condition, cum_affinity) %>%
         unique(),
       aes(x = factor(cum_affinity), y = reporter_activity_condition)) +
  geom_quasirandom(alpha = .4) +
  geom_boxplot(alpha = 0.4) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))+
  ylab("reporter activity (a.u.)")+
  xlab("predicted cumulative reporter activity (%)")+
  labs(title = "reporter activity per affinity")+
  facet_wrap(~condition, nrow = 2)









## correlation bc-counts pDNA vs. cDNA
sp <- ggscatter(cDNA_df %>%
                  filter(neg_ctrls  == "No", rand_promoter == "No", outlier == "No"),
               x = "pDNA_counts_rpm", y = "rpm",
               size = 0.1, alpha = 0.3, color = "#2D3047", 
   add = "reg.line",
   add.params = list(color = "blue", fill = "lightgray"), title = "normalized barcode counts cDNA vs. pDNA",
   conf.int = TRUE, ylab = "cDNA", xlab = "pDNA")
sp + stat_cor(method = "pearson", label.x = 150, label.y = 2) + 
  geom_abline(linetype = "dashed") + facet_wrap(~condition) + xlim(0,2000) + ylim(0,2000)


## correlation bc-counts pDNA vs. reporter activity
sp <- ggscatter(cDNA_df %>%
                  filter(neg_ctrls  == "No", rand_promoter == "No", outlier == "No"),
          size = 0.75, alpha = 0.1, color = neg_ctrls,
          y = "log_activity", x = "pDNA_counts_rpm",
          title = "reporter pDNA counts vs log2 reporter activity",
          ylab = "log2 reporter activity",
          xlab = "pDNA counts per million",
          facet.by = "condition")
sp + xlim(0,200) + ylim(-5,5) +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10), 
          axis.text.y = element_text(size = 10)) +
  theme(text = element_text(size = 12))
```



### Plot mean activities per affinity
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# Plot mean activites per affinity rank
ggplot(cDNA_df %>% 
         filter(outlier == "No", neg_ctrls == "No", promoter == "mCMV", positive_ctrl == "No", 
                str_detect(condition, "MCF7-WT")) %>%
         dplyr::select(cum_affinity, condition, reporter_activity_condition) %>%
         group_by(cum_affinity, condition) %>%
         mutate(reporter_activity_condition = mean(reporter_activity_condition, na.rm = T)) %>%
         ungroup() %>%
         dplyr::select(-reporter_id) %>%
         unique() %>%
         na.omit() %>%
         arrange(cum_affinity) %>%
         group_by(condition) %>%
         mutate(cum_affinity = 1:20) %>%
         ungroup(), 
       aes(x = cum_affinity, y = reporter_activity_condition, color = condition)) +
  geom_smooth(method = "lm", formula = y ~ poly(x, 8), se = F)+
  ylab("fitted mean reporter activity (a.u.)") +
  xlab("affinity rank")+
  labs(title = "reporter activity per affinity", subtitle = "DMSO = green, Nutlin = black")+
  scale_color_manual(values = colors)+
  theme_bw() +
  theme(legend.position = "none")

# Now plot the difference between Nutlin and DMSO
ggplot(cDNA_df %>% 
         filter(outlier == "No", neg_ctrls == "No", promoter == "mCMV", positive_ctrl == "No", 
                str_detect(condition, "MCF7-WT")) %>%
         dplyr::select(cum_affinity, condition, reporter_activity_condition) %>%
         group_by(cum_affinity, condition) %>%
         mutate(reporter_activity_condition = mean(reporter_activity_condition, na.rm = T)) %>%
         ungroup() %>%
         dplyr::select(-reporter_id) %>%
         unique() %>%
         na.omit() %>%
         group_by(condition) %>%
         arrange(cum_affinity) %>%
         mutate(cum_affinity = 1:20) %>%
         ungroup() %>%
         dcast(cum_affinity ~ condition) %>%
         mutate(dif = `MCF7-WT_Nutlin` / `MCF7-WT_DMSO`), 
       aes(x = cum_affinity, y = dif)) +
  geom_smooth(method = "lm", formula = y ~ poly(x, 8), se = F, color = "#1B998B")+
  geom_hline(yintercept = 1, linetype = "dashed") +
  ylab("reporter activity, Nutlin/DMSO") +
  xlab("affinity rank")+
  labs(title = "differential reporter activity per affinity")+
  theme_bw() +
  theme(legend.position = "none")
```









### Compute enrichment over 0 affinity
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# Compute for each reporter the activity relative to the 0 activity reporter
cDNA_df <- cDNA_df %>%
  dplyr::mutate(reporter_id_2 = gsub("_a1.*", "", reporter_id))

zero_activity <- cDNA_df %>%
  filter(affinity_id == "1_null_only", outlier == "No") %>%
  dplyr::select('zero_activity' = matches("^reporter_activity$"), matches("^reporter_id_2$"), matches("^sample$")) %>% 
  unique()
zero_activity <- zero_activity[,-1] %>% unique()
zero_activity$zero_activity <- ave(zero_activity$zero_activity, zero_activity$reporter_id_2, 
                                  zero_activity$sample, FUN = function(x) mean(x))

cDNA_df <- merge(cDNA_df, zero_activity, all = T)

cDNA_df <- cDNA_df %>% 
  dplyr::mutate(rel_reporter_activity = reporter_activity / zero_activity,
                rel_reporter_activity_condition = 
                  ave(rel_reporter_activity, condition, reporter_id, FUN = function(x) mean(x)))

# Plot relative reporter activities for each condition per affinity
ggplot(cDNA_df %>%
         filter(outlier == "No", neg_ctrls == "No", promoter == "mCMV") %>%
         dplyr::select(cum_affinity, rel_reporter_activity, sample, reporter_id) %>%
         unique(),
       aes(x = factor(cum_affinity), y = rel_reporter_activity)) +
  geom_quasirandom(alpha = 0.3, color =  "#2D3047") +
  geom_boxplot(alpha = 0.4)+
  theme_bw() +
  ylim(0,10)+
  labs(title = "reporter activity per cum_aff(P53_mCMV)")+
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  facet_wrap(~sample)

ggplot(cDNA_df %>%
         filter(outlier == "No", neg_ctrls == "No", promoter == "mCMV", str_detect(sample,"MCF")) %>%
         dplyr::select(cum_affinity, rel_reporter_activity_condition, condition, reporter_id) %>%
         unique(),
       aes(x = factor(cum_affinity), y = rel_reporter_activity_condition)) +
  geom_quasirandom(alpha = 0.3, color =  "#2D3047") +
  geom_boxplot(alpha = 0.4)+
  theme_bw() +
  ylim(0,6)+
  labs(title = "reporter activity vs reporter affinity")+
  xlab("reporter affinity (%, relative to 4x consensus motif)")+
  ylab("reporter activity (fold change over 0 affinity reporter)")+
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  facet_wrap(~condition)
```



### Compare conditions
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# Compare per affinity the reporter activity distributions per condition
ggplot(cDNA_df %>%
         filter(outlier == "No", neg_ctrls == "No", promoter == "mCMV") %>%
         dplyr::select(cum_affinity, rel_reporter_activity_condition, condition, reporter_id) %>%
         unique(), 
       aes(x = condition, y = rel_reporter_activity_condition))+
  geom_quasirandom(alpha = 0.3, color =  "#2D3047") +
  geom_boxplot(alpha = 0.4)+
  theme_bw()+  
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  facet_wrap(~cum_affinity)

ggplot(cDNA_df %>%
         dplyr::select(cum_affinity, rel_reporter_activity_condition, condition, reporter_id, outlier, neg_ctrls, promoter) %>%
         filter(outlier == "No", neg_ctrls == "No", promoter == "mCMV", str_detect(condition,"MCF")) %>%
         unique(), 
       aes(x = condition, y = rel_reporter_activity_condition))+
  geom_quasirandom(alpha = 0.3, color =  "#2D3047") +
  geom_boxplot(alpha = 0.4)+
  theme_bw()+  
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 12)) +
  facet_wrap(~cum_affinity)

for (i in unique(cDNA_df$cum_affinity)){
  if (nrow(cDNA_df[cDNA_df$cum_affinity == i,]) > 1){
  p <- ggplot(cDNA_df[cDNA_df$cum_affinity == i,] %>%
         dplyr::select(cum_affinity, rel_reporter_activity_condition, condition, reporter_id, outlier, neg_ctrls, promoter) %>%
         filter(outlier == "No", neg_ctrls == "No", promoter == "mCMV", str_detect(condition,"MCF")) %>%
         unique(), 
       aes(x = condition, y = rel_reporter_activity_condition))+
  geom_quasirandom(alpha = 0.3, color =  "#2D3047") +
  geom_boxplot(alpha = 0.4)+
  theme_bw()+  
    xlab("")+
    ylab("reporter activity")+
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10), 
        axis.text.y = element_text(size = 10)) +
    ylim(0,5)+
  ggtitle(paste("reporter affinity: ", i, "%", sep = ""))
  print(p)
  }
}
```


### Compare all special affinity groups
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
ggplot(cDNA_df %>%
         dplyr::select(affinity_id, rel_reporter_activity_condition, condition, reporter_id, outlier, neg_ctrls, promoter) %>%
         filter(outlier == "No", neg_ctrls == "No", promoter == "mCMV", str_detect(condition,"MCF")) %>%
         unique(), 
       aes(x = affinity_id, y = rel_reporter_activity_condition))+
  geom_quasirandom(alpha = 0.3, color =  "#2D3047") +
  geom_boxplot(alpha = 0.4)+
  theme_bw()+  
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  facet_wrap(~condition)
```



### Compare activities in one plot
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# Plot mean reporter activity per affinity per condition - connected line plot
cDNA_df_aff<- cDNA_df %>%
         filter(outlier == "No", neg_ctrls == "No", promoter == "mCMV", str_detect(condition,"MCF7")) %>%
         dplyr::select(rel_reporter_activity_condition, condition, cum_affinity) %>%
         unique() %>%
         group_by(condition, cum_affinity) %>%
         mutate(activity = mean(rel_reporter_activity_condition, na.rm = T)) %>%
         select(-rel_reporter_activity_condition) %>%
         unique()
cDNA_df_aff_wide <- cDNA_df_aff %>% 
  unique() %>%
  ungroup() %>%
  mutate(stimulation = gsub(".*_(.*$)", "\\1", condition),
         condition = gsub("_.*$", "\\1", condition)) %>%
  dcast(cum_affinity + stimulation ~ condition, value.var = "activity")


ggplot(cDNA_df_aff_wide %>% group_by(stimulation) %>%
         mutate(cum_affinity = 1:20) %>%
         ungroup(), aes(x = cum_affinity, y = `MCF7-WT`)) +
  geom_point(color = "#1B998B", size = 2)+
  geom_line(color = "#1B998B") + 
  geom_point(aes(y = `MCF7-KO`), color = "#2D3047", size = 2) +
  geom_line(aes(y = `MCF7-KO`)) + 
  geom_ribbon(aes(ymin = `MCF7-KO`, ymax = `MCF7-WT`), fill = "#2D3047", alpha = .3)+
  ylab("reporter activity P53+/+ (green) and P53-/- (black)")+
  xlab("affinity rank")+
  theme_bw() +
  facet_wrap(~stimulation)

# Plot difference between KO and WT
cDNA_df_aff_wide <- cDNA_df_aff_wide %>%
  mutate(dif = `MCF7-WT` / `MCF7-KO`)

ggplot(cDNA_df_aff_wide %>% group_by(stimulation) %>%
         mutate(cum_affinity = 1:20) %>%
         ungroup(), aes(x = cum_affinity, y = dif)) +
  geom_point(aes(fill = "#2D3047"))+
  geom_line() +
  geom_hline(yintercept = 1) + 
  geom_ribbon(aes(ymin = 1, ymax = dif), fill = "#2D3047", alpha = .5)+
  ylab("FC reporter activity P53+/+ over P53-/-")+
  xlab("affinity rank")+
  theme_bw() +
  facet_wrap(~stimulation)

ggplot(cDNA_df_aff_wide %>% group_by(stimulation) %>%
         mutate(cum_affinity = 1:20) %>%
         ungroup(), aes(x = cum_affinity, y = dif)) +
  geom_point(aes(fill = "#2D3047"))+ 
  geom_smooth(method = "loess", fill = "#2D3047", color = "#2D3047", alpha = 0.4)+
  geom_hline(yintercept = 1) + 
  ylab("FC reporter activity WT/KO")+
  xlab("affinity rank")+
  ggtitle("MCF7+DMSO: WT over KO activity")+
  theme_bw()+
  facet_wrap(~stimulation)

ggplot(cDNA_df_aff_wide %>% group_by(stimulation) %>%
         mutate(cum_affinity = 1:20) %>%
         ungroup(), aes(x = cum_affinity, y = dif, group = stimulation, color = stimulation, fill = stimulation)) +
  geom_point()+
  geom_line() +
  geom_hline(yintercept = 1) + 
  geom_ribbon(aes(ymin = 1, ymax = dif), alpha = .2)+
  ylab("FC reporter activity P53+/+ over P53-/-")+
  xlab("reporter affinity rank")+
  scale_color_manual(values = colors[c(2,1)])+
  scale_fill_manual(values = colors[c(2,1)])+
  theme_bw()

ggplot(cDNA_df_aff_wide %>% group_by(stimulation) %>%
           mutate(cum_affinity = 1:20) %>%
           ungroup(), aes(x = cum_affinity, y = dif, group = stimulation, color = stimulation, fill = stimulation)) +
    geom_smooth(method = "lm", formula = y ~ poly(x, 8), se = F)+
    geom_hline(yintercept = 1) + 
  ylab("FC reporter activity P53+/+ over P53-/-")+
  xlab("reporter affinity rank")+
    scale_color_manual(values = colors[c(2,1)])+
    scale_fill_manual(values = colors[c(2,1)])+
    theme_bw()

# Same plot but now show the differential track for Nutlin vs. DMSO (for both KO & WT)
cDNA_df_aff_wide_2 <- cDNA_df_aff %>% 
  unique() %>%
  ungroup() %>%
  mutate(cells = gsub(".*-(.*)_.*", "\\1", condition),
         condition = gsub(".*_", "\\1", condition)) %>%
  dcast(cum_affinity + cells ~ condition, value.var = "activity") %>%
  mutate(dif = Nutlin / DMSO)

ggplot(cDNA_df_aff_wide_2 %>% filter(cells == "WT") %>% group_by(cells) %>%
           mutate(cum_affinity = 1:20) %>%
           ungroup(), aes(x = cum_affinity, y = Nutlin)) +
  geom_point(color = "#1B998B", size = 2)+
  geom_line(color = "#1B998B") + 
  geom_point(aes(y = DMSO), color = "#2D3047", size = 2) +
  geom_line(aes(y = DMSO)) + 
  ylab("mean reporter activity (fold change over 0 affinity)")+ 
  ggtitle("Reporter activity vs affinity in DMSO (black) and Nutlin (green)")+
  xlab("affinity rank")+
  theme_bw()


ggplot(cDNA_df_aff_wide_2 %>% group_by(cells) %>%
         mutate(cum_affinity = 1:20) %>%
         ungroup(), aes(x = cum_affinity, y = dif, group = cells, color = cells, fill = cells)) +
  geom_point()+
  geom_line() +
  geom_hline(yintercept = 1) + 
  geom_ribbon(aes(ymin = 1, ymax = dif), alpha = .5)+
  ylab("FC reporter activity Nutlin/DMSO")+
  xlab("affinity rank")+
  ggtitle("Nutlin over DMSO activity")+
  scale_color_manual(values = colors[c(2,1)])+
  scale_fill_manual(values = colors[c(2,1)])+
  theme_bw()

ggplot(cDNA_df_aff_wide_2 %>% group_by(cells) %>%
         mutate(cum_affinity = 1:20) %>%
         ungroup(), aes(x = cum_affinity, y = dif, group = cells, color = cells, fill = cells)) +
  geom_hline(yintercept = 1) + 
  geom_smooth(method = "lm", formula = y ~ poly(x, 8), se = F)+
  ylab("FC reporter activity Nutlin/DMSO")+
  xlab("affinity rank")+
  ggtitle("Nutlin over DMSO activity")+
  scale_color_manual(values = colors[c(2,1)])+
  scale_fill_manual(values = colors[c(2,1)])+
  theme_bw()
```






### Linear model - only unmixed reporter population (include spacing & background)
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
## Rule of thumb for fitting models: e^n_features < n_observations -> I will try to fit 4 features for 114 observations

## Subselect p53 only 
cDNA_df_p53 <- cDNA_df %>%
  filter(neg_ctrls == "No", positive_ctrl == "No", outlier == "No", str_detect(affinity_id, "[1-5]_"),
         promoter == "mCMV", str_detect(condition, "MCF")) %>%
  dplyr::select(reporter_id, condition, reporter_activity_condition, 
                affinity_pos1,spacing, background) %>%
  mutate(spacing = as.factor(spacing),
         background = as.factor(background)) %>%
  unique()

# Rename 0,1,2,3,4 in null, very-weak, weak, medium, strong
replace <- data.frame("old" = c(0:4), "new" = c("_4_strong", "_3_medium", "_2_weak", "_1_very-weak", "_0_null"), stringsAsFactors=FALSE)
for (i in unique(replace$old)) {
  cDNA_df_p53$affinity_pos1[cDNA_df_p53$affinity_pos1 == i] <- replace$new[replace$old == i]
}

## Explain which affinities are essential at which position to boost activity
### Calculate expression variance
exp_variance <- data.frame("feature" = c(colnames(cDNA_df_p53 %>% 
                                                    dplyr::select(-reporter_id, -condition, -reporter_activity_condition)),
                                         "Residuals"), stringsAsFactors=FALSE)

prediction <- data.frame("id" = 1:(nrow(cDNA_df_p53)), stringsAsFactors=FALSE)

weight <- data.frame("feature" = merge(colnames(cDNA_df_p53 %>% 
                                                  dplyr::select(-reporter_id, -condition, -reporter_activity_condition,
                                                                -spacing, -background)),
                                       unique(cDNA_df_p53$affinity_pos1), all = T)) %>%
  mutate(feature = paste(feature.x, feature.y, sep = "")) %>% dplyr::select("feature") 

weight_2 <- data.frame("feature" = merge(colnames(cDNA_df_p53 %>% 
                                                  dplyr::select(spacing)),
                                       unique(cDNA_df_p53$spacing), all = T)) %>%
  mutate(feature = paste(feature.x, feature.y, sep = "")) %>% dplyr::select("feature")

weight_3 <- data.frame("feature" = merge(colnames(cDNA_df_p53 %>% 
                                                  dplyr::select(background)),
                                       unique(cDNA_df_p53$background), all = T)) %>%
  mutate(feature = paste(feature.x, feature.y, sep = "")) %>% dplyr::select("feature") 

weight <- rbind(weight, weight_2, weight_3, "(Intercept)", stringsAsFactors=FALSE)



cDNA_df_p53 <- cDNA_df_p53[order(cDNA_df_p53$condition),-1]
cDNA_df_p53$row <- rownames(cDNA_df_p53)

for (i in unique(cDNA_df_p53$condition)) {
  x <- lm(reporter_activity_condition ~ affinity_pos1 + spacing + background, 
                     cDNA_df_p53[cDNA_df_p53$condition == i,])
  y <- data.frame(x$fitted.values, stringsAsFactors=FALSE) %>% rownames_to_column()
  par(mfrow=c(2,2))
  plot(x, main = i)
  names(y) <- c("row", i)
  y$id <- 1:nrow(y)
  prediction <- merge(prediction,y, all = T)
  
  w <- data.frame(x$coefficients, stringsAsFactors=FALSE) %>% rownames_to_column()
  names(w) <- c("feature", i)
  weight <- merge(weight,w, all = T)
  
  x <- anova(x) %>% rownames_to_column("feature") %>%
    setnames("Sum Sq", "sum_sq") %>%
    dplyr::select(feature, sum_sq) %>%
    mutate(sum = sum(sum_sq)) %>%
    mutate(rel_sum_sq = (sum_sq/sum)*100) %>%
    setnames("rel_sum_sq", i)
  x <- x[,c(1,4)]
  
  exp_variance <- merge(exp_variance, x)
}


## Prepare lm parameters for plotting
exp_variance <- melt(exp_variance, variable.name = "condition")
exp_variance$group <- "no residual"
exp_variance$group[grep("Residual", exp_variance$feature)] <- "residual"
prediction <- prediction %>% 
  dplyr::select(-id) %>% 
  melt(id.vars = "row", variable.name = "condition", value.name = "activity_predicted") %>% 
  dplyr::select(-condition) %>% 
  na.omit()
cDNA_df_p53 <- merge(cDNA_df_p53, prediction)
weight <- weight %>% filter(feature != "(Intercept)")
weight <- melt(weight, variable.name = "condition")
categorie <- "affinity_pos"
weight$features <- gsub(paste(categorie, collapse="|"), "",weight$feature)
feature <- weight$features
feature <- gsub("^.{1}", "", feature)
weight$cond <- gsub(paste(feature, collapse="|"), "", weight$feature)
weight$value[is.na(weight$value)] <- 0

# Further plotting preparations
exp_variance$value[exp_variance$feature == "Residuals"] <- 100 - exp_variance$value[exp_variance$feature == "Residuals"]
exp_variance$feature <- revalue(exp_variance$feature, c("Residuals"="total"))

# Plotting
ggplot(exp_variance, 
       aes(x = reorder(feature, -value), y = value, fill = group)) +
  scale_fill_manual(values = c("#1B998B", "#2D3047")) +
  geom_bar(stat = "identity") + 
  ylab("Variance explained (%)") + xlab("") + 
  labs(title = "Linear variance modelling", 
       subtitle = "Expression variance predicted by features") +
  theme_bw() +
  theme(text = element_text(size = 12)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    guides(fill = F) + facet_wrap(~condition)

ggscatter(cDNA_df_p53, x = "reporter_activity_condition", y = "activity_predicted",
          add = "reg.line",
          add.params = list(color = "blue", fill = "lightgray"),
          conf.int = TRUE, ylab = "Predicted log2-activity", size = 0.5,
          xlab = "Average log2-activity") + 
  theme_bw() + 
  facet_wrap(~condition) +
  stat_cor(aes(label = ..r.label..), method = "pearson", label.x = 0, label.y = 3)

ggplot(weight, 
       aes(x = features, y = value, fill = cond)) +
  scale_fill_manual(values = c("#1B998B", "#2D3047", "#FF9B71", "#ECDD7B", "#C2B2B4", "#A54657")) +
  ggtitle(i)+
  geom_bar(stat = "identity") + 
  ylab("Weight") + xlab("") +
  theme_bw() +
  theme(text = element_text(size = 12)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    guides(fill = F) + 
  facet_wrap(~condition)
```
















### Explain expression differences betweeen the different affinities
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
## Rule of thumb for fitting models: e^n_features < n_observations -> I will try to fit 4 features for 114 observations

## Subselect p53 only 
cDNA_df_p53 <- cDNA_df %>%
  filter(neg_ctrls == "No", positive_ctrl == "No", outlier == "No", 
         spacing == "7", background == "1", affinity_pos1 %in% c(3,4), affinity_pos2 %in% c(3,4), 
           affinity_pos3 %in% c(3,4), affinity_pos4 %in% c(3,4),
         promoter == "mCMV", str_detect(condition, "MCF")) %>%
  dplyr::select(reporter_id, condition, reporter_activity_condition, 
                affinity_pos1, affinity_pos2, affinity_pos3, affinity_pos4) %>% 
  unique()

# Rename 0,1,2,3,4 in null, very-weak, weak, medium, strong
replace <- data.frame("old" = c(0:4), "new" = c("_4_strong", "_3_medium", "_2_weak", "_1_very-weak", "_0_null"), stringsAsFactors=FALSE)
for (i in unique(replace$old)) {
  cDNA_df_p53$affinity_pos1[cDNA_df_p53$affinity_pos1 == i] <- replace$new[replace$old == i]
  cDNA_df_p53$affinity_pos2[cDNA_df_p53$affinity_pos2 == i] <- replace$new[replace$old == i]
  cDNA_df_p53$affinity_pos3[cDNA_df_p53$affinity_pos3 == i] <- replace$new[replace$old == i]
  cDNA_df_p53$affinity_pos4[cDNA_df_p53$affinity_pos4 == i] <- replace$new[replace$old == i]
}

## Explain which affinities are essential at which position to boost activity
### Calculate expression variance
exp_variance <- data.frame("feature" = c(colnames(cDNA_df_p53 %>% 
                                                    dplyr::select(-reporter_id, -condition, -reporter_activity_condition)),
                                         "Residuals"), stringsAsFactors=FALSE)

prediction <- data.frame("id" = 1:(nrow(cDNA_df_p53)), stringsAsFactors=FALSE)

weight <- data.frame("feature" = merge(colnames(cDNA_df_p53 %>% 
                                                  dplyr::select(-reporter_id, -condition, -reporter_activity_condition)),
                                       unique(cDNA_df_p53$affinity_pos1), all = T)) %>%
  mutate(feature = paste(feature.x, feature.y, sep = "")) %>% dplyr::select("feature") %>% 
  rbind("(Intercept)", stringsAsFactors=FALSE)


cDNA_df_p53 <- cDNA_df_p53[order(cDNA_df_p53$condition),-1]
cDNA_df_p53$row <- rownames(cDNA_df_p53)

for (i in unique(cDNA_df_p53$condition)) {
  x <- lm(reporter_activity_condition ~ affinity_pos1 * affinity_pos4 + affinity_pos4 * affinity_pos2 + affinity_pos4 * affinity_pos3, 
                     cDNA_df_p53[cDNA_df_p53$condition == i,])
  y <- data.frame(x$fitted.values, stringsAsFactors=FALSE) %>% rownames_to_column()
  par(mfrow=c(2,2))
  plot(x, main = i)
  names(y) <- c("row", i)
  y$id <- 1:nrow(y)
  prediction <- merge(prediction,y, all = T)
  
  w <- data.frame(x$coefficients, stringsAsFactors=FALSE) %>% rownames_to_column()
  names(w) <- c("feature", i)
  weight <- merge(weight,w, all = T)
  
  x <- anova(x) %>% rownames_to_column("feature") %>%
    setnames("Sum Sq", "sum_sq") %>%
    dplyr::select(feature, sum_sq) %>%
    mutate(sum = sum(sum_sq)) %>%
    mutate(rel_sum_sq = (sum_sq/sum)*100) %>%
    setnames("rel_sum_sq", i)
  x <- x[,c(1,4)]
  
  exp_variance <- merge(exp_variance, x)
}


## Prepare lm parameters for plotting
exp_variance <- melt(exp_variance, variable.name = "condition")
exp_variance$group <- "no residual"
exp_variance$group[grep("Residual", exp_variance$feature)] <- "residual"
prediction <- prediction %>% 
  dplyr::select(-id) %>% 
  melt(id.vars = "row", variable.name = "condition", value.name = "activity_predicted") %>% 
  dplyr::select(-condition) %>% 
  na.omit()
cDNA_df_p53 <- merge(cDNA_df_p53, prediction)
weight <- weight %>% filter(feature != "(Intercept)")
weight <- melt(weight, variable.name = "condition")
categorie <- "affinity_pos"
weight$features <- gsub(paste(categorie, collapse="|"), "",weight$feature)
feature <- weight$features
feature <- gsub("^.{1}", "", feature)
weight$cond <- gsub(paste(feature, collapse="|"), "", weight$feature)
weight$value[is.na(weight$value)] <- 0

# Further plotting preparations
exp_variance$value[exp_variance$feature == "Residuals"] <- 100 - exp_variance$value[exp_variance$feature == "Residuals"]
exp_variance$feature <- revalue(exp_variance$feature, c("Residuals"="total"))

# Plotting
ggplot(exp_variance, 
       aes(x = reorder(feature, -value), y = value, fill = group)) +
  scale_fill_manual(values = c("#1B998B", "#2D3047")) +
  geom_bar(stat = "identity") + 
  ylab("Variance explained (%)") + xlab("") + 
  labs(title = "Linear variance modelling", 
       subtitle = "Expression variance predicted by features") +
  theme_bw() +
  theme(text = element_text(size = 12)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    guides(fill = F) + facet_wrap(~condition)

ggscatter(cDNA_df_p53, x = "reporter_activity_condition", y = "activity_predicted",
          add = "reg.line",
          add.params = list(color = "blue", fill = "lightgray"),
          conf.int = TRUE, ylab = "Predicted log2-activity", size = 0.5,
          xlab = "Average log2-activity") + 
  theme_bw() + 
  facet_wrap(~condition) +
  stat_cor(aes(label = ..r.label..), method = "pearson", label.x = 0, label.y = 10)

ggplot(weight, 
       aes(x = features, y = value, fill = cond)) +
  scale_fill_manual(values = c("#1B998B", "#2D3047", "#FF9B71", "#ECDD7B", "#C2B2B4", "#A54657")) +
  ggtitle(i)+
  geom_bar(stat = "identity") + 
  ylab("Weight") + xlab("") +
  theme_bw() +
  theme(text = element_text(size = 12)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    guides(fill = F) + 
  facet_wrap(~condition)

ggplot(weight %>% filter(str_detect(feature, "_very-weak$"), str_detect(condition,"WT")), 
       aes(x = features, y = value, fill = cond)) +
    scale_fill_manual(values = c("#1B998B", "#2D3047", "#FF9B71", "#ECDD7B", "#C2B2B4", "#A54657")) +
    geom_bar(stat = "identity") + 
    ylab("Weight") + xlab("") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    guides(fill = F) + 
    facet_wrap(~condition)



enrichment_df <- cDNA_df %>%
         filter(str_detect(condition, "MCF7-WT"), outlier == "No", neg_ctrls == "No", promoter == "mCMV",
                affinity_pos1 %in% c(2,3), affinity_pos2 %in% c(2,3), 
                affinity_pos3 %in% c(2,3), affinity_pos4 %in% c(2,3)) %>%
         dplyr::select(reporter_activity_condition, affinity_pos4, affinity_pos3, affinity_pos2, affinity_pos1, condition) %>%
  mutate(position_4 = ave(reporter_activity_condition, condition, affinity_pos4, FUN = function(x) mean(x)),
         position_3 = ave(reporter_activity_condition, condition, affinity_pos3, FUN = function(x) mean(x)),
         position_2 = ave(reporter_activity_condition, condition, affinity_pos2, FUN = function(x) mean(x)),
         position_1 = ave(reporter_activity_condition, condition, affinity_pos1, FUN = function(x) mean(x))) %>%
  dplyr::select(-reporter_activity_condition) %>%
  melt(id.vars = c("condition","affinity_pos1", "affinity_pos2", "affinity_pos3", "affinity_pos4"), variable.name = "affinity_pos", value.name = "activity") %>%
  unique()


ggplot(enrichment_df,
       aes(x = affinity_pos, y = activity, color = condition)) +
  geom_quasirandom(dodge.width = 0.75) +
  geom_boxplot(position = position_dodge(0.75), alpha = .4) +
  scale_color_manual(values = colors)+
  guides(fill = F)+
  ylab("reporter activity (a.u.)")+
  ggtitle("weak (green) vs very weak (black) motif at position 4")+
  theme_bw()
```
*The site closest to the minimal promoter determines the activity. Can neighboring sites even inhibit this effect?*






### Ridge/Lasso regression
```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# Train model
fit <- cva.glmnet(log_reporter_activity ~ affinity_pos1 + affinity_pos2 + affinity_pos3 + affinity_pos4, data=cDNA_df_p53)

# Get all parameters
get_model_params <- function(fit) {
  alpha <- fit$alpha
  lambdaMin <- sapply(fit$modlist, `[[`, "lambda.min")
  lambdaSE <- sapply(fit$modlist, `[[`, "lambda.1se")
  error <- sapply(fit$modlist, function(mod) {min(mod$cvm)})
  best <- which.min(error)
  data.frame(alpha = alpha[best], lambdaMin = lambdaMin[best],
             lambdaSE = lambdaSE[best], eror = error[best])
}

x <- get_model_params(fit)

# Perform GLM
cDNA_dfMod <- glm(log_reporter_activity ~ affinity_pos1 + affinity_pos2 + affinity_pos3 + affinity_pos4, data=cDNA_df_p53, family = "gaussian")
print(cDNA_dfMod)

```








### Random forest implementation
```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# Select features and activity to predict
cDNA_df_p53 <- cDNA_df[cDNA_df$tf == "P53" &
                         cDNA_df$neg_ctrls == "No" &
                         cDNA_df$positive_ctrl == "No" &
                         cDNA_df$sample == "MCF7-WT_Nutlin" &
                         cDNA_df$promoter == "mCMV",] %>%
  dplyr::select(log_reporter_activity, ddG, spacing, position, background) %>% unique()

# for (i in unique(replace$old)) {
#   cDNA_df_p53$affinity_pos1[cDNA_df_p53$affinity_pos1 == i] <- replace$new[replace$old == i]
#   cDNA_df_p53$affinity_pos2[cDNA_df_p53$affinity_pos2 == i] <- replace$new[replace$old == i]
#   cDNA_df_p53$affinity_pos3[cDNA_df_p53$affinity_pos3 == i] <- replace$new[replace$old == i]
#   cDNA_df_p53$affinity_pos4[cDNA_df_p53$affinity_pos4 == i] <- replace$new[replace$old == i]
# }

cDNA_df_p53$spacing <- as.character(cDNA_df_p53$spacing)
cDNA_df_p53$background <- as.character(cDNA_df_p53$background)
cDNA_df_p53$position <- as.character(cDNA_df_p53$position)

x <- randomForest(log_reporter_activity ~ ddG + background + spacing + position, 
                  data = cDNA_df_p53, ntree=100, mtry = 2, importance=TRUE)

feature_df <- data.frame(importance(x), stringsAsFactors=FALSE) %>% rownames_to_column("feature") %>% dplyr::select(-IncNodePurity)


plot_ly(feature_df, x = ~feature, y = ~X.IncMSE, type = 'bar',
             marker = list(color = '#D6D5C9',
                           line = list(color = 'rgb(8,48,107)', width = 1.5))) %>% 
  layout(title = "Random forest - factor importance plot (P53-mCMV reporters)",
         xaxis = list(title = "Feature"),
         yaxis = list(title = "%IncMSE"))


```





# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

