---
title: "Barcode processing"
author: "Max Trauernicht"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output: 
  prettydoc::html_pretty:
    theme: leonids
    highlight: github
  #   toc: true
  #   toc_float: true
  #   code_folding: show
  # editor_options:
  #   chunk_output_type: console
---

# knitr document van Steensel lab

# TF reporter barcode processing

# Introduction
18,000 TF reporters on pMT02 were transfected into mESCs, U2OS & A549, sequencing data yielded barcode counts of these experiments. These counts will be processed in this script. 

# Data processing
## Path, Libraries, Parameters and Useful Functions
```{r setup, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(data.table)
library(plyr)
library(stringr)
library(ggpubr)
library(GGally)
library(vwr)
library(dplyr)
library(tibble)
library(plotly)
library(ggbeeswarm)
library(haven)
library(readr)
library(parallel)
library(RColorBrewer)
library(gridExtra)
library(pheatmap)
```

### Custom functions
Functions used thoughout this script.
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}


# From Fede:
# ggpairs custom functions
corColor <- function(data, mapping, color = I("black"), sizeRange = c(1, 3), ...) {

  x   <- eval_data_col(data, mapping$x)
  y   <- eval_data_col(data, mapping$y)
  r   <- cor(x, y, "pairwise.complete.obs")
  rt  <- format(r, digits = 3)
  tt  <- as.character(rt)
  cex <- max(sizeRange)

  # helper function to calculate a useable size
  percent_of_range <- function(percent, range) {
    percent * diff(range) + min(range, na.rm = TRUE)
  }

  # plot correlation coefficient
  p <- ggally_text(label = tt, mapping = aes(), xP = 0.5, yP = 0.5,
                   size = I(percent_of_range(cex * abs(r), sizeRange)), color = color, ...) +
    theme(panel.grid.minor=element_blank(),
          panel.grid.major=element_blank())

  corColors <- RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")[2:6]

  if (r <= boundaries[1]) {
    corCol <- corColors[1]
  } else if (r <= boundaries[2]) {
    corCol <- corColors[2]
  } else if (r < boundaries[3]) {
    corCol <- corColors[3]
  } else if (r < boundaries[4]) {
    corCol <- corColors[4]
  } else {
    corCol <- corColors[5]
  }

  p <- p +
    theme(panel.background = element_rect(fill = corCol))

  return(p)
}
```


## Data import
```{r data import, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Import barcode counts per experiment
bc_files = list.files('/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/gcf6210/results/',
                       full.names=T, patter='*_barcode_counts.tsv')
bc_list <- lapply(bc_files, fread, header = FALSE)
names(bc_list)<- gsub('.*//6210_[0-9]{1,2}_(.*?)_[CGAT]{6}.*_barcode_counts.tsv', 
                                    '\\1', 
                                    bc_files)

# Import barcode annotation
bc_annotation <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/library_design/output/tf_df_complete.csv", header = T) %>% 
  select(barcode, tf, oligo.barcode, spacing, promoter,
         position, distance, background, affinity_pos1,
         affinity_pos2, affinity_pos3, affinity_pos4, seq.name) %>%
  setnames("seq.name", "reporter_id")
```

# Analysis

## Write bc-counts into a long df
```{r cluster_compare, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Generate long dfs
for (i in 1:length(bc_list)) {
  if (i == 1) {
  bc_df <- data.frame(bc_list[i])
  bc_df[3] <- names(bc_list[i])
  names(bc_df) <- c("barcode", "count", "name")
  bc_df <- reshape2::dcast(bc_df, barcode ~ name, value.var = "count")
  }
  else {
  bc_df_i <- data.frame(bc_list[i])
  bc_df_i[3] <- names(bc_list[i])
  names(bc_df_i) <- c("barcode", "count", "name")
  bc_df_i <- reshape2::dcast(bc_df_i, barcode ~ name, value.var = "count")
  bc_df <- merge(bc_df, bc_df_i, all = T)
  }
}

# NAs to 0
bc_df[is.na(bc_df)] <- 0

# Convert to long df
bc_df <- melt(bc_df, id.vars = "barcode",
              variable.name = "condition", value.name = "starcode_counts", as.is = T)

# Match designed barcodes
bc_df <- merge(bc_df, bc_annotation, all = T)
bc_df <- bc_df[!is.na(bc_df$condition),]

# Rename conditions
bc_df$condition <- gsub("_pMT06", "", bc_df$condition)
```


## Data exploration
```{r data inspection, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# First compute reads per million to estimate the relative counts in their respective sample
for (i in unique(bc_df$condition)) {
  bc_df$rpm[bc_df$condition == i] <- bc_df$starcode_counts[bc_df$condition == i] + 1/ # Adds a pseudocount of 1
    sum(bc_df$starcode_counts[bc_df$condition == i]) *1e6
}

# I want to show the following:
## 1: Read distribution of matched barcodes vs. unmatched barcode
ggplot(bc_df[-grep("random", bc_df$tf),], aes(x = tf, y = rpm)) +
  geom_jitter(alpha = 0.1) +
  theme_bw() +
  ylim(0,10000) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1)) +
  facet_wrap(~condition)

bc_df_2 <- bc_df[bc_df$rpm <= 500,]
bc_df_2 <- bc_df_2[bc_df_2$rpm >= 0.5,]
bc_df_2 <- bc_df_2[!is.na(bc_df_2$tf),]

ggplot(bc_df_2, aes(x = rpm)) +
  geom_histogram(binwidth = 20) +
  theme_bw() +
  xlim(0,750)+
  ylim(0,1400)+
  facet_wrap(~condition)+
  theme(strip.background =element_rect(fill="#D6D5C9"))

ggplot(bc_df[bc_df$rpm >= 500 & !is.na(bc_df$tf),], aes(x = rpm)) +
  geom_histogram(binwidth = 20) +
  theme_bw() +
  xlim(500,2000)+
  ylim(0,100)+
  facet_wrap(~condition)+
  theme(strip.background =element_rect(fill="#D6D5C9"))

n_highly_expressed <- data.frame("condition" = unique(bc_df$condition),
                                 "n_bc" = "")
for (i in unique(bc_df$condition)) {
  n_highly_expressed$n_bc[n_highly_expressed$condition == i] <- length(bc_df$barcode[bc_df$rpm > 500 & bc_df$condition == i])
}

plot_ly(n_highly_expressed, x = ~condition, y = ~as.numeric(n_bc), type = 'bar',
             marker = list(color = '#D6D5C9',
                           line = list(color = 'rgb(8,48,107)', width = 1.5))) %>% 
  layout(title = "Highly expressed barcodes",
         xaxis = list(title = "Number of barcodes with > 500 rpm"),
         yaxis = list(title = "Condition"))


## 2: How many barcodes can I find back at which cutoff? + What is the percentage of barcode reads that match the design at which cutoff?
## Identify the unmapped fraction
bc_fraction <- data.frame("condition" = unique(bc_df$condition),
                          "bcs_found" = "")
rpm_cutoff <- data.frame("cutoff" = seq(0.0001,15,0.5))
bc_fraction <- merge(bc_fraction, rpm_cutoff)

for (i in unique(bc_fraction$cutoff)) {
  for (j in unique(bc_df$condition)) {
    bc_n <- bc_df[bc_df$rpm >= i & bc_df$condition == j,]
    bc_fraction$bcs_found[bc_fraction$cutoff == i & bc_fraction$condition == j] <- nrow(bc_n[!is.na(bc_n$tf),])/
      nrow(bc_annotation) *100
  }
}



## How many reads match to designed barcodes?
bc_reads <- data.frame("condition" = unique(bc_df$condition),
                          "bc_reads" = "")
bc_reads <- merge(bc_reads, rpm_cutoff)

for (i in unique(bc_reads$cutoff)) {
  for (j in unique(bc_df$condition)) {
    bc_n <- bc_df[bc_df$rpm >= i & bc_df$condition == j,]
    bc_reads$bc_reads[bc_reads$cutoff == i & bc_reads$condition == j] <- sum(bc_n$rpm[!is.na(bc_n$tf)])/
      sum(bc_n$rpm) *100
  }
}

bc_fraction <- merge(bc_fraction, bc_reads)
bc_fraction$bcs_found <- as.numeric(bc_fraction$bcs_found)
bc_fraction$bc_reads <- as.numeric(bc_fraction$bc_reads)

#c("#1B998B", "#2D3047", "#FF9B71", "#ECDD7B")
# Plot to evaluate data quality per cutoff
ggplot(bc_fraction) +
  geom_point(aes(x = cutoff, y = bcs_found), color = '#1B998B', size = 1) +
  geom_line(aes(x = cutoff, y = bcs_found), color = '#1B998B', size = 1) +
  geom_point(aes(x = cutoff, y = bc_reads), color = 'black', size = 1) +
  geom_line(aes(x = cutoff, y = bc_reads), color = 'black', size = 1) +
  theme_bw()+
  xlab("rpm cutoff")+
  ylab("total barcodes (black) and matched barcode reads (green) detected (%)")+
  facet_wrap(~condition)+
  theme(strip.background =element_rect(fill="#D6D5C9"))

## 3: What is the correlation of the 24 cDNA bc counts with the pDNA bc counts? 
pDNA <- data.frame("pDNA" = bc_df$rpm[bc_df$condition == "pMT06_pDNA_rep1"],
                   "barcode"= bc_df$barcode[bc_df$condition == "pMT06_pDNA_rep1"])
bc_df_2 <- merge(pDNA, bc_df, all = T)

ggplot(bc_df_2, aes(x = pDNA, y = rpm)) +
  geom_bin2d(bins = 100)+
  xlim(0,400) +
  ylim(0,400)+
  theme_bw()+
  facet_wrap(~condition)

cor <- data.frame("condition" = unique(bc_df_2$condition), "cor" = "")

for (i in unique(bc_df_2$condition)) {
  x <- bc_df_2$rpm[bc_df_2$condition == i]
  y <- bc_df_2 %>% select(barcode, pDNA) %>% unique()
  cor$cor[cor$condition == i] <- cor(x, y$pDNA)
}

bc_df <- merge(bc_df, cor, by = "condition", all = T)

## 4: Correlation plots of the replicates
## Combine replicates of normalized data in 3 different columns
bc_df$rep <- gsub(".*([1-3]{1}$)","\\1",bc_df$condition)
bc_df_rep <- bc_df[!is.na(bc_df$tf),] %>% select(rep, rpm, reporter_id, condition) %>% unique()
bc_df_rep$condition <- gsub("(.*?)_rep[1-3]$", "\\1", bc_df_rep$condition)
rep1 <- bc_df_rep[bc_df_rep$rep == 1,]
rep2 <- bc_df_rep[bc_df_rep$rep == 2,]
rep3 <- bc_df_rep[bc_df_rep$rep == 3,]
rep1 <- rep1 %>% select(-rep)
rep2 <- rep2 %>% select(-rep)
rep3 <- rep3 %>% select(-rep)

names(rep1) <- c("rep1", "reporter", "condition")
names(rep2) <- c("rep2", "reporter", "condition")
names(rep3) <- c("rep3", "reporter", "condition")

bc_df_rep <- merge(rep1, rep2, all = TRUE)
bc_df_rep <- merge(bc_df_rep, rep3, all = TRUE)
bc_df_rep$neg_ctrl <- "No"
bc_df_rep$neg_ctrl[grep("random", bc_df_rep$reporter)] <- "Yes"

colors <- c("#2D3047", "#1B998B")

ggscatter(bc_df_rep, x = "rep1", y = "rep2",
   add = "reg.line",
   color = "neg_ctrl",
   size = 0.5,
   alpha = 0.2,
   add.params = list(color = "blue", fill = "lightgray"), title = "rep1 vs rep2",
   conf.int = TRUE, ylab = "rep2", xlab = "rep1") + 
  stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
  geom_abline(linetype = "dashed") +
  xlim(0,10000) + ylim(0,10000) +
  scale_color_manual(values = colors)+facet_wrap(~condition)

ggscatter(bc_df_rep, x = "rep1", y = "rep3",
   add = "reg.line",
   color = "neg_ctrl",
   size = 0.5,
   alpha = 0.2,
   add.params = list(color = "blue", fill = "lightgray"), title = "rep1 vs rep3",
   conf.int = TRUE, ylab = "rep3", xlab = "rep1") + 
  stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
  geom_abline(linetype = "dashed")+
  xlim(0,10000) + ylim(0,10000) +
  scale_color_manual(values = colors)+facet_wrap(~condition)

ggscatter(bc_df_rep, x = "rep3", y = "rep2",
   add = "reg.line",
   color = "neg_ctrl",
   size = 0.5,
   alpha = 0.2,
   add.params = list(color = "blue", fill = "lightgray"), title = "rep3 vs rep2",
   conf.int = TRUE, ylab = "rep2", xlab = "rep3") + 
  stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
  geom_abline(linetype = "dashed")+
  xlim(0,10000) + ylim(0,10000) +
  scale_color_manual(values = colors)+facet_wrap(~condition)



## 5: What is the correlation of the pDNA bc counts with the pDNA bc counts from the insert-seq?
### Highligh matched vs. unmatched with color
pDNA_insert_counts <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/pDNA_insert_seq/bc_counts.csv") %>% select(-X)
pDNA <- data.frame("pDNA" = bc_df$rpm[bc_df$condition == "pMT06_pDNA_rep1"],
                   "barcode"= bc_df$barcode[bc_df$condition == "pMT06_pDNA_rep1"],
                   "tf" = bc_df$tf[bc_df$condition == "pMT06_pDNA_rep1"])
pDNA <- merge(pDNA, pDNA_insert_counts, all = T)
pDNA$pDNA[is.na(pDNA$pDNA)] <- 0
pDNA$rpm[is.na(pDNA$rpm)] <- 0
pDNA$tf[!is.na(pDNA$tf)] <- "match"
pDNA$tf[is.na(pDNA$tf)] <- "no_match"

ggplot(pDNA, aes(x = pDNA, y = rpm, color = tf)) +
  geom_point(alpha = 0.4)+
  scale_color_brewer(palette = "Dark2")+
  xlab("bc counts insert-seq")+
  ylab("bc counts bc-seq")+
  theme_bw()

# Plot pDNA distribution vs. cutoff
pDNA_fraction <- bc_df[grep("pDNA", bc_df$condition),]
pDNA_fraction <- pDNA_fraction[!is.na(pDNA_fraction$tf),]
pDNA_fraction <- pDNA_fraction %>% select(barcode, condition, rpm)
pDNA_fraction <- dcast(pDNA_fraction, barcode ~ condition)
pDNA_fraction <- pDNA_fraction %>% 
  mutate(pDNA = (pMT06_pDNA_rep1 + pMT06_pDNA_rep2) / 2) %>%
  select(-pMT06_pDNA_rep1, -pMT06_pDNA_rep2)
pDNA_fr <- data.frame("cutoff" = seq(0,100, 1),
                      "bcs_missing" = "")
for (i in pDNA_fr$cutoff) {
  pDNA_fr$bcs_missing[pDNA_fr$cutoff == i] <- length(pDNA_fraction$barcode) -
    length(pDNA_fraction$barcode[pDNA_fraction$pDNA >= i])
}

ggplot(pDNA_fr) +
  geom_point(aes(x = cutoff, y = as.numeric(bcs_missing)), color = '#1B998B', size = 1) +
  geom_line(aes(x = cutoff, y = as.numeric(bcs_missing)), color = '#1B998B', size = 1) +
  theme_bw()+
  xlab("rpm cutoff")+
  ylab("barcodes excluded from analysis (12,000 in total)") +
  geom_vline(xintercept = 30, linetype = "dashed", color = "black")
```
## Stuff learned from above figures:
1: Most samples have high matched read counts, but some seem to have lower counts:
**High unmatched counts**:
- A549_Dex-100_rep1/3
**Low matched counts**:
- MCF7-KO_DMSO_rep1/2
- MCF7-KO_Nutlin_rep2/3 (maybe also 1)
- MCF7-WT_DMSO_rep2/3 
- MCF7-WT_Nutlin_rep3 
- mES_N2B27-HQ_rep1 (maybe also 2)
- mES_N2B27-RA_rep1/2
-> Remove everything with less than 1700 matched barcodes with more than 500 rpm's

2: Samples with high counts usually also have a high matched barcode percentage.

3: Barcode counts diverge from pDNA data for almost all samples (except MCF7-WT-DMSO-rep2, and maybe mES_N2B27-RA_rep3) - with this we can exclude that we mainly sequenced barcode from pDNA.

4: Bc counts match with insert-seq -> The GC bias is also present here. I should possibly correct for this. 

-> Based on these figures I will exclude all samples that have less than 1700 barcodes with at least 500 normalized counts.

# Filtering data
```{r}
# Remove samples with low dynamic range
for (i in unique(bc_df$condition)) {
  bc_df$n_bc[bc_df$condition == i] <- length(bc_df$barcode[bc_df$rpm > 500 & bc_df$condition == i])
}
bc_df <- bc_df[bc_df$n_bc >= 1700,] %>% select(-n_bc)

# Remove samples with high pDNA contamination
pDNA <- bc_df[grep("pDNA", bc_df$condition),] %>% select(-cor)
bc_df <- bc_df[bc_df$cor <= 0.85,] %>% select(-cor)
bc_df <- rbind(bc_df, pDNA)

# Remove all non-matching reads
bc_df <- bc_df[!is.na(bc_df$tf),]
```



## Normalization of barcode counts:
Divide cDNA barcode counts through pDNA barcode counts, if more than 30 pDNA counts for that barcode
```{r normalization, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Normalize data using pDNA data
## Compute activity by dividing through pDNA data
 n <- 0
 for (i in unique(bc_df$barcode)) {
    # Take pDNA data for each barcode
    pDNA_1 <- bc_df$rpm[bc_df$barcode == i & bc_df$condition == "pMT06_pDNA_rep1"]
    pDNA_2 <- bc_df$rpm[bc_df$barcode == i & bc_df$condition == "pMT06_pDNA_rep2"]
    pDNA <- (pDNA_1+pDNA_2)/2
    
    # cDNA data for each barcode
    cDNA <- bc_df$rpm[bc_df$barcode == i]
  
    # Compute activity by dividing cDNA rpm data by mean pDNA rpm data 
    if (pDNA >= 30){
      bc_df$activity[bc_df$barcode == i] <- cDNA / pDNA
      }
    else {
      bc_df$activity[bc_df$barcode == i] <- NA
      }
    
    # Keep track of the progress
    n <- n + length(i)
    percent <- round((n / length(unique(bc_df$barcode)))*100,2)
    progress <- paste("progress:", percent, "%")
    if (percent %% 1 == 0) {
    print(progress)
  }
 }
```







## Calculate mean activity - filter out outlier barcodes 
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# First identify and remove outlier barcodes - this removes the noise created by faulty barcode clustering etc. 
## Calculate mean and SD for each reporter
bc_df_cDNA <- bc_df[-grep("pDNA", bc_df$condition),]
bc_df_cDNA$reporter_id <- gsub(".{5}$", "", bc_df_cDNA$reporter_id)
bc_df_cDNA$reporter_activity <- ave(bc_df_cDNA$activity, bc_df_cDNA$reporter_id, 
                                bc_df_cDNA$condition, FUN =
                                  function(x) mean(x))
bc_df_cDNA$reporter_activity_sd <- ave(bc_df_cDNA$activity, bc_df_cDNA$reporter_id, 
                              bc_df_cDNA$condition,  FUN =
                                  function(x) sd(x))

## Remove data points that are 2xSD away from the mean
bc_df_cDNA$upper_activity <- bc_df_cDNA$reporter_activity + (2 * bc_df_cDNA$reporter_activity_sd)
bc_df_cDNA$lower_activity <- bc_df_cDNA$reporter_activity - (2 * bc_df_cDNA$reporter_activity_sd)
bc_df_cDNA_filt <- bc_df_cDNA[bc_df_cDNA$activity <= bc_df_cDNA$upper_activity & 
                                bc_df_cDNA$activity >= bc_df_cDNA$lower_activity,]

## Recalculate mean and sd
bc_df_cDNA_filt$reporter_activity <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id, 
                                bc_df_cDNA_filt$condition, FUN =
                                  function(x) mean(x))
bc_df_cDNA_filt$reporter_activity_sd <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id, 
                              bc_df_cDNA_filt$condition,  FUN =
                                  function(x) sd(x))
```







## Filter out data using pDNA-insert-seq data
## pDNA data seems ambigous - data doesn't match sometimes - exclude for now
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# # Remove barcodes with multiple inserts attached
# bc_exclude <- read.csv("/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/pDNA_seq/bc_exclude.csv") %>% select(-X) %>% setnames("x", "barcode")
# exclude <- bc_df_cDNA_filt[!bc_df_cDNA_filt$barcode %in% bc_exclude$barcode,]
# 
# # Reassign barcodes with wrong inserts attached
# bc_replace <- read.csv("/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/pDNA_seq/bc_replace.csv") %>% select(-X, -bc.match)
# change.df <- exclude[exclude$barcode %in% bc_replace$barcode,]
# exclude <- exclude[!exclude$barcode %in% bc_replace$barcode,]
# 
# change.df$bc.number <- 9
# change.df <- merge(change.df, bc_replace)
# e <- c("e11", "e93", "e97")
# change.df.e <- change.df[grep(paste(e, collapse = "|"), change.df$insert.match),]
# change.df <- change.df[-grep(paste(e, collapse = "|"), change.df$insert.match),]
# change.df <- change.df %>%
#   mutate(TF = gsub("(^.*?)_.*", "\\1", insert.match),
#          promoter = gsub(".*(minP|mCMV|hBGm|Random)_.*", "\\1", insert.match),
#          distance = gsub(".*_d-([0-9]{1,2}bp)_.*", "\\1", insert.match),
#          spacing = gsub(".*_s-([0-9]{1,2}bp)_.*", "\\1", insert.match),
#          background = gsub(".*([0-9]{1}$)", "\\1", insert.match),
#          reporter_id = paste(TF, spacing, distance, promoter, background, sep = "_"))
# change.df <- change.df %>% select(-insert.match)
# 
# change.df.e <- change.df.e %>%
#   mutate(TF = gsub("(.*?)_(minP|mCMV|hBGm|Random)$", "\\1", insert.match),
#          promoter = gsub(".*(minP|mCMV|hBGm|Random)", "\\1", insert.match),
#          spacing = "",
#          distance = "",
#          background = "",
#          reporter_id = paste(TF, promoter, sep ="_"))
# change.df.e <- change.df.e %>% select(-insert.match)
# 
# 
# exclude$reporter_id <- gsub("___", "_", exclude$reporter_id)
# exclude$reporter_id <- gsub("_0", "", exclude$reporter_id)
# 
# 
# 
# exclude <- rbind(exclude, change.df, change.df.e)
# 
# bc_df_cDNA_filt <- exclude
```





## Add pDNA data
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
pDNA <- bc_df[grep("pMT06_pDNA_rep1", bc_df$condition),] %>% 
  select(barcode, rpm) %>% unique() %>%
  setnames("rpm", "pDNA_counts_rpm") %>%
  mutate(pDNA_counts_rpm = ave(pDNA_counts_rpm, barcode, FUN = function(x) mean(x))) %>%
  unique()
bc_df_cDNA_filt <- merge(pDNA, bc_df_cDNA_filt, all = T)
#bc_df_cDNA_filt <- bc_df_cDNA_filt[!is.na(bc_df_cDNA_filt$condition),]
```




## Annotate controls
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# Annotate the mutated motif of each TF
bc_df_cDNA_filt$neg_ctrls <- "No"
bc_df_cDNA_filt$neg_ctrls[grep("random", bc_df_cDNA_filt$tf)] <- "Yes"

# Annotate random promoter control
bc_df_cDNA_filt$rand_promoter <- "No"
bc_df_cDNA_filt$rand_promoter[grep("random", bc_df_cDNA_filt$promoter)] <- "Yes"
```


## Scale data to 1, use negative controls as reference
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
mean_neg_ctrl <- mean(bc_df_cDNA_filt$activity[bc_df_cDNA_filt$neg_ctrls == "Yes"])
bc_df_cDNA_filt$activity <- bc_df_cDNA_filt$activity / mean_neg_ctrl
bc_df_cDNA_filt$mean_activity <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id, 
                                bc_df_cDNA_filt$condition, FUN =
                                  function(x) mean(x))
```



## Calculate correlations between technical replicates
```{r correlations_2, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
## Combine replicates in 8 different columns
bc_df_rep <- bc_df_cDNA_filt[bc_df_cDNA_filt$rand_promoter == "No",] %>% 
  select(oligo.barcode, activity, condition, reporter_id)
rep1 <- bc_df_rep[bc_df_rep$oligo.barcode == 1,] %>% select(-oligo.barcode)
rep2 <- bc_df_rep[bc_df_rep$oligo.barcode == 2,] %>% select(-oligo.barcode)
rep3 <- bc_df_rep[bc_df_rep$oligo.barcode == 3,] %>% select(-oligo.barcode)
rep4 <- bc_df_rep[bc_df_rep$oligo.barcode == 4,] %>% select(-oligo.barcode)
rep5 <- bc_df_rep[bc_df_rep$oligo.barcode == 5,] %>% select(-oligo.barcode)



setnames(rep1, "activity", "bc1")
setnames(rep2, "activity", "bc2")
setnames(rep3, "activity", "bc3")
setnames(rep4, "activity", "bc4")
setnames(rep5, "activity", "bc5")



bc_df_rep <- merge(rep1, rep2, all = T)
bc_df_rep <- bc_df_rep[rowSums(is.na(bc_df_rep)) != ncol(bc_df_rep), ]
bc_df_rep <- merge(bc_df_rep, rep3, all = T)
bc_df_rep <- merge(bc_df_rep, rep4, all = T)
bc_df_rep <- bc_df_rep[rowSums(is.na(bc_df_rep)) != ncol(bc_df_rep), ]
bc_df_rep <- merge(bc_df_rep, rep5, all = T)



# Correlation matrix plot
n <- sample(1:nrow(bc_df_rep), 5000)
boundaries <- seq(from = 0.8, by = 0.05, length.out = 4)
plt <- ggpairs(bc_df_rep %>% select(bc1, bc2, bc3, bc4, bc5),
               upper = list(continuous = corColor),
               lower = list(continuous = function(data, mapping, ...) {
                   ggally_points(data = data[n, ], mapping = mapping, alpha = 0.1, size = 0.5) +
                   geom_abline(slope = 1, lty = "dashed", col = "red") +
                   theme_bw()}),
               diag = list(continuous = function(data, mapping, ...) {
                   ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red") +
                   theme_bw()})) +
  ggtitle("Correlation Between Technial Replicates") +
  theme(text = element_text(size = 20)) +
  xlab("Reporter activity") +
  ylab("Reporter activity") + 
  theme_light()

print(plt)
```



## Data quality plots - correlation between replicates
```{r correlations_3, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# Correlation plots of the replicates
## Combine replicates of normalized data in 3 different columns
bc_df_rep <- bc_df_cDNA_filt %>% select(rep, reporter_activity, reporter_id, condition) %>% unique()
bc_df_rep$condition <- gsub("(.*?)_rep[1-3]$", "\\1", bc_df_rep$condition)
rep1 <- bc_df_rep[bc_df_rep$rep == 1,]
rep2 <- bc_df_rep[bc_df_rep$rep == 2,]
rep3 <- bc_df_rep[bc_df_rep$rep == 3,]
rep1 <- rep1 %>% select(-rep)
rep2 <- rep2 %>% select(-rep)
rep3 <- rep3 %>% select(-rep)

names(rep1) <- c("rep1", "reporter", "condition")
names(rep2) <- c("rep2", "reporter", "condition")
names(rep3) <- c("rep3", "reporter", "condition")

bc_df_rep <- merge(rep1, rep2, all = TRUE)
bc_df_rep <- merge(bc_df_rep, rep3, all = TRUE)
bc_df_rep$neg_ctrl <- "No"
bc_df_rep$neg_ctrl[grep("neg", bc_df_rep$reporter)] <- "Yes"

colors <- c("#2D3047", "#1B998B")

ggscatter(bc_df_rep, x = "rep1", y = "rep2",
   add = "reg.line",
   color = "neg_ctrl",
   size = 0.5,
   add.params = list(color = "blue", fill = "lightgray"), title = "rep1 vs rep2",
   conf.int = TRUE, ylab = "rep2", xlab = "rep1") + 
  stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
  geom_abline(linetype = "dashed") +
  xlim(0,10) + ylim(0,10) +
  scale_color_manual(values = colors)+facet_wrap(~condition)

ggscatter(bc_df_rep, x = "rep1", y = "rep3",
   add = "reg.line",
   color = "neg_ctrl",
   size = 0.5,
   add.params = list(color = "blue", fill = "lightgray"), title = "rep1 vs rep3",
   conf.int = TRUE, ylab = "rep3", xlab = "rep1") + 
  stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
  geom_abline(linetype = "dashed")+
  xlim(0,10) + ylim(0,10) +
  scale_color_manual(values = colors)+facet_wrap(~condition)

ggscatter(bc_df_rep, x = "rep3", y = "rep2",
   add = "reg.line",
   color = "neg_ctrl",
   size = 0.5,
   add.params = list(color = "blue", fill = "lightgray"), title = "rep3 vs rep2",
   conf.int = TRUE, ylab = "rep2", xlab = "rep3") + 
  stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
  geom_abline(linetype = "dashed")+
  xlim(0,10) + ylim(0,10) +
  scale_color_manual(values = colors)+facet_wrap(~condition)
```



## Summarize replicates: summmarize activity values between biological replicates by their geometric mean
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# # Create similarity heatmap
# bc_df_cDNA_filt <- bc_df_cDNA_filt[!is.na(bc_df_cDNA_filt$condition),]
# cor_heatmap <- bc_df_cDNA_filt %>% select(barcode, condition, activity)
# cor_heatmap <- dcast(cor_heatmap, barcode~condition) %>% column_to_rownames("barcode")
# cor_heatmap <- round(cor(cor_heatmap, use = "pairwise.complete.obs"),2)
# cor_heatmap_mcf <- cor_heatmap[grep("MCF", colnames(cor_heatmap)),grep("MCF", rownames(cor_heatmap))]
# cor_heatmap_a549 <- cor_heatmap[grep("A549", colnames(cor_heatmap)),grep("A549", rownames(cor_heatmap))]
# cor_heatmap_mES <- cor_heatmap[grep("mES", colnames(cor_heatmap)),grep("mES", rownames(cor_heatmap))]
# pheatmap(cor_heatmap)
# pheatmap(cor_heatmap_mcf)
# pheatmap(cor_heatmap_a549)
# pheatmap(cor_heatmap_mES)


# Mean of the three replicates
bc_df_cDNA_filt$condition <- gsub("(.*?)_rep[1-3]$", "\\1", bc_df_cDNA_filt$condition)
bc_df_cDNA_filt$reporter_activity <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id, bc_df_cDNA_filt$condition, FUN = function(x) mean(x))
bc_df_cDNA_filt$reporter_activity_sd <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id, bc_df_cDNA_filt$condition, FUN = function(x) sd(x))
```


## Exporting data
```{r data export, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# Polish export dataframe
bc_df_cDNA_filt <- bc_df_cDNA_filt %>% 
  select(-upper_activity, -lower_activity) %>% 
  mutate(log_activity = log2(activity),
         log_reporter_activity = log2(reporter_activity))


# Export bc_df for cDNA analysis
filename <- SetFileName("_reporter_activity_filt", "mt")
setwd("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/gcf6210/results/")
write.csv(bc_df_cDNA_filt, file = paste(filename,".csv", sep = ""), row.names = F)
```

# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

